---
title: "Patho FFPE human tumor vs. normal pancreas (AS1000Cov50, physeq_original)"
author: "Nina & Christoph"
date: "Report created: `r Sys.Date()`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
editor_options: 
  chunk_output_type: console
---
Adapter by Linh

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

In this experiment, human tumor vs. normal pancreas FFPE tissue samples from the Pathology department are examined. The DNA had been isolated with the Invitrogen FFPE kit (by me). The sample size is n=24 for both tumor and normal pancreas (matched patients). Additionally, n=6 FFPE tumor tissue samples from the establishment experiment (patho_ffpe_self) without matched normal pancreas tissue are implemented (so we have n=30 tumor, n=24 normal pancreas). n=75 NTCs (n=23 paraffin tumor, n=24 paraffin normal, n=10 buffer-only, n=7 PCR CTRLs) plus n=11 NTCs from the establishment experiment (patho_ffpe_self) are implemented. 

```{r, echo=FALSE}
library(PERFect)
library(microbiome)
library(phyloseq)
library(decontam)
library(tidyverse)
library(pairwiseAdonis)
library(kableExtra)
library(vegan)
library(tidytree)
library(ape)
library(ggrepel)
library(ggpubr) # for manual p-value assigning
library(rstatix) # for wilcox test
```

# Combining Phyloseq objects {.tabset}

## Loading all physeq-objects (AS1000Cov50, physeq_original)  

Run by run was separately analysed and a phyloseq object per run was created. By using a barplot for microbial composition and the beta diversity (weighted UF) outliers are detected per each run. Every outlier is categorized and a reason is described, why we identify this sample as an outlier. 

```{r}
Path2PhyloSeqObj <- "./raw_data/PhyloSeqObj/"
physeq1 <- readRDS(paste0(Path2PhyloSeqObj,"2022_05_27_patho_FFPE_tum_panc_1/physeq_original_50.rds"))
physeq2 <- readRDS(paste0(Path2PhyloSeqObj,"2022_05_27_patho_FFPE_tum_panc_2/physeq_original_50.rds"))
physeq3 <- readRDS(paste0(Path2PhyloSeqObj,"2022_05_28_patho_FFPE_tum_panc_3/physeq_original_50.rds"))
physeq4 <- readRDS(paste0(Path2PhyloSeqObj,"2022_05_28_patho_FFPE_tum_panc_4/physeq_original_50.rds"))

# sample_data(physeq1)
# sample_data(physeq2)
# sample_data(physeq3)
# sample_data(physeq4)

p.true.original <- merge_phyloseq(physeq1,physeq2,physeq3,physeq4)

```

```{r}
nrow(sample_data(p.true.original))
nrow(otu_table(p.true.original))
```


Now, we have **`r nrow(sample_data(p.true.original))` samples and `r nrow(otu_table(p.true.original))` taxa**. 


## Implement samples from establishment experiment (patho_ffpe_self)

n=6 FFPE tumor tissue samples (without matching normal pancreas).

```{r}
physeq5 <- readRDS(paste0(Path2PhyloSeqObj,"2021_12_01_patho_FFPE_self_tumor/physeq_original.rds"))

# sample_data(physeq5)

physeq5sub <- subset_samples(physeq5, sample_nr!="P19269-20" & sample_nr!="P15970-20" & sample_nr!="P12758-20" & sample_nr!="P3434-20")

# sample_data(physeq5sub)

p.true.original <- merge_phyloseq(p.true.original,physeq5sub)

# sample_data(physeq)

```

## Implementing additional phyloseq objects (repeaters)

XX.XX.22: 
XX.XX.22: 
XX.XX.22: 

```{r}
# physeq6 <- readRDS("")
# physeq7 <- readRDS("")
# physeq8 <- readRDS("")
# 
# physeq <- merge_phyloseq(physeq,physeq6,physeq7,physeq8)

```

Now, we have **`r nrow(sample_data(p.true.original))` samples and `r nrow(otu_table(p.true.original))` taxa**.

## Identifying Outliers 

Here 2 main groups of outliers are defined: without_antibiotic and with antibiotic treatment besides the outliers excluded because of a technical error.
Details without_ab:  
1. oral_flora = higher abundance of oral microorganism in rectal swab samples
2. slight_dermal_cont = some of the higher abundant species belong to residual dermal flora indicating a possible not sufficient rectal swab
3. dermal_cont = most high abundant species (> 33%) belong to dermal residual flora indicating a not sufficiant rectal swab
4. dom_1_spec = One species dominate the microbial composition (> 50%), e.g. E. coli indicating a storage of the sample for longer time (> 3h)
5. untypical_spec = High abundant species which are not typically found in other samples (1% of the samples)
6. vaginal_flora = higher abundance of vaginal microorganism in rectal swab samples


# Normalization {.tabset}

## Overview 

In order to normalize the data we use used 5 methods:  

1. Filtering for bacterial reads
2. Deleting all taxIDs without resolution to species level (tax_glom)  
3. Prevalence filtering (PERfect package) ## you mean abundant filtering?  
4. Decontamination (decontam package)    
5. Rarefaction (phyloseq package)   


The first one was already applied during "phyloseq-object-creation" for every single run. Hereby, less abundant taxa were considered as technical/ bioinformatical mistakes rather than true species. The underlying algorithm was the simultanous method. Here we apply the PERfect packge again to eliminate more taxa.  

To decontaminate the TRUE samples decontam is applied. The Negative controls derived from DNA-extraction (blank swab medium was extracted) and library preparation for both approaches (16S rRNA and metagenomics).  

Due to different sequencing depth all samples have to be rarified. To this end rarefaction is used. Details below. Before we filtering all reads which do not contain bacterial reads (also already done in "phyloseq-object-creation"). Furthermore all TaxIDs are removed which do not have a resolution until species level (also already done in "phyloseq-object-creation"). Step 3-4 are conducted to have all normalization steps included, though their impact of the data is 0, because they were done before.

## Filtering

```{r}
p.true.filt <- subset_taxa(p.true.original, superkingdom=="Bacteria")
p.true.filt.s <- tax_glom(p.true.filt, taxrank = "species", NArm = TRUE)
```

Now we have `r nrow(sample_data(p.true.filt.s))` samples and `r nrow(otu_table(p.true.filt.s))` taxa.


## Prevalence filtering

```{r}
otu = as(otu_table(p.true.filt.s), "matrix")
if(taxa_are_rows(p.true.filt.s)){otu <- t(otu)}
otu = as_tibble(otu)
res_sim <- PERFect_sim(X = otu)
# dim(res_sim$filtX) 
ids.sim<- colnames(res_sim$filtX) 
p.true.filt.s.prev <-  prune_taxa(ids.sim, p.true.filt.s)
```

Now we have `r nrow(sample_data(p.true.filt.s.prev))` samples and `r nrow(sample_data(p.true.filt.s.prev))` taxa. 
```{r}
nrow(sample_data(p.true.filt.s.prev))
nrow(otu_table(p.true.filt.s.prev))
```

## Decontam {.tabset}

The negative controls are merged and all three steps of the normalization process are performed as well.

```{r}

physeq.nct1 <- readRDS(paste0(Path2PhyloSeqObj,"2022_06_02_patho_FFPE_tum_panc_ntc_1/physeq_original_50.rds"))
physeq.nct2 <- readRDS(paste0(Path2PhyloSeqObj,"2022_06_02_patho_FFPE_tum_panc_ntc_2/physeq_original_50.rds"))
physeq.nct3 <- readRDS(paste0(Path2PhyloSeqObj,"2022_06_03_patho_FFPE_tum_panc_ntc_3/physeq_original_50.rds"))
physeq.nct4 <- readRDS(paste0(Path2PhyloSeqObj,"2022_06_03_patho_FFPE_tum_panc_ntc_4/physeq_original_50.rds"))
physeq.nct5 <- readRDS(paste0(Path2PhyloSeqObj,"2022_06_04_patho_FFPE_tum_panc_ntc_5/physeq_original_50.rds"))
physeq.nct6 <- readRDS(paste0(Path2PhyloSeqObj,"2022_06_04_patho_FFPE_tum_panc_ntc_6/physeq_original_50.rds"))
physeq.nct7 <- readRDS(paste0(Path2PhyloSeqObj,"2022_06_10_patho_FFPE_tum_panc_ntc_7/physeq_original_50.rds"))
physeq.nct8 <- readRDS(paste0(Path2PhyloSeqObj,"2022_07_13_patho_FFPE_tum_panc_ntc_8/physeq_original_50.rds"))
physeq.nct9 <- readRDS(paste0(Path2PhyloSeqObj,"2022_07_13_patho_FFPE_tum_panc_ntc_9/physeq_original_50.rds"))
physeq.nct10 <- readRDS(paste0(Path2PhyloSeqObj,"2021_12_02_patho_FFPE_self_ntc/physeq_original.rds"))
physeq.nct11 <- readRDS(paste0(Path2PhyloSeqObj,"2022_01_24_patho_FFPE_self_ntc_2/physeq_original.rds"))


physeq.nct.original <- merge_phyloseq(physeq.nct1, physeq.nct2, physeq.nct3, physeq.nct4, physeq.nct5, physeq.nct6, physeq.nct7, physeq.nct8, physeq.nct9, physeq.nct10, physeq.nct11)

# sample_data(physeq.nct)$sample_side <- ifelse(sample_data(physeq.nct)$sample_side == "paraffin", "paraffin_tumor", sample_data(physeq.nct)$sample_side)

physeq.filt.nct <- subset_taxa(physeq.nct.original, superkingdom=="Bacteria")
physeq.filt.nct.s <- tax_glom(physeq.filt.nct, taxrank = "species", NArm = TRUE)

otu = as(otu_table(physeq.filt.nct.s), "matrix")
if(taxa_are_rows(physeq.filt.nct.s)){otu <- t(otu)}
otu = as_tibble(otu)
res_sim <- PERFect_sim(X = otu)
dim(res_sim$filtX)
ids.sim<- colnames(res_sim$filtX)
physeq.filt.nct.s.prev <-  prune_taxa(ids.sim, physeq.filt.nct.s)

physeq.filt.combined <- merge_phyloseq(p.true.filt.s.prev, physeq.filt.nct.s.prev)
sample_data(physeq.filt.combined)$is.neg <- sample_data(physeq.filt.combined)$TRUE_control == "control"#first identify your negative controls 
```

Now we have `r nrow(sample_data(physeq.filt.combined))` samples and `r nrow(otu_table(physeq.filt.combined))` taxa. 


### Inspect the library size

The following plots show the library size, whereas the point color separate negative controls and TRUE samples.

```{r}
df <- as.data.frame(sample_data(physeq.filt.combined)) # Put sample_data into a ggplot-friendly data.frame

true_count <- as_tibble(sample_data(physeq.filt.combined)) %>% filter(TRUE_control == "TRUE") %>% count(TRUE_control) %>% pull(n)
control_count <- as_tibble(sample_data(physeq.filt.combined)) %>% filter(TRUE_control == "control") %>% count(TRUE_control) %>% pull(n)

df$LibrarySize <- sample_sums(physeq.filt.combined)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
ggplot(data=df, aes(x=Index, y=LibrarySize, color=TRUE_control)) + geom_point()  + theme_bw() + ggtitle("Library size")+
  scale_color_brewer(palette = "Set1", labels=c(paste("Negative controls\nn =", control_count, sep = " "), paste("Tissue samples\nn =", true_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12),
        legend.title = element_blank())
```

### Beta diversity negativ controls vs true samples

```{r}
bray_dist = phyloseq::distance(physeq.filt.combined, method="bray")
bc_ordination = ordinate(physeq.filt.combined, method="PCoA", distance=bray_dist)

true_count <- as_tibble(sample_data(physeq.filt.combined)) %>% filter(TRUE_control == "TRUE") %>% count(TRUE_control) %>% pull(n)
control_count <- as_tibble(sample_data(physeq.filt.combined)) %>% filter(TRUE_control == "control") %>% count(TRUE_control) %>% pull(n)

pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")

set.seed(40)
p.adonis <- adonis2(bray_dist ~ sample_data(physeq.filt.combined)$TRUE_control)

p <- case_when(
  p.adonis$`Pr(>F)`[1] > 0.05 ~ "n.s.",
  p.adonis$`Pr(>F)`[1] < 0.05 &  p.adonis$`Pr(>F)`[1] > 0.01 ~ paste(p.adonis$`Pr(>F)`[1], "*", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.01 & p.adonis$`Pr(>F)`[1] > 0.001  ~ paste(p.adonis$`Pr(>F)`[1], "**", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.001 ~ paste(p.adonis$`Pr(>F)`[1], "***", sep = " "),
)
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2) ,
        vjustvar = c(1.5))

plot_ordination(physeq.filt.combined, bc_ordination, color="TRUE_control") + 
  geom_point(aes(colour=TRUE_control), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_color_brewer(palette = "Set1", labels=c(paste("Tissue samples\nn =", true_count, sep = " "), paste("Negative controls\nn =", control_count, sep = " ")))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("Differences Negative controls and True samples - Bray curtis - Not rarified") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12),
        legend.title = element_blank())+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)


sample.size <- 1000
seed <- 711
physeq.filt.combined2 <- rarefy_even_depth(physeq = physeq.filt.combined, sample.size = sample.size, rngseed = seed)

true_count <- as_tibble(sample_data(physeq.filt.combined2)) %>% filter(TRUE_control == "TRUE") %>% count(TRUE_control) %>% pull(n)
control_count <- as_tibble(sample_data(physeq.filt.combined2)) %>% filter(TRUE_control == "control") %>% count(TRUE_control) %>% pull(n)

bray_dist = phyloseq::distance(physeq.filt.combined2, method="bray")
bc_ordination = ordinate(physeq.filt.combined2, method="PCoA", distance=bray_dist)

pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")

set.seed(40)
p.adonis <- adonis2(bray_dist ~ sample_data(physeq.filt.combined2)$TRUE_control)

p <- case_when(
  p.adonis$`Pr(>F)`[1] > 0.05 ~ "n.s.",
  p.adonis$`Pr(>F)`[1] < 0.05 &  p.adonis$`Pr(>F)`[1] > 0.01 ~ paste(p.adonis$`Pr(>F)`[1], "*", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.01 & p.adonis$`Pr(>F)`[1] > 0.001  ~ paste(p.adonis$`Pr(>F)`[1], "**", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.001 ~ paste(p.adonis$`Pr(>F)`[1], "***", sep = " "),
)
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2) ,
        vjustvar = c(1.5))

plot_ordination(physeq.filt.combined2, bc_ordination, color="TRUE_control") + 
  geom_point(aes(colour=TRUE_control), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
scale_color_brewer(palette = "Set1", labels=c(paste("Tissue samples\nn =", true_count, sep = " "), paste("Negative controls\nn =", control_count, sep = " ")))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("Differences Negative controls and True samples - Bray curtis - rarified") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12),
        legend.title = element_blank())+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)
```

### Beta diversity within negative controls

```{r}
partum_count <- as_tibble(sample_data(physeq.filt.nct.s.prev)) %>% filter(sample_side == "paraffin_tumor") %>% count(sample_side) %>% pull(n)
parpanc_count <- as_tibble(sample_data(physeq.filt.nct.s.prev)) %>% filter(sample_side == "paraffin_normal") %>% count(sample_side) %>% pull(n)
buffer_count <- as_tibble(sample_data(physeq.filt.nct.s.prev)) %>% filter(sample_side == "buffer") %>% count(sample_side) %>% pull(n)
h2o_count <- as_tibble(sample_data(physeq.filt.nct.s.prev)) %>% filter(sample_side == "PCR_ctrl_H2O") %>% count(sample_side) %>% pull(n)
par_count <- as_tibble(sample_data(physeq.filt.nct.s.prev)) %>% filter(sample_side == "paraffin") %>% count(sample_side) %>% pull(n)

bray_dist = phyloseq::distance(physeq.filt.nct.s.prev, method="bray")
bc_ordination = ordinate(physeq.filt.nct.s.prev, method="PCoA", distance=bray_dist)

pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")

set.seed(40)
p.adonis <- adonis2(bray_dist ~ sample_data(physeq.filt.nct.s.prev)$sample_side)

p <- case_when(
  p.adonis$`Pr(>F)`[1] > 0.05 ~ "n.s.",
  p.adonis$`Pr(>F)`[1] < 0.05 &  p.adonis$`Pr(>F)`[1] > 0.01 ~ paste(p.adonis$`Pr(>F)`[1], "*", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.01 & p.adonis$`Pr(>F)`[1] > 0.001  ~ paste(p.adonis$`Pr(>F)`[1], "**", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.001 ~ paste(p.adonis$`Pr(>F)`[1], "***", sep = " "),
)
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2) ,
        vjustvar = c(1.5))


plot_ordination(physeq.filt.nct.s.prev, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
scale_color_brewer(palette = "Set1", labels=c(paste("DNA Ext. buffer\nn =", buffer_count, sep = " "), paste("Paraffin\nn =", par_count, sep = " "), paste("Paraffin pancreas\nn =", parpanc_count, sep = " "), paste("Paraffin tumor\nn =", partum_count, sep = " "), paste("PCR H2O\nn =", h2o_count, sep = " ")))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("Differences Negative controls and True samples - Bray curtis - rarified") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12),
        legend.title = element_blank())+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

ps.disper <- betadisper(bray_dist, as.matrix(sample_data(physeq.filt.nct.s.prev)$sample_side))
permutest(ps.disper, pair=TRUE)
```

Perfect! That is the first proof that there might be a true signal in the tissue samples compared to negative controls

### Identify Contaminants - Prevalence  

Now we will apply decontam by using the prevalence method. 

```{r}
threshold <- 0.5
contamdf.prev <- isContaminant(physeq.filt.combined, method="prevalence", neg="is.neg", threshold = threshold)
cont <- as.data.frame(table(contamdf.prev$contaminant))
contprev <- contamdf.prev %>% filter (contaminant==TRUE)
contprev <- contprev %>% mutate (taxID=rownames(contprev))
which(contamdf.prev$contaminant)
length(which(contamdf.prev$contaminant))
```

Hereby **`r cont[2,2]` taxa are considered as contaminants** and will be removed subsequently. As the matrix above point out, a lot of high abundant taxa will be removed. **The `r (which(contamdf.prev$contaminant))[1]`th most abundant taxon** is the most prevalent taxon which will be removed. The threshold of `r threshold` is determined with the following histogram.

```{r}
hist(contamdf.prev$p, 100, ylim = c(0,300), xlim = c(0,1))
```

We will control the "definition of contaminants" by looking to the distribution of contaminated taxa within true and control samples.


```{r}
# Make phyloseq object of presence-absence in negative controls and true samples
ps.pa <- transform_sample_counts(physeq.filt.combined, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$TRUE_control == "control", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$TRUE_control == "TRUE", ps.pa)
# Make data.frame of prevalence in positive and negative samples
df.pa <- data.frame(pa.pos=taxa_sums(ps.pa.pos), pa.neg=taxa_sums(ps.pa.neg),
                      contaminant=contamdf.prev$contaminant)
ggplot(data=df.pa, aes(x=pa.neg, y=pa.pos, color=contaminant)) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)") + ggtitle("Contaminants prevalence in tumor vs NTCs") + theme_bw()
```

It seems that we identify really contaminants! 


```{r}
physeq <- prune_taxa(!contamdf.prev$contaminant, physeq.filt.combined)
physeq <- subset_samples(physeq, is.neg ==FALSE )
```

## Rarefaction {.tabset}

### Sequencing depth

First we give us an overview how many bacterial reads we have per sample to determine a reasonable threshold. 
```{r}
bac.count <- as_tibble(colSums(as.data.frame(otu_table(physeq))))%>% mutate (id=sample_data(physeq)$id)
sum.bac.count <- bac.count %>% summarise(mean=mean(value),
                                          min=min(value),
                                          max=max(value),
                                          median=median(value)) 

knitr::kable (sum.bac.count, caption="Sequencing summary", booktabs =TRUE) %>%
  kable_styling(font_size = 15)

bac.count %>% 
  arrange(value) %>% 
  kable (caption="Bacterial reads", booktabs =TRUE) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>% 
  kable_classic(full_width = F, html_font = "Cambria") %>%
  scroll_box(width = "100%", height = "600px")

bac.count %>% 
  filter(value<= 1000) %>% 
  arrange(value) %>% 
  kable (caption="Bacterial reads", booktabs =TRUE) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>% 
  kable_classic(full_width = F, html_font = "Cambria") %>%
  scroll_box(width = "100%", height = "600px")
```

According to this summary all samples are above `r sum.bac.count$min` reads. 

## Rarifying


```{r}
# sample.size <- sum.bac.count$min
sample.size <- 1000
seed <- 711
physeq.rar <- rarefy_even_depth(physeq = physeq, sample.size = sample.size, rngseed = seed)
rarecurve(t(otu_table(physeq.rar)), step=50, cex=0.5)

phy_tree = rtree(ntaxa(physeq.rar), rooted=TRUE, tip.label=taxa_names(physeq.rar))
physeq.rar  <- merge_phyloseq(physeq.rar , phy_tree)
```

For rarifaction `r seed` seed and a threshold of  **`r format(sample.size, scientific=FALSE)`** was used.
Now we have `r nrow(sample_data(physeq.rar))` samples and `r nrow(otu_table(physeq.rar))` taxa. 


# Results {.tabset}  

## Alpha diversity 

```{r}
rich = estimate_richness(physeq.rar)
rich$id <- gsub("X","", rownames(rich))
rownames(rich) <- rich$id
rich_meta <- merge(sample_data(physeq.rar),rich, by="row.names")
rich_meta %>% group_by(sample_side) %>% count()

#Shapiro-Wilk
sh.ob <- rich_meta %>% shapiro_test(Observed)
sh.sh <- rich_meta %>% shapiro_test(Shannon)
sh.in <- rich_meta %>% shapiro_test(InvSimpson)

sh.ob
sh.sh
sh.in

# #Levene Test
l.ob <- rich_meta %>% levene_test(Observed ~ sample_side)
l.sh <- rich_meta %>% levene_test(Shannon ~ sample_side)
l.in <- rich_meta %>% levene_test(InvSimpson ~ sample_side)

l.ob
l.sh
l.in

##Shannon data is normally distributed according to Shapiro-Wilk and Levene (p>0.05), therefore, we can apply a t-test
##Observed and InvSimpson are not normally distributed according to Shapiro-Wilk (p<0.05), therefore, we have to apply the Wilcoxon test


# # stat-test
ob <- rich_meta %>% wilcox_test(Observed ~ sample_side) %>% adjust_pvalue(method = "BH") %>%  add_significance("p.adj") %>% add_xy_position()
sh <- rich_meta %>% t_test(Shannon ~ sample_side) %>% adjust_pvalue(method = "BH") %>%  add_significance("p.adj") %>% add_xy_position()
iv <- rich_meta %>% wilcox_test(InvSimpson ~ sample_side) %>% adjust_pvalue(method = "BH") %>%  add_significance("p.adj") %>% add_xy_position()

ob
sh
iv

group.colors <- c(normal = "dodgerblue3", tumor = "firebrick2")
my.labels <- c("Normal pancreas", "Tumor tissue")

#Observed species
ggplot (rich_meta, aes (x=sample_side, y=Observed, fill=sample_side))+ 
  geom_boxplot()+
  geom_point (position=position_jitterdodge( jitter.width = 0.05))+
  theme_gray() + 
  xlab("")+
  # facet_grid(.~tp)+
  scale_x_discrete(labels= my.labels)+
  scale_fill_manual(values=group.colors, labels = c("\nNormal pancreas\nn = 15\n", "\nTumor tissue\nn = 25\n"))+
  ggtitle("FFPE human tumor tissue vs. normal pancreas - Observed species") +
  theme(axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.text = element_text(size = 12),
        legend.title = element_blank(),
        # legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"),
        axis.text.x = element_text(size=12, hjust = .5, angle=0, vjust = .5),
        plot.title = element_text(size = 14))+
   stat_pvalue_manual(ob, label = "p.adj.signif", inherit.aes = FALSE, tip.length = 0.01)

#Shannon index
ggplot (rich_meta, aes (x=sample_side, y=Shannon, fill=sample_side))+ 
  geom_boxplot()+
  geom_point (position=position_jitterdodge( jitter.width = 0.05))+
  theme_gray() + 
  xlab("")+
  # facet_grid(.~tp)+
  scale_x_discrete(labels= my.labels)+
  scale_fill_manual(values=group.colors, labels = c("\nNormal pancreas\nn = 15\n", "\nTumor tissue\nn = 25\n"))+
  ggtitle("FFPE human tumor tissue vs. normal pancreas - Shannon index") +
  theme(axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.text = element_text(size = 12),
        legend.title = element_blank(),
        # legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"),
        axis.text.x = element_text(size=12, hjust = .5, angle=0, vjust = .5),
        plot.title = element_text(size = 14))+
   stat_pvalue_manual(sh, label = "p.adj.signif", inherit.aes = FALSE, tip.length = 0.01)

#InvSimpson index
ggplot (rich_meta, aes (x=sample_side, y=InvSimpson, fill=sample_side))+ 
  geom_boxplot()+
  geom_point (position=position_jitterdodge( jitter.width = 0.05))+
  theme_gray() + 
  xlab("")+
  # facet_grid(.~tp)+
  scale_x_discrete(labels= my.labels)+
  scale_fill_manual(values=group.colors, labels = c("\nNormal pancreas\nn = 15\n", "\nTumor tissue\nn = 25\n"))+
  ggtitle("FFPE human tumor tissue vs. normal pancreas - Inverse Simpson index") +
  theme(axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.text = element_text(size = 12),
        legend.title = element_blank(),
        # legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"),
        axis.text.x = element_text(size=12, hjust = .5, angle=0, vjust = .5),
        plot.title = element_text(size = 14))+
   stat_pvalue_manual(iv, label = "p.adj.signif", inherit.aes = FALSE, tip.length = 0.01)

```


## Beta diversity

```{r}
panc_count <- as_tibble(sample_data(physeq.rar)) %>% filter(sample_side == "normal") %>% count(sample_side) %>% pull(n)
tum_count <- as_tibble(sample_data(physeq.rar)) %>% filter(sample_side == "tumor") %>% count(sample_side) %>% pull(n)

#unweighted UF
unifrac_dist = phyloseq::distance(physeq.rar, method="unifrac", weighted=F)
uf_ordination = ordinate(physeq.rar, method="PCoA", distance=unifrac_dist)

pcoa1 <- paste("PCoA 1", " [", round(uf_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(uf_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")

set.seed(40)
p.adonis <- adonis2(unifrac_dist ~ sample_data(physeq.rar)$sample_side)
p <- case_when(
  p.adonis$`Pr(>F)`[1] > 0.05 ~ "n.s.",
  p.adonis$`Pr(>F)`[1] < 0.05 &  p.adonis$`Pr(>F)`[1] > 0.01 ~ paste(p.adonis$`Pr(>F)`[1], "*", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.01 & p.adonis$`Pr(>F)`[1] > 0.001  ~ paste(p.adonis$`Pr(>F)`[1], "**", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.001 ~ paste(p.adonis$`Pr(>F)`[1], "***", sep = " "),
)
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2) ,
        vjustvar = c(1.5))

plot_ordination(physeq.rar, uf_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  scale_color_manual(values=group.colors, labels=c(paste("Normal pancreas\nn =", panc_count, sep = " "), paste("Tumor tissue\nn =", tum_count, sep = " ")))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("FFPE human tumor tissue vs. normal pancreas - unweighted UniFrac") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12),
        legend.title = element_blank())+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)


#weighted UF
w_unifrac_dist = phyloseq::distance(physeq.rar, method="unifrac", weighted=T)
wuf_ordination = ordinate(physeq.rar, method="PCoA", distance=w_unifrac_dist)

pcoa1 <- paste("PCoA 1", " [", round(wuf_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(wuf_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")


set.seed(40)
p.adonis <- adonis2(w_unifrac_dist ~ sample_data(physeq.rar)$sample_side)

p <- case_when(
  p.adonis$`Pr(>F)`[1] > 0.05 ~ "n.s.",
  p.adonis$`Pr(>F)`[1] < 0.05 &  p.adonis$`Pr(>F)`[1] > 0.01 ~ paste(p.adonis$`Pr(>F)`[1], "*", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.01 & p.adonis$`Pr(>F)`[1] > 0.001  ~ paste(p.adonis$`Pr(>F)`[1], "**", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.001 ~ paste(p.adonis$`Pr(>F)`[1], "***", sep = " "),
)
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2) ,
        vjustvar = c(1.5))

plot_ordination(physeq.rar, wuf_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  scale_color_manual(values=group.colors, labels=c(paste("Normal pancreas\nn =", panc_count, sep = " "), paste("Tumor tissue\nn =", tum_count, sep = " ")))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("FFPE human tumor tissue vs. normal pancreas - weigthed UniFrac") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12),
        legend.title = element_blank())+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)


#Bray curtis
bray_dist = phyloseq::distance(physeq.rar, method="bray")
bc_ordination = ordinate(physeq.rar, method="PCoA", distance=bray_dist)

pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")

set.seed(40)
p.adonis <- adonis2(bray_dist ~ sample_data(physeq.rar)$sample_side)

p <- case_when(
  p.adonis$`Pr(>F)`[1] > 0.05 ~ "n.s.",
  p.adonis$`Pr(>F)`[1] < 0.05 &  p.adonis$`Pr(>F)`[1] > 0.01 ~ paste(p.adonis$`Pr(>F)`[1], "*", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.01 & p.adonis$`Pr(>F)`[1] > 0.001  ~ paste(p.adonis$`Pr(>F)`[1], "**", sep = " "),
  p.adonis$`Pr(>F)`[1] <= 0.001 ~ paste(p.adonis$`Pr(>F)`[1], "***", sep = " "),
)
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2) ,
        vjustvar = c(1.5))

plot_ordination(physeq.rar, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_color_manual(values=group.colors, labels=c(paste("Normal pancreas\nn =", panc_count, sep = " "), paste("Tumor tissue\nn =", tum_count, sep = " ")))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("FFPE human tumor tissue vs. normal pancreas - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12),
        legend.title = element_blank())+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)
```


## Differential abundance

Once we have observed significant differences regarding all beta diversity measurements,  we like to define which taxa are responsible for these differences. This methodical step is called differential abundance. There are several approaches to calculate the DA. This recent [paper](https://www.nature.com/articles/s41467-022-28034-z) nicely reviews the different approaches with and without prevalence filtering. So far I just established the most common way: Huttenhowers LEFSE, which is broadly applied in several microbiome studies. This tool applies Liniar discriminant analysis (LDA).

We have to different ways of generate the important figures.  

1. approach: all is done in R  
2. approach: a table is created in R which can be uploaded at huttenhower galaxy  

For both approaches we need yingtools2.

For using the yingtool2 it is not allowed to have columns  in the sample_data containing the word "sample". Therefore, we have to adjust the sample_data a little bit.

```{r}
# library(yingtools2)
# df <- as_tibble(as.data.frame(sample_data(physeq.rar)))
# df <- df %>% mutate(site = sample_side,
#                     nr = sample_nr) %>% 
#   select(-sample_side, -sample_nr, -sample) %>% as.data.frame()
# rownames(df) <- paste(df$uniqueID, df$site, sep = ".")
# df <- sample_data(df)
# phy <- merge_phyloseq(otu_table(physeq.rar), tax_table(physeq.rar), df)
# 
# lda <- lda.effect(phy, class="site")
# 
# lda.plot <- lda %>% filter(pass==TRUE & lda > 3.5) %>%
#   mutate(lda=if_else(as.numeric(factor(direction))==2,-lda,lda)) %>%
#   arrange(lda) %>%
#   mutate(taxon=fct_inorder(taxon))
# 
# ggplot(lda.plot,aes(x=taxon,y=lda,fill=direction)) + 
#   geom_col() + 
#   coord_flip()
# 
# lvls <- rank_names(phy)
# for (i in 1:length(lvls)) {
#   lda[[lvls[i]]] <- str_split(lda$taxonomy,"\\|") %>% map_chr(~str_c(.[1:i],collapse="|"))
# }
# 
# library(ggtree)
# lefse.phy <- as.phylo.formula2(as.formula(paste("~",paste(lvls,collapse="/"))),data=lda)
# gt <- ggtree(lefse.phy,layout="circular")
# 
# get.children.yrange <- function(node,gd) {
#   hits <- gd$node[gd$parent==node]
#   if (length(hits)==0 | node %in% hits) {
#     return(gd$y[gd$node==node])
#   } else {
#     return(unlist(lapply(hits,get.children.yrange,gd)))
#   }
# }
# gd <- gt$data %>% left_join(lda,by=c("label"="taxonomy")) %>%
#   mutate(y.range=lapply(node,get.children.yrange,cur_data()),
#   ymin=map_dbl(y.range,min)-0.5,
#   ymax=map_dbl(y.range,max)+0.5,
#   xmin=x,
#   xmax=1+2*length(lvls)-x,
#   ymid=(ymin+ymax)/2,
#   xtext=xmax-0.5,
#   angle.label=scales::rescale(ymid,from=range(y),to=c(0,360)),
#   angle.label=if_else(is.between(angle.label,0,180),angle.label-90,angle.label+90),
#   short.label=map_chr(str_split(label,"\\|"),last))
# 
# gt + geom_point(data=gd,aes(size=log.max),color="dark gray",fill="gray",shape=21) +
#   geom_rect(data=filter(gd,pass),aes(xmin=xmin,xmax=xmax,ymin=ymin,ymax=ymax,fill=direction),color="dark gray",alpha=0.2) +
#   geom_point(data=filter(gd,pass),aes(fill=direction,size=log.max),shape=21) +
#   geom_text(data=filter(gd,pass),aes(x=xtext,y=ymid,label=short.label,angle=angle.label)) + 
#   theme(legend.position="right")
# ```
# 
# Huttenhowers Lefse uses python. Happily some guys have written a function which converts a phyloseq object to a text file which can be uploaded [here](https://huttenhower.sph.harvard.edu/galaxy/).
# 
# The output is stored in the current working directory and can be uploaded [here](https://huttenhower.sph.harvard.edu/galaxy/). Get data: upload file --> Lefse
# 
# ```{r}
# library(reshape2)
# lefse2 <- function (phy, class, subclass = NA, subject = NA, anova.alpha = 0.05,
#                     wilcoxon.alpha = 0.05, lda.cutoff = 2, wilcoxon.within.subclass = FALSE,
#                     one.against.one = FALSE, mult.test.correction = 0, make.lefse.plots = FALSE,
#                     by_otus = FALSE, levels = rank_names(phy)) {
#   keepvars <- c(class, subclass, subject, "sample")
#   keepvars <- unique(keepvars[!is.na(keepvars)])
#   samp <- get.samp(phy)[, keepvars]
#   if (by_otus) {
#     otu <- get.otu.melt(phy, sample_data = FALSE)
#     otu.levels <- otu %>%
#       mutate(taxon = otu) %>%
#       group_by(sample,taxon) %>%
#       dplyr::summarize(pctseqs = sum(pctseqs)) %>%
#       mutate(taxon = gsub(" ", "_", taxon))
#   }else{
#     otu <- get.otu.melt(phy, sample_data = FALSE)
#     otu.list <- lapply(1:length(levels), function(i) {
#       lvls <- levels[1:i]
#       lvl <- levels[i]
#       otu.level <- otu
#       otu.level$taxon <- do.call(paste, c(lapply(lvls, function(l) otu[[l]]), sep = "|"))
#       otu.level$rank <- lvl
#       otu.level2 <- otu.level %>%
#         group_by(sample, taxon,  rank) %>%
#         dplyr::summarize(pctseqs = sum(pctseqs)) %>%
#         ungroup()
#       return(otu.level2)
#     })
#     otu.levels <- bind_rows(otu.list) %>%
#       mutate(taxon = gsub(" ","_", taxon))
#   }
#   otu.tbl <- otu.levels %>%
#     dcast(sample ~ taxon, value.var = "pctseqs",fill = 0) %>%
#     left_join(samp, by = "sample") %>%
#     select_(.dots = c(keepvars,lazyeval::interp(~everything())))
#   if (is.na(subject) | subject != "sample") {
#     otu.tbl <- otu.tbl %>% select(-sample)
#   }
#   tbl <- otu.tbl %>% t()
#   write.table(tbl, "lefse.Cov50.ori.txt", quote = FALSE, sep = "\t",
#               col.names = FALSE)
# }
# lefse2(phy,class="site")
```


## Microbial composition at order level

```{r, fig.width=10, fig.height=4}
physeq.fam <- transform_sample_counts(physeq.rar, function(x){x/sum(x)})
df.fam.melt <- psmelt(physeq.fam)

df.fam.melt$order.v2 <- ifelse(df.fam.melt$Abundance>.03,as.character(df.fam.melt$order), "others")
df.fam.melt$order.v2 <- factor(df.fam.melt$order.v2, levels=rev(unique(df.fam.melt$order.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33")


ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=order.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at order level")+
  theme( axis.text.x = element_text(angle=45,  vjust = .5),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

```

## Microbial composition at family level

```{r, fig.width=10, fig.height=4}
physeq.fam <- transform_sample_counts(physeq.rar, function(x){x/sum(x)})
df.fam.melt <- psmelt(physeq.fam)

df.fam.melt$family.v2 <- ifelse(df.fam.melt$Abundance>.03,as.character(df.fam.melt$family), "others")
df.fam.melt$family.v2 <- factor(df.fam.melt$family.v2, levels=rev(unique(df.fam.melt$family.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33")


ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=family.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at family level")+
  theme( axis.text.x = element_text(angle=45,  vjust = .5),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

```


## Microbial composition at genus level

```{r, fig.width=10, fig.height=4}
physeq.fam <- transform_sample_counts(physeq.rar, function(x){x/sum(x)})
df.fam.melt <- psmelt(physeq.fam)

df.fam.melt$genus.v2 <- ifelse(df.fam.melt$Abundance>.04,as.character(df.fam.melt$genus), "others")
df.fam.melt$genus.v2 <- factor(df.fam.melt$genus.v2, levels=rev(unique(df.fam.melt$genus.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33")


ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at genus level")+
  theme( axis.text.x = element_text(angle=45,  vjust = .5),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

```

## Microbial composition at species level

```{r, fig.width=10, fig.height=4}
physeq.fam <- transform_sample_counts(physeq.rar, function(x){x/sum(x)})
df.fam.melt <- psmelt(physeq.fam)

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>.04,as.character(df.fam.melt$species), "others")
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "magenta2", "palegreen", "salmon", "maroon", "cyan2")


ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~sample_side,  scales = "free_x")+
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_text(angle=45,  vjust = .5),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

```


```{r}
# Exiguobacterium sp. AT1b
contamdf.prev %>% filter(row.names(contamdf.prev) %in% 360911)
# Methylobacterium sp.
contamdf.prev %>% filter(row.names(contamdf.prev) %in% 2696060)
```

# Decontamination approach ala Nejman et al.

Details are found under following [link](https://www.science.org/action/downloadSupplement?doi=10.1126%2Fscience.aay9189&file=aay9189_nejman_sm.pdf). A step by step abstract is provided for the decontamination protocol (hit calling pipeline" by Nejman et al. and information about adjustments to our special type of data set are given.

The whole set of Nejman et al. included 1526 samples, whereas 811 were controls (53%). In comparison our set of data has 129 samples with 75 controls (58%)

## Exclude low read count NTCs

1. Nejman: All NTC with read counts < 1000 are excluded (in total 10 samples from 811 samples).  

Here I read in again all NTCs. In comparison to the above deconatam approach I at first will not apply any Perfect filtering (Prevalence filtering) to keep as much reads as possible. Only the filtering for bacterial reads and the tax_glom on species level will be conducted.

```{r}
physeq.nct1 <- readRDS("2022_06_02_patho_FFPE_tum_panc_ntc_1/physeq_original_50.rds")
physeq.nct2 <- readRDS("2022_06_02_patho_FFPE_tum_panc_ntc_2/physeq_original_50.rds")
physeq.nct3 <- readRDS("2022_06_03_patho_FFPE_tum_panc_ntc_3/physeq_original_50.rds")
physeq.nct4 <- readRDS("2022_06_03_patho_FFPE_tum_panc_ntc_4/physeq_original_50.rds")
physeq.nct5 <- readRDS("2022_06_04_patho_FFPE_tum_panc_ntc_5/physeq_original_50.rds")
physeq.nct6 <- readRDS("2022_06_04_patho_FFPE_tum_panc_ntc_6/physeq_original_50.rds")
physeq.nct7 <- readRDS("2022_06_10_patho_FFPE_tum_panc_ntc_7/physeq_original_50.rds")
physeq.nct8 <- readRDS("2022_07_13_patho_FFPE_tum_panc_ntc_8/physeq_original_50.rds")
physeq.nct9 <- readRDS("2022_07_13_patho_FFPE_tum_panc_ntc_9/physeq_original_50.rds")
physeq.nct10 <- readRDS("2021_12_02_patho_FFPE_self_ntc/physeq_original.rds")
physeq.nct11 <- readRDS("2022_01_24_patho_FFPE_self_ntc_2/physeq_original.rds")


physeq.nct <- merge_phyloseq(physeq.nct1, physeq.nct2, physeq.nct3, physeq.nct4, physeq.nct5, physeq.nct6, physeq.nct7, physeq.nct8, physeq.nct9, physeq.nct10, physeq.nct11)

sample_data(physeq.nct)$sample_side <- ifelse(sample_data(physeq.nct)$sample_side == "paraffin", "paraffin_tumor", sample_data(physeq.nct)$sample_side)

physeq.filt.nct <- subset_taxa(physeq.nct, superkingdom=="Bacteria")
physeq.filt.nct.s <- tax_glom(physeq.filt.nct, taxrank = "species", NArm = TRUE)

bac.count.ntc <- tibble(bac.count = sample_sums(physeq.filt.nct.s), 
                        sample = sample_data(physeq.filt.nct.s)$id) %>% 
  arrange (bac.count)

tibble(
  ">250 reads" = length(which(bac.count.ntc$bac.count > 250)),
  ">500 reads" = length(which(bac.count.ntc$bac.count > 500)),
  ">750 reads" = length(which(bac.count.ntc$bac.count > 750)),
  ">1000 reads" = length(which(bac.count.ntc$bac.count > 1000)),
)  %>% 
  kable (caption="NTC read counts", booktabs =TRUE) %>%
  kable_classic(full_width = F, html_font = "Cambria") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) 

id.ex <- bac.count.ntc %>% filter(bac.count < 750) %>% pull (sample)
```


According to the table above we choose a cut-off of 750 reads. So we have nearly the same read count as Nejman et al and we still have almost 50% NTC vs. true samples. 

```{r}
physeq.filt.nct.s <- subset_samples(physeq.filt.nct.s, !(id %in% id.ex))
```

## Prevalence filtering

2. Nejman et al. applied a prevalence filtering according the rule of thumb (excluding all bacteria with a relative abundance < 10⁻⁴ %)  

```{r}
p.comb.filt.s <- merge_phyloseq(physeq.filt.nct.s, p.true.filt.s)
# Nejman prevalence filtering for controls

physeq.filt.nct.s.prev.filt <- transform_sample_counts(physeq.filt.nct.s, function(x){x/sum(x)})
physeq.filt.nct.s.prev.filt <- filter_taxa(physeq.filt.nct.s.prev.filt, function(x) mean(x) > 1e-4, TRUE)

# Nejman prevalence filtering for true samples

p.true.filt.s.rel <- transform_sample_counts(p.true.filt.s, function(x){x/sum(x)})
physeq.true.filt.s.prev.filt <- filter_taxa(p.true.filt.s.rel, function(x) mean(x) > 1e-4, TRUE)

# combining ntc with true samples

p.comb.filt.s.prev <- merge_phyloseq(physeq.true.filt.s.prev.filt, physeq.filt.nct.s.prev.filt)

# Perfect prevalence filtering

otu = as(otu_table(physeq.filt.nct.s), "matrix")
if(taxa_are_rows(physeq.filt.nct.s)){otu <- t(otu)}
otu = as_tibble(otu)
res_sim <- PERFect_sim(X = otu)
dim(res_sim$filtX)
ids.sim<- colnames(res_sim$filtX)
physeq.filt.nct.s.perfect <-  prune_taxa(ids.sim, physeq.filt.nct.s)
```

Hereby there were only `r nrow(otu_table(physeq.filt.nct.s.prev.filt))` taxa from initially `r nrow(otu_table(physeq.filt.nct.s))` kept. With our perfect prevalence method we kept `r nrow(otu_table(physeq.filt.nct.s.perfect))`. So our approach does not filter as much as the rule of thumb applied by Nejman et al..

## Filter 1 - Removing global environmental and paraffin contaminants

3. Here Nejman et al. removed all species that were prevalent across more than 7.5% of negative DNA/PCR/Sequencing controls or more than 7.5% of empty paraffin controls. The threshold of 7.5% was determined on the inflection point in the species prevalence abundance distribution (figure S9b in Nejman et al.). Here the x-axis represents the relative prevalence of all species oin control samples plotted against frequencies of species (whole data set (true samples and controls combined)).

### Prevalence - Abundance distribution

![Nejman et al. Prevalence - abundance distribution](add_files/abundance_prev_dist_controls_nejaman.png)

In the following chunk I create such prevalence-abundance distribution for our setting.

```{r}
prev = tibble(
  prev = prevalence(physeq.filt.nct.s.prev.filt, detection  = 0, sort = TRUE, count = FALSE),
  OTU =as.factor(names(prevalence(physeq.filt.nct.s.prev.filt , detection = 0, sort = TRUE, count = FALSE)))
)

df <- psmelt(p.comb.filt.s.prev)
abu <- df %>% group_by(OTU) %>% 
  summarise(across(Abundance, mean)) 

prev_abu <- left_join(abu, prev) %>% 
  mutate(prev = ifelse(is.na(prev), 0, prev)) %>% 
  arrange (-Abundance) %>% as_tibble()


ggplot(prev_abu, aes(x= prev, y = Abundance))+
  geom_point()+
  xlab ("Relative prevalence of species in controls") +
  ylab("Relative abundance of species in controls and samples") +
  theme_bw() +
  ggtitle("Pervalence - Abundance Distribution")+
  theme( axis.text.x = element_text(size=12),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

```

### Microbial composition true samples vs. controls

In Comparison, we have another distribution than Nejman et al.. Highly abundant species are found in the most control samples. We further want to make a glimpse in the microbial composition of pooled control samples and true samples to further visualize this finding.

```{r, fig.height=6, fig.width=10}
physeq.fam <- transform_sample_counts(p.comb.filt.s.prev, function(x){x/sum(x)})
df.fam.melt <- psmelt(p.comb.filt.s.prev) 

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>.1,as.character(df.fam.melt$species), "others")
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "magenta2", "palegreen", "salmon", "maroon", "cyan2")


ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~TRUE_control,  scales = "free_x")+
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

```

This barplot verify the findings of the prevalence - abundance distribution. The most abundant taxa are found in controls and true samples. Most of them belong to the genus Sphingomonas a known bacterium from the environment.

### Applying filter 1

Following we will apply the filter 1. Here we do not separate the control physeq object in paraffin and the other controls. We have too less buffer/ PCR H2O controls (> 7.5% would mean a whole sample, meaning all taxa from these controls will be removed). Thus we define all species as contaminant which occur in more than 7.5% of ALL controls. That means a species have to occur in `r round(nrow(sample_data(physeq.filt.nct.s.prev.filt))*0.075, digits = 0)` controls to be defined as contaminant.  

#### Nejman threshold 

```{r}
filter <- 0.075

filtered_taxa1 <- tibble(
  Cont = genefilter_sample(physeq.filt.nct.s.prev.filt, filterfun_sample(function(x) x > 0), A=filter*nsamples(physeq.filt.nct.s.prev.filt)),
  OTU = names(genefilter_sample(physeq.filt.nct.s.prev.filt, filterfun_sample(function(x) x > 0), A=filter*nsamples(physeq.filt.nct.s.prev.filt)))
) %>% 
  arrange (-Cont)

cont.count <- filtered_taxa1 %>% count(Cont) 
cont.count %>% 
  kable (caption=paste("Count of contaminants (filter", filter*100, " %)"), booktabs =TRUE) %>%
  kable_classic(full_width = F, html_font = "Cambria") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) 
```

##### List of Contaminants

```{r}
df.fam.melt %>% select(OTU, species) %>% 
  unique %>% 
  right_join(filtered_taxa1) %>% 
  filter(Cont == TRUE) %>% 
  select(-Cont) %>% 
  kable (caption=paste("Contaminants (filter", filter*100, " %)"), booktabs =TRUE) %>%
  kable_classic(full_width = F, html_font = "Cambria") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>% 
  scroll_box(width = "70%", height = "600px")
```


#### Alternative threshold 

Applying the Nejman et al. threshold of 7.5 % will define most taxa as Contaminants (n = `r cont.count[2,2]`) which occur in the controls. However, our prevalence - abundance curve shows a different distribution. Therefore, another cut-off according to this distribution seems reasonable.

```{r}
filter <- 0.7

filtered_taxa1 <- tibble(
  Cont = genefilter_sample(physeq.filt.nct.s.prev.filt, filterfun_sample(function(x) x > 0), A=filter*nsamples(physeq.filt.nct.s.prev.filt)),
  OTU = names(genefilter_sample(physeq.filt.nct.s.prev.filt, filterfun_sample(function(x) x > 0), A=filter*nsamples(physeq.filt.nct.s.prev.filt)))
) %>% 
  arrange (-Cont)

cont.count <- filtered_taxa1 %>% count(Cont) 
cont.count %>% 
  kable (caption=paste("Count of contaminants (filter", filter*100, " %)"), booktabs =TRUE) %>%
  kable_classic(full_width = F, html_font = "Cambria") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) 
```

##### List of Contaminants

```{r}
df.fam.melt %>% select(OTU, species) %>% 
  unique %>% 
  right_join(filtered_taxa1) %>% 
  filter(Cont == TRUE) %>% 
  select(-Cont) %>% 
  kable (caption=paste("Contaminants (filter", filter*100, " %)"), booktabs =TRUE) %>%
  kable_classic(full_width = F, html_font = "Cambria") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive")) %>% 
  scroll_box(width = "70%", height = "600px")
```

Now, only `r cont.count[2,2]` were defined as Contaminants. That means `r round(cont.count[2,2]/sum(cont.count$n)*100, digits = 1)` % of the taxa. In comparison Nejman et al. removed `r round(100-9023/9190*100, digits = 1)` % of the taxa with the first general filter.

### Filter 2 - 3

In order to remove the less prevalent contaminations, Nejman et al. compared taxa prevalences in samples to their prevalences in controls based on the assumption that "even in a low-biomass regime, the non-contaminants will appear in a larger fraction of true samples than in negative controls". This approach veers away from skewed relative abundance data which occur when contaminations are high and relies more heavily on absent/present calls, which along with flooring by relative abundance (see above), reduces the noise variation of presence/absence calling at the very low abundance levels.
As the RIDE guidelines and additional research indicate, different contamination profiles can vary across processing batches and time.
Nejman et al. thus shifted to a 'percondition' (a combined tissue+tumor/NAT/normal status) contamination analysis. Therefore, our comparisons between taxon prevalence in samples and controls were done on a per-condition, per-batch basis, comparing samples from the same condition to the controls that were processed alongside them. This was done for the three different processing batch types; DNA extraction batch (filter 2), PCR amplification batch (filter 3), and sequencing library batch (filter 4). **They employed the non-parametric, exact, binomial test on each taxa's prevalence, x=number of successes in the sample & n=sample size to the background p= taxa's prevalence in the relevant batch controls**. Only taxa that passed a p-value cutoff of 0.05 in all batch comparisons (filters 2-4) were considered in the next filtering stage.

We only have controls for filter 2 (DNA extraction) and filter 3 (PCR). First of all we separate the phyloseq objects accordingly (samples + certain control batch (PCR or buffer).
Here we take the not prevalence filtered physeq object, because we need absolute abundances. Due to our prevalence filtering we irreversibly transformed the abundances to relative values. 
```{r}
p.comb.pcr <- subset_samples(p.comb.filt.s, sample_side == "PCR_ctrl_H2O" | TRUE_control == "TRUE")
p.comb.buffer <- subset_samples(p.comb.filt.s, sample_side == "buffer" | TRUE_control == "TRUE")
```

We then removed all species identified in filter 1 (defined in ont.filt1) from both phyloseq objects. 

```{r}
cont.filt1 <- df.fam.melt %>% select(OTU, species) %>% 
  unique %>% 
  right_join(filtered_taxa1) %>% 
  filter(Cont == TRUE) %>% 
  pull(species)

p.comb.pcr <- subset_taxa(p.comb.pcr,!species %in% cont.filt1)
p.comb.buffer <- subset_taxa(p.comb.buffer,!species %in% cont.filt1)
```

#### DNA extraction filter

```{r}
#test

df <- psmelt(p.comb.buffer) 

# binomal test for Acidovorax sp.

rp <- subset_taxa(p.comb.buffer, species == "Acidovorax sp." | species == "Ramlibacter tataouinensis")
otu_table(rp)
rp.ntc <- subset_samples(rp, sample_side == "buffer")
rp.true <- subset_samples(rp, TRUE_control == "TRUE")

p <- sum(prevalence(rp.ntc, detection = 0, count = TRUE))/length(prevalence(rp.ntc, detection = 0, count = TRUE))


x <- otu_tibble(rp.true) %>% rowwise(FeatureID) %>% summarise(sum = sum(c_across(contains("tumor") | contains("normal"))))
#
```

https://www.statology.org/binomial-test-r/