---
title: "Comprehensive Workflow"
author: "Christoph Petrynowski"
date: "2022-12-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

In this experiment, human tumor vs. normal pancreas FFPE tissue samples from the Pathology department are examined. The DNA had been isolated with the Invitrogen FFPE kit (by me). The sample size is n=24 for both tumor and normal pancreas (matched patients). Additionally, n=6 FFPE tumor tissue samples from the establishment experiment (patho_ffpe_self) without matched normal pancreas tissue are implemented (so we have n=30 tumor, n=24 normal pancreas). n=75 NTCs (n=23 paraffin tumor, n=24 paraffin normal, n=10 buffer-only, n=7 PCR CTRLs) plus n=11 NTCs from the establishment experiment (patho_ffpe_self) are implemented. Further, several samples <1000 reads were repeated and implemented as well (merge with their first sequencing).

{22.12.2022 CP} Who actually wrote that? Physeq1-5 alrady contain all 58 true samples - respective of removing 4 samples from phyloseq5

# Software-Library

```{r library, echo=FALSE}
library(PERFect)
library(phyloseq)
library(decontam)
library(tidyverse)
library(pairwiseAdonis)
library(kableExtra)
library(vegan)
library(tidytree)
library(ape)
library(ggrepel)
library(microbiome)
library(ggpubr) # for manual p-value assigning
library(rstatix) # for wilcox test
library(yingtools2)
#if( !require(ggtree) ){
#  install.packages(ggtree)
#}
library(readxl) # load excel data
if( !require(rlist) ){
  install.packages(rlist)
}
library(rlist) # work with lists
if( !require(ggsignif) ){
  install.packages(ggsignif)
}
library(ggsignif) # plot p.values on a ggplot
set.seed(3)
library(plyr) #for ldply - performs function on every element in list & combines the output into a dataframe
library(reshape2) # for melting data
if( !require(doParallel) ){
  install.packages("doParallel")
}
library(doParallel) # ConQuR´s GitHub recomends this step
if( !require(ConQuR) ){
  devtools::install_github("wdl2459/ConQuR")
}
library(ConQuR)
library("VennDiagram")
if( !require("ggvenn") ){
  install.packages("ggvenn")
}
library("ggvenn")
```

# Raw Phyloseq 1500.80s
## Loading
### Loading phyloseq "1500.80s true samples" objects
```{r loading phyloseq "1500.80s true samples" objects}
physeq1.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_05_27_patho_FFPE_tum_panc_1/physeq_original_1500.80.rds")
physeq2.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_05_27_patho_FFPE_tum_panc_2/physeq_original_1500.80.rds")
physeq3.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_05_28_patho_FFPE_tum_panc_3/physeq_original_1500.80.rds")
physeq4.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_05_28_patho_FFPE_tum_panc_4/physeq_original_1500.80.rds")
physeq5.1500.80 <- readRDS("raw_data/PhyloSeqObj/2021_12_01_patho_FFPE_self_tumor/physeq_original_1500.80.rds")

physeq6.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_08_09_patho_FFPE_tum_panc_rep1/physeq_original_1500.80.rds") #09.08.22: rep1: 10 samples (tumor+normal panc mixed)
physeq7.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_08_09_patho_FFPE_tum_panc_rep2/physeq_original_1500.80.rds") #09.08.22: rep2: 10 samples (tumor+normal panc mixed)
p.true.original.1500.80 <- merge_phyloseq(physeq1.1500.80,physeq2.1500.80,physeq3.1500.80,physeq4.1500.80,physeq5.1500.80)#merge the different phyloseq objects into 1 phyloseq object
# sample_data of new phyloseq-object

p.true.filt.1500.80 <- subset_taxa(p.true.original.1500.80, superkingdom=="Bacteria")
p.true.filt.s.1500.80 <- tax_glom(p.true.filt.1500.80, taxrank = "species", NArm = TRUE)
```
##LibrarySize
### LibrarySize tissue samples
```{r tissue samples´s librarysize of the 1500.80s}
normal_count <- as_tibble(sample_data(p.true.filt.s.1500.80)) %>% filter(sample_side == "normal") %>% count("sample_side") %>% pull(freq)
tumor_count <- as_tibble(sample_data(p.true.filt.s.1500.80)) %>% filter(sample_side == "tumor") %>% count("sample_side") %>% pull(freq)
df <- meta(p.true.filt.s.1500.80) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.true.filt.s.1500.80) # add LibrarySize for each sample
df <- df[order(df$LibrarySize),] # order samples (dataframe) based on LibrarySize in ascending order (this will be our y_value for ggplot)
df$Index <- seq(nrow(df)) # give each row an index (this will be our x_value for ggplot)
ggplot(data=df, aes(x=Index, y=LibrarySize, color=sample_side)) + geom_point()  + theme_bw() + ggtitle("Library size true samples 1500.80")+
  scale_color_brewer(palette = "Set1", labels=c(paste("Normal samples\nn =", normal_count, sep = " "), paste("Tumor samples\nn =", tumor_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12),
        legend.title = element_blank())
```
## Loading
### Loading phyloseq "1500.80s ntc samples" objects
```{r loading phyloseq "1500.80s ntc samples" objects}
#loading 1500.80s ntc samples
physeq.nct1.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_06_02_patho_FFPE_tum_panc_ntc_1/physeq_original_1500.80.rds")
physeq.nct2.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_06_02_patho_FFPE_tum_panc_ntc_2/physeq_original_1500.80.rds")
physeq.nct3.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_06_03_patho_FFPE_tum_panc_ntc_3/physeq_original_1500.80.rds")
physeq.nct4.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_06_03_patho_FFPE_tum_panc_ntc_4/physeq_original_1500.80.rds")
physeq.nct5.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_06_04_patho_FFPE_tum_panc_ntc_5/physeq_original_1500.80.rds")
physeq.nct6.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_06_04_patho_FFPE_tum_panc_ntc_6/physeq_original_1500.80.rds")
physeq.nct7.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_06_10_patho_FFPE_tum_panc_ntc_7/physeq_original_1500.80.rds")
physeq.nct8.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_07_13_patho_FFPE_tum_panc_ntc_8/physeq_original_1500.80.rds")
physeq.nct9.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_07_13_patho_FFPE_tum_panc_ntc_9/physeq_original_1500.80.rds")
physeq.nct10.1500.80 <- readRDS("raw_data/PhyloSeqObj/2021_12_02_patho_FFPE_self_ntc/physeq_original_1500.80.rds")
physeq.nct11.1500.80 <- readRDS("raw_data/PhyloSeqObj/2022_01_24_patho_FFPE_self_ntc_2/physeq_original_1500.80.rds")

physeq.nct.original.1500.80 <- merge_phyloseq(physeq.nct1.1500.80, physeq.nct2.1500.80, physeq.nct3.1500.80, physeq.nct4.1500.80, physeq.nct5.1500.80, physeq.nct6.1500.80, physeq.nct7.1500.80, physeq.nct8.1500.80, physeq.nct9.1500.80,physeq.nct10.1500.80, physeq.nct11.1500.80)

physeq.filt.nct.1500.80 <- subset_taxa(physeq.nct.original.1500.80, superkingdom=="Bacteria")
physeq.filt.nct.s.1500.80 <- tax_glom(physeq.filt.nct.1500.80, taxrank = "species", NArm = TRUE)

#replace sample_side "paraffin" with "paraffin_tumor"
sample_data(physeq.filt.nct.s.1500.80)$sample_side <- ifelse(sample_data(physeq.filt.nct.s.1500.80)$sample_side == "paraffin", "paraffin_tumor", sample_data(physeq.filt.nct.s.1500.80)$sample_side)
```
## LibrarySize
### LibrarySize NCT samples
```{r NCT samples´s librarysize of the 1500.80s}
df <- meta(physeq.filt.nct.s.1500.80) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(physeq.filt.nct.s.1500.80) # add LibrarySize for each sample
df <- df[order(df$LibrarySize),] # order samples (dataframe) based on LibrarySize in ascending order (this will be our y_value for ggplot)
df$Index <- seq(nrow(df)) # give each row an index (this will be our x_value for ggplot)
ggplot(data=df, aes(x=Index, y=LibrarySize, color=sample_side)) + geom_point()  + theme_bw() + ggtitle("Library size negative controls 1500.80")+ scale_fill_discrete(name="Sample\nSide")+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12),
        legend.title = element_blank())
```
## Merge TRUE/NCT samples
```{r merge true/nct samples}
p.comb.filt.s.1500.80 <- merge_phyloseq(physeq.filt.nct.s.1500.80,p.true.filt.s.1500.80) # merge the phyloseq objects containing true samples (p.true.filt.s.1500.80) and nct samples (physeq.filt.nct.s.1500.80)
p.comb.filt.s.1500.80 <- tax_glom(p.comb.filt.s.1500.80,taxrank = "species", NArm = TRUE) # agglomerate taxa to species level. This combines different OTUs {hypothecially: 450 / 453} assigned to the same species into just 1 OTU {this progess accounts for } 
```
## Loading
### Loading Batch metadata
```{r introduce Batch metadata}
df <- read_excel("raw_data/Metadata.xlsx", 
    col_types = c("numeric", "text", "text", 
        "text", "text", "text", "text", "text", 
        "text", "numeric", "text", "numeric", 
        "numeric", "text", "text")) #load metadata excel-sheet containing DNA extraction batch information & PCR batch infromation
df <- as.data.frame(df)

sd.new <- meta(p.comb.filt.s.1500.80) %>% arrange(uniqueID)
df <- df %>% arrange(uniqueID)
sd.new$DNAex_round <- df$DNAex_round
sd.new$PCR_round <- df$PCR_round
rownames(sd.new) <- sd.new$id

# the above code arranges sampledata of merged phyloseq-object by uniqueID & the metadata excel-sheet by uniqueID. This guarantees the ordering of the rows for both objects is the same. It follows that we can now create the columns "DNAex_round" & "PCR_round" in sd.new using our df. As sample_data from phyloseq-objects have rownames equivalent to the names of their samples we define the rownames of sd.new using the $id colums

p.comb.filt.s.1500.80 <- merge_phyloseq(otu_table(p.comb.filt.s.1500.80), tax_table(p.comb.filt.s.1500.80), sample_data(sd.new)) #we merge the otu_table and tax_table of our old phyloseq-object and the newly created sample_data - now containing the DNA extraction batch information and PCR batch information
sample_data(p.comb.filt.s.1500.80)$is.neg <- sample_data(p.comb.filt.s.1500.80)$TRUE_control == "control"#first identify your negative controls 
```
## LibrarySize
### LibrarySize Comb - colored by TRUE/control, DNAex_round, PCR_round
```{r LibrarySize Tissue/NCT - colored by TRUE/control, DNAex_round, PCR_round }
#prepare dataframe for plotting
df <- meta(p.comb.filt.s.1500.80)
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
#prepare TRUE_control counts for plotting
true_count <- meta(p.comb.filt.s.1500.80) %>% filter(TRUE_control == "TRUE") %>% count("TRUE_control") %>% pull()
control_count <- meta(p.comb.filt.s.1500.80) %>% filter(TRUE_control == "control") %>% count("TRUE_control") %>% pull()
#plot librarysize colored by TRUE_control
ggplot(data=df, aes(x=Index, y=LibrarySize, color=TRUE_control)) + geom_point()  + theme_bw() + ggtitle("Library size p.comb.filt.s.1500.80")+
  scale_color_brewer(palette = "Set1", labels=c(paste("Negative controls\nn =", control_count, sep = " "), paste("Tissue samples\nn =", true_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare DNAex_batch counts for plotting
DNAex_batch1_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "1") %>% count("DNAex_round") %>% pull()
DNAex_batch2_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "2") %>% count("DNAex_round") %>% pull()
DNAex_batch3_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "3") %>% count("DNAex_round") %>% pull()
DNAex_batch4_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "4") %>% count("DNAex_round") %>% pull()
DNAex_batch5_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "5") %>% count("DNAex_round") %>% pull()
DNAex_batch6_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "6") %>% count("DNAex_round") %>% pull()
DNAex_batch7_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "7") %>% count("DNAex_round") %>% pull()
DNAex_batch4_5_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "4_5") %>% count("DNAex_round") %>% pull()
DNAex_batchNA_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "NA") %>% count("DNAex_round") %>% pull()
#plot librarysize colored by DNAex_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(DNAex_round))) + geom_point()  + theme_bw() + ggtitle("Library size -  p.comb.filt.s.1500.80")+scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c(paste("Batch 1\nn =", DNAex_batch1_count, sep = " "),paste("Batch 2\nn =", DNAex_batch2_count, sep = " "),paste("Batch 3\nn =", DNAex_batch3_count, sep = " "),paste("Batch 4\nn =", DNAex_batch4_count, sep = " "),paste("Batch 4_5\nn =", DNAex_batch4_5_count, sep = " "),paste("Batch 5\nn =", DNAex_batch5_count, sep = " "),paste("Batch 6\nn =", DNAex_batch6_count, sep = " "),paste("Batch 7\nn =", DNAex_batch7_count, sep = " "),paste("Batch NA\nn =", DNAex_batchNA_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare PCR_batch counts for plotting
PCR_batch1_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "1") %>% count("PCR_round") %>% pull()
PCR_batch2_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "2") %>% count("PCR_round") %>% pull()
PCR_batch3_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "3") %>% count("PCR_round") %>% pull()
PCR_batch4_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "4") %>% count("PCR_round") %>% pull()
PCR_batch5_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "5") %>% count("PCR_round") %>% pull()
PCR_batch6_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "6") %>% count("PCR_round") %>% pull()
PCR_batch7_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "7") %>% count("PCR_round") %>% pull()
PCR_batch8_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "8") %>% count("PCR_round") %>% pull()
PCR_batch9_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "9") %>% count("PCR_round") %>% pull()
#plot librarysize colored by PCR_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(PCR_round))) + geom_point()  + theme_bw() + ggtitle("Library size - p.comb.filt.s.1500.80")+scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c(paste("Batch 1\nn =", PCR_batch1_count, sep = " "),paste("Batch 2\nn =", PCR_batch2_count, sep = " "),paste("Batch 3\nn =", PCR_batch3_count, sep = " "),paste("Batch 4\nn =", PCR_batch4_count, sep = " "),paste("Batch 5\nn =", PCR_batch5_count, sep = " "),paste("Batch 6\nn =", PCR_batch6_count, sep = " "),paste("Batch 7\nn =", PCR_batch7_count, sep = " "),paste("Batch 8\nn =", PCR_batch8_count, sep = " "),paste("Batch 9\nn =", PCR_batch9_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#https://stackoverflow.com/questions/43359050/error-continuous-value-supplied-to-discrete-scale-in-default-data-set-example#43359104

#p.comb.filt.s.1500.80 <- subset_samples(p.comb.filt.s.1500.80, sample_nr!="P19269-20" & sample_nr!="P15970-20" & sample_nr!="P12758-20" & sample_nr!="P3434-20") #we remove 4 samples as these patients do not have buccal/ rectal swaps for later analysis
```
## DEMERGE!
```{r}
p.comb.filt.s.1500.80.true <- subset_samples(p.comb.filt.s.1500.80,TRUE_control=="TRUE") #only true samples
p.comb.filt.s.1500.80.true <- prune_taxa(x = p.comb.filt.s.1500.80.true,taxa = taxa_names(p.comb.filt.s.1500.80.true)[-(which(taxa_sums(p.comb.filt.s.1500.80.true)==0))]) #only taxa found in true samples

p.comb.filt.s.1500.80.nct <- subset_samples(p.comb.filt.s.1500.80,TRUE_control == "control") #only NCT samples
p.comb.filt.s.1500.80.nct <- prune_taxa(x = p.comb.filt.s.1500.80.nct,taxa = taxa_names(p.comb.filt.s.1500.80.nct)[-(which(taxa_sums(p.comb.filt.s.1500.80.nct)==0))]) #only taxa found in NCT samples
```


### LibrarySize - tissue samples - boxplots
```{r visualize LibrarySize of tissue samples as a boxplot(normal vs tumor) - colored by DNAex_round, PCR_round}
#prepare dataframe for plotting
df <- meta(p.comb.filt.s.1500.80.true) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80.true)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))

#set levels
df$sample_side <- factor(df$sample_side)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$sample_side), 2)),seq(nrow(t(combn(levels(df$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df, aes(x = sample_side, y = LibrarySize)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by sample side") +
  theme(text = element_text(size = 10))

#set levels
df$DNAex_round <- factor(df$DNAex_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$DNAex_round), 2)),seq(nrow(t(combn(levels(df$DNAex_round), 2)))))
#boxplot comparing librarysizes of different DNAex batches
ggplot(df, aes(x = DNAex_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "DNA extraction round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.02,comparisons = comb, map_signif_level = FALSE,y_position = seq.int(55000,200000,8000),color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by DNA extraction batch") +
    theme(text = element_text(size = 10))

#set levels
df$PCR_round <- factor(df$PCR_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$PCR_round), 2)),seq(nrow(t(combn(levels(df$PCR_round), 2)))))
#boxplot comparing librarysizes of different PCR batches
ggplot(df, aes(x = PCR_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "PCR round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.1,comparisons = comb, map_signif_level = FALSE,color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by PCR extraction batch") +
    theme(text = element_text(size = 10))
```

We compare librarysizes of true samples via boxplots. 
Plot1: LibSizes of tissue samples grouped by sample side 
Plot2: LibSizes of tissue samples grouped by DNA extraction batch
Plot3: LibSizes of tissue samples grouped by PCR batch
Conclusion: Plot 1 scores a pvalue = 0.066 = 6.6% - i.e. just slightly not significant. In Plot2 and Plot3 we can clearly see the main driver of any differences in librarysizes is DNA batch 6 (Plot 2), PCR batch 8 (Plot 3). Both of these batches represent the same tissue samples (uniqueID: 326 - 335). Combining this information with the microbial composition of these 10 samples yields the conclusion that high amounts of contaminants are present in this batch. 
Visual intra-batch (not DNA extraction batch 6 / PCR batch 8) comparisons indicate no significant differences between librarysizes for normal samples versus tumor samples (i.e. Plot3: take PCR batch 1 and visually compare normal samples versus tumor samples)
Small note: Plot 2 and Plot 3 show that DNA extraction batch 6 / PCR batch 8 are made up of only tumor samples -> bad study design!

Building on our above conclusions: we now go one step further and compare normal samples versus tumor samples for each batch (DNA extraction, PCR) on a statistical basis - if possible (ie PCR batch 8 only contains tumor samples)
### Boxplot of tissue samples´LibrarySize comparing sample_side per DNA extraction batch, PCR batch
```{r Boxplot of tissue samples´LibrarySize comparing sample_side per DNA extraction batch, PCR batch}
#prepare dataframe for plotting
df <- meta(p.comb.filt.s.1500.80.true) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80.true)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
#DNA
#now we need to subset the dataframe for each batch seperatly
#DNA batch 1
df_mod <- df %>% filter(DNAex_round == "1")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = DNAex_round)) +
  ggtitle("LibrarySize of DNA extraction batch 1: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
#DNA batch 2
df_mod <- df %>% filter(DNAex_round == "2")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = DNAex_round)) +
  ggtitle("LibrarySize of DNA extraction batch 2: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
#DNA batch 3
df_mod <- df %>% filter(DNAex_round == "3")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = DNAex_round)) +
  ggtitle("LibrarySize of DNA extraction batch 3: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
#DNA batch 4
df_mod <- df %>% filter(DNAex_round == "4")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = DNAex_round)) +
  ggtitle("LibrarySize of DNA extraction batch 4: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
#DNA batch 5
df_mod <- df %>% filter(DNAex_round == "5")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = DNAex_round)) +
  ggtitle("LibrarySize of DNA extraction batch 5: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
#PCR
#now we need to subset the dataframe for each batch seperatly
df_mod <- df %>% filter(PCR_round == "1")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = PCR_round)) +
  ggtitle("LibrarySize of PCR batch 1: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
  
#now we need to subset the dataframe for each batch seperatly
df_mod <- df %>% filter(PCR_round == "2")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = PCR_round)) +
  ggtitle("LibrarySize of PCR batch 2: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))

#shorter code version using facet_wrap
data <- reshape2::melt(df,measure.vars = c("LibrarySize"),value.name = c("LibrarySize"))
data$sample_side <- factor(data$sample_side)
comb <- split(t(combn(levels(data$sample_side), 2)),seq(nrow(t(combn(levels(data$sample_side), 2)))))
ggplot(data, aes(x = sample_side,y = LibrarySize)) +geom_boxplot(outlier.shape = NA) + 
    geom_jitter(width = 0.2) + labs(x = "sample type") +
    geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black",y_position = c(14000)) + facet_wrap(vars(DNAex_round), scales = "free_y") +
    ggtitle("LibrarySize - DNA extraction batches grouped by sample side")

ggplot(data, aes(x = sample_side,y = LibrarySize)) +geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black",step_increase = 0.1) + facet_wrap(vars(PCR_round), scales = "free_y") + 
  ggtitle("LibrarySize - PCR batches grouped by sample side")
#notice DNAex batch 6 / PCR batch 8 have no normal sample type 
#helpful https://www.statology.org/facet_wrap/
```

Plot1: LibrarySize of DNA extraction batch 1 grouped by sample side
Plot2: LibrarySize of DNA extraction batch 2 grouped by sample side
Plot3: LibrarySize of DNA extraction batch 3 grouped by sample side
Plot4: LibrarySize of DNA extraction batch 4 grouped by sample side
Plot5: LibrarySize of DNA extraction batch 5 grouped by sample side
Plot6: LibrarySize of PCR batch 1 grouped by sample side
Plot7: LibrarySize of PCR batch 2 grouped by sample side

As we expected from visualizing LibrarySize of tissue samples comparing normal tissue vs. tumor tissue - coloring by DNA extraction batch/ PCR batch (Plots 2 and 3) - the differences between LibrarySize for sample side in each batch are not significant.
## Microbial composition
### Microbial composition - species level - tissue samples - grouped by DNA, PCR batch
```{r Microbial composition of tissue samples, DNA extraction batch, PCR batch}

physeq.fam <- transform_sample_counts(p.comb.filt.s.1500.80.true, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.04,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

#microbial composition at species level - grouped by DNA Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level - grouped by DNA Batch")+
  facet_wrap(vars(DNAex_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

#microbial composition at species level - grouped by PCR Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level - grouped by PCR Batch")+
  facet_wrap(vars(PCR_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```

### Microbial Composition at genus level
```{r}
physeq.fam <- transform_sample_counts(p.comb.filt.s.1500.80.true, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$genus.v2 <- ifelse(df.fam.melt$Abundance>0.01,as.character(df.fam.melt$genus), "others") # define cut-off for genus to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$genus.v2 <- factor(df.fam.melt$genus.v2, levels=rev(unique(df.fam.melt$genus.v2)))
header2 <- c("normal" = "Normal pancreas","tumor" = "Tumor tissue") # for plot
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at genus level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~sample_side, scales = "free", labeller = as_labeller(header2))+
  ggtitle("Microbial composition at genus level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
#microbial composition at genus level - grouped by DNA Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at genus level - grouped by DNA Batch")+
  facet_wrap(vars(DNAex_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

#microbial composition at genus level - grouped by PCR Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at genus level - grouped by PCR Batch")+
  facet_wrap(vars(PCR_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```

## Beta diversity
### Beta diversity - tissue samples - sample side
```{r TRUE - BrayCurtis - sample side}
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.true, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.true, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.true)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.true, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix of p.comb.filt.s.1500.80.true, compairing tumor vs. normal
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(p.comb.filt.s.1500.80.true)$sample_side)
p.adonis.pairwise
```
### Beta diversity - tissue samples - DNA extraction batch
```{r TRUE - BrayCurtis - DNAex batches}
#Bray-Curtis for True samples only
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.true, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.true, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.true)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.true, bc_ordination, color="DNAex_round") + 
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 4_5","Batch 5","Batch 6","Batch 7","Batch NA"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix - pairwise compairing DNA extraction batches
#p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.true)$DNAex_round)
#p.adonis.pairwise
```
### Beta diversity - tissue samples - PCR batch
```{r TRUE - BrayCurtis - PCR batches}
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.true, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.true, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.true)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.true, bc_ordination, color="PCR_round", shape = "sample_side") + 
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 5","Batch 6","Batch 7","Batch 8","Batch 9"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix of p.comb.filt.s.1500.80.true, pairwise compairing PCR batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.true)$PCR_round)
p.adonis.pairwise
```
### Beta diversity - NCT samples - sample_side
```{r NCT - BrayCurtis - sample side}
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.nct, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.nct, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.nct)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.nct, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("NCT samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix of p.comb.filt.s.1500.80.nct, pairwise compairing sample_side
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.nct)$sample_side)
p.adonis.pairwise
```
### Beta diversity - NCT samples - DNA extraction batch
```{r NCT - BrayCurtis - DNAex batches}
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.nct, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.nct, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.nct)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.nct, bc_ordination, color="DNAex_round") + 
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 4_5","Batch 5","Batch 6","Batch 7","Batch NA"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("NCT samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix of p.comb.filt.s.1500.80.true, pairwise compairing DNA extraction batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.nct)$DNAex_round)
p.adonis.pairwise
```
### Beta diversity - NCT samples - PCR batch
```{r NCT - BrayCurtis - PCR batches}
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.nct, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.nct, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.nct)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.nct, bc_ordination, color="PCR_round") + 
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 5","Batch 6","Batch 7","Batch 8","Batch 9"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("NCT samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix of p.comb.filt.s.1500.80.true, pairwise compairing PCR batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.nct)$PCR_round)
p.adonis.pairwise
```

```{r create a multitude of PCoA plots using different distance methods}
#https://joey711.github.io/phyloseq/distance.html

#color: DNAex_round | shape: sample_side | pvalue: DNAex_round
dist_methods <- unlist(distanceMethodList) #all available distance methods
dist_methods <- dist_methods[-(1:3)] #remove methods that require a tree
dist_methods = dist_methods[-which(dist_methods=="ANY")] # remove the user-defined distance
plist <- vector("list", length(dist_methods))
names(plist) = dist_methods
for( i in dist_methods ){
    # Calculate distance matrix
    iDist <- distance(p.comb.filt.s.1500.80.true, method=i)
    # Calculate ordination
    iPCoA  <- ordinate(p.comb.filt.s.1500.80.true, "PCoA", distance=iDist)
    ## Make plot
    # Don't carry over previous plot (if error, p will be blank)
    p <- NULL
    # Create plot, store as temp variable, p
    p <- plot_ordination(p.comb.filt.s.1500.80.true, iPCoA, color="DNAex_round", shape="sample_side")
    # Add title to each plot
    p <- p + ggtitle(paste0("PCoA using distance method ", i))
    # Save the graphic to file.
    plist[[i]] = p
}
df = ldply(plist, function(x) x$data)
names(df)[1] <- "distance"
vec0 <- rep(0,dim(df)[1])
# expand dataframe "df" that contains the pvalue for each distance method and add a geom_text to p
for (i in seq(1,dim(df)[1])) {
  iDist <- distance(p.comb.filt.s.1500.80.true, method=df$distance[i])  
  p.adonis <- adonis2(iDist ~ sample_data(p.comb.filt.s.1500.80.true)$DNAex_round)
  pv <- p.adonis$`Pr(>F)`[1]
  vec0[i] <- pv
}
df$xpos <- rep(x = c(-Inf),dim(df)[1])
df$ypos <- rep(x = c(Inf),dim(df)[1])
df$hjustvar <- rep(x = c(-0.2),dim(df)[1])
df$vjustvar <- rep(x = c(1.5),dim(df)[1])
df$pv <- vec0
df2 <- df %>% group_by(distance) %>% summarise_at(vars(pv), list(name = mean)) # https://www.statology.org/r-mean-by-group/ [Method2]
colnames(df2)[2] <- c("pv2")
df2$pv2 <- round(df2$pv2, digits = 3)
df <- left_join(df, df2, by = "distance") # https://statisticsglobe.com/r-dplyr-join-inner-left-right-full-semi-anti
p = ggplot(df, aes(Axis.1, Axis.2, color=DNAex_round, shape=sample_side))
p = p + geom_point(size=3, alpha=0.5)
p = p + geom_text(data=df,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=paste0("mean p-value: ",pv2), inherit.aes = FALSE))
p = p + stat_ellipse()
p = p + facet_wrap(~distance, scales="free")
p = p + ggtitle("PCoA on various distance metrics for True samples")
p
#small issue: pv2 needs to be rounded + why is it in pink?
ggsave(file="PCoA_DNA_true-nct_distmetric.pdf", width=24, height=24, dpi=900)


#color: PCR_round | shape: sample_side | pvalue: PCR_round
plist <- vector("list", length(dist_methods))
names(plist) = dist_methods
for( i in dist_methods ){
    # Calculate distance matrix
    iDist <- distance(p.comb.filt.s.1500.80.true, method=i)
    # Calculate ordination
    iPCoA  <- ordinate(p.comb.filt.s.1500.80.true, "PCoA", distance=iDist)
    ## Make plot
    # Don't carry over previous plot (if error, p will be blank)
    p <- NULL
    # Create plot, store as temp variable, p
    p <- plot_ordination(p.comb.filt.s.1500.80.true, iPCoA, color="PCR_round", shape="sample_side")
    # Add title to each plot
    p <- p + ggtitle(paste0("PCoA using distance method ", i))
    # Save the graphic to file.
    plist[[i]] = p
}
df = ldply(plist, function(x) x$data)
names(df)[1] <- "distance"
vec0 <- rep(0,dim(df)[1])
# expand dataframe "df" that contains the pvalue for each distance method, calculate mean p value for each distance method and plot the calculated p value in each plot
for (i in seq(1,dim(df)[1])) {
  iDist <- distance(p.comb.filt.s.1500.80.true, method=df$distance[i])  
  p.adonis <- adonis2(iDist ~ sample_data(p.comb.filt.s.1500.80.true)$PCR_round)
  pv <- p.adonis$`Pr(>F)`[1]
  vec0[i] <- pv
}
# food for thought - currently we take each ENTRY into df$distance and calculate the pvalue. A quicker method would be to calculate distance() for each unique method and calculate adonis2 | pvalue based on that → this would also create 1 pvalue for each method and we won´t have to calculate an average pvalue. TL:DR inefficient code calculates unnessesary amounts of pvalue-s. this drives up computational time
df$xpos <- rep(x = c(-Inf),dim(df)[1])
df$ypos <- rep(x = c(Inf),dim(df)[1])
df$hjustvar <- rep(x = c(-0.2),dim(df)[1])
df$vjustvar <- rep(x = c(1.5),dim(df)[1])
df$pv <- vec0
df2 <- df %>% group_by(distance) %>% summarise_at(vars(pv), list(name = mean)) # https://www.statology.org/r-mean-by-group/ [Method2]
colnames(df2)[2] <- c("pv2")
df2$pv2 <- round(df2$pv2, digits = 3)
df <- left_join(df, df2, by = "distance") # https://statisticsglobe.com/r-dplyr-join-inner-left-right-full-semi-anti
p = ggplot(df, aes(Axis.1, Axis.2, color=PCR_round, shape=sample_side))
p = p + geom_point(size=3, alpha=0.5)
p = p + geom_text(data=df,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=paste0("mean p-value: ",pv2), inherit.aes = FALSE))
p = p + stat_ellipse()
p = p + facet_wrap(~distance, scales="free")
p = p + ggtitle("PCoA on various distance metrics for True samples")
p
ggsave(file="PCoA_PCR_true-nct_distmetric.pdf", width=24, height=24, dpi=900)

#color: sample_side | shape: sample_side | pvalue: sample_side
plist <- vector("list", length(dist_methods))
names(plist) = dist_methods
for( i in dist_methods ){
    # Calculate distance matrix
    iDist <- distance(p.comb.filt.s.1500.80.true, method=i)
    # Calculate ordination
    iPCoA  <- ordinate(p.comb.filt.s.1500.80.true, "PCoA", distance=iDist)
    ## Make plot
    # Don't carry over previous plot (if error, p will be blank)
    p <- NULL
    # Create plot, store as temp variable, p
    p <- plot_ordination(p.comb.filt.s.1500.80.true, iPCoA, color="sample_side", shape="sample_side")
    # Add title to each plot
    p <- p + ggtitle(paste0("PCoA using distance method ", i))
    # Save the graphic to file.
    plist[[i]] = p
}
df = ldply(plist, function(x) x$data)
names(df)[1] <- "distance"
vec0 <- rep(0,dim(df)[1])
# expand dataframe "df" that contains the pvalue for each distance method, calculate mean p value for each distance method and plot the calculated p value in each plot
for (i in seq(1,dim(df)[1])) {
  iDist <- distance(p.comb.filt.s.1500.80.true, method=df$distance[i])  
  p.adonis <- adonis2(iDist ~ sample_data(p.comb.filt.s.1500.80.true)$sample_side)
  pv <- p.adonis$`Pr(>F)`[1]
  vec0[i] <- pv
}
# food for thought - currently we take each ENTRY into df$distance and calculate the pvalue. A quicker method would be to calculate distance() for each unique method and calculate adonis2 | pvalue based on that → this would also create 1 pvalue for each method and we won´t have to calculate an average pvalue. TL:DR inefficient code calculates unnessesary amounts of pvalue-s. this drives up computational time
df$xpos <- rep(x = c(-Inf),dim(df)[1])
df$ypos <- rep(x = c(Inf),dim(df)[1])
df$hjustvar <- rep(x = c(-0.2),dim(df)[1])
df$vjustvar <- rep(x = c(1.5),dim(df)[1])
df$pv <- vec0
df2 <- df %>% group_by(distance) %>% summarise_at(vars(pv), list(name = mean)) # https://www.statology.org/r-mean-by-group/ [Method2]
colnames(df2)[2] <- c("pv2")
df2$pv2 <- round(df2$pv2, digits = 3)
df <- left_join(df, df2, by = "distance") # https://statisticsglobe.com/r-dplyr-join-inner-left-right-full-semi-anti
p = ggplot(df, aes(Axis.1, Axis.2, color=sample_side, shape=sample_side))
p = p + geom_point(size=3, alpha=0.5)
p = p + geom_text(data=df,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=paste0("mean p-value: ",pv2), inherit.aes = FALSE))
p = p + stat_ellipse()
p = p + facet_wrap(~distance, scales="free")
p = p + ggtitle("PCoA on various distance metrics for True samples")
p
ggsave(file="PCoA_true-nct_distmetric.pdf", width=24, height=24, dpi=900)
```

#ContaminantRemover: remove taxa from tissue samples - if taxa is present in NCT samples
## Venn diagram: taxa found in tissue samples, NCT samples or both
```{r ContaminantRemover: Venn diagram showing taxa found in tissue samples, NCT samples or both}
ntaxa <- as.integer(ntaxa(p.comb.filt.s.1500.80))
zerotaxasum <- which(taxa_sums(p.comb.filt.s.1500.80) == 0) #which taxa from p.comb.filt.s.1500.80 have 0 total read counts
nzerotaxasum <- length(zerotaxasum)
zerotaxasum.true <- which(taxa_sums(p.comb.filt.s.1500.80.true) == 0) #which taxa have 0 total read counts in just TRUE samples -> ONLY present in NCT samples
nzerotaxasum.true <- length(zerotaxasum.true)
zerotaxasum.nct <- which(taxa_sums(p.comb.filt.s.1500.80.nct) == 0) #which taxa have 0 total read counts in just NCT samples -> ONLY present in TRUE samples
nzerotaxasum.nct <- length(zerotaxasum.nct)
nmixedtaxa <- ntaxa - (nzerotaxasum.true+nzerotaxasum.nct)

# move to new plotting page
grid.newpage()
# create pairwise Venn diagram
draw.pairwise.venn(area1=c(nzerotaxasum.nct+nmixedtaxa), area2=c(nzerotaxasum.true+nmixedtaxa),cross.area=nmixedtaxa,
                   category=c("Tissue","NCT"),fill=c("Blue","Yellow"))

tissue <- p.comb.filt.s.1500.80@tax_table@.Data[-as.vector(zerotaxasum.true),7]
nct <- p.comb.filt.s.1500.80@tax_table@.Data[-as.vector(zerotaxasum.nct),7]
A <- list(Tissue = tissue ,NCT = nct)
ggvenn(A)
```

## Phyloseq: Taxa found ONLY tissue samples
```{r ContaminantRemover}
p.comb.filt.s.1500.80.true.nozero <- subset_taxa(p.comb.filt.s.1500.80.true, taxa_names(p.comb.filt.s.1500.80.true) %in% taxa_names(p.comb.filt.s.1500.80)[zerotaxasum.nct]) #take TRUE samples, keep taxa only present in tissue samples
```


```{r contaminant filtering: remove all taxa found in negative controls - no correction for batch effect := ContaminantRemover}
df <- meta(p.comb.filt.s.1500.80.true.nozero) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80.true.nozero)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
## Scatterplots of library size - naive, colored by DNA extraction batch / PCR batch
#plot librarysize
ggplot(data=df, aes(x=Index, y=LibrarySize)) + geom_point()  + theme_bw() + ggtitle("Library size p.comb.filt.s.1500.80.true.nozero")+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare DNAex_batch counts for plotting
DNAex_batch1_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(DNAex_round == "1") %>% count("DNAex_round") %>% pull()
DNAex_batch2_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(DNAex_round == "2") %>% count("DNAex_round") %>% pull()
DNAex_batch3_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(DNAex_round == "3") %>% count("DNAex_round") %>% pull()
DNAex_batch4_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(DNAex_round == "4") %>% count("DNAex_round") %>% pull()
DNAex_batch5_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(DNAex_round == "5") %>% count("DNAex_round") %>% pull()
DNAex_batch6_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(DNAex_round == "6") %>% count("DNAex_round") %>% pull()
DNAex_batch7_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(DNAex_round == "7") %>% count("DNAex_round") %>% pull()
DNAex_batch4_5_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(DNAex_round == "4_5") %>% count("DNAex_round") %>% pull()
DNAex_batchNA_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(DNAex_round == "NA") %>% count("DNAex_round") %>% pull()
#plot librarysize colored by DNAex_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(DNAex_round))) + geom_point()  + theme_bw() + ggtitle("Library size -  p.comb.filt.s.1500.80.true.nozero")+scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c(paste("Batch 1\nn =", DNAex_batch1_count, sep = " "),paste("Batch 2\nn =", DNAex_batch2_count, sep = " "),paste("Batch 3\nn =", DNAex_batch3_count, sep = " "),paste("Batch 4\nn =", DNAex_batch4_count, sep = " "),paste("Batch 4_5\nn =", DNAex_batch4_5_count, sep = " "),paste("Batch 5\nn =", DNAex_batch5_count, sep = " "),paste("Batch 6\nn =", DNAex_batch6_count, sep = " "),paste("Batch 7\nn =", DNAex_batch7_count, sep = " "),paste("Batch NA\nn =", DNAex_batchNA_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare PCR_batch counts for plotting
PCR_batch1_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(PCR_round == "1") %>% count("PCR_round") %>% pull()
PCR_batch2_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(PCR_round == "2") %>% count("PCR_round") %>% pull()
PCR_batch3_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(PCR_round == "3") %>% count("PCR_round") %>% pull()
PCR_batch4_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(PCR_round == "4") %>% count("PCR_round") %>% pull()
PCR_batch5_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(PCR_round == "5") %>% count("PCR_round") %>% pull()
PCR_batch6_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(PCR_round == "6") %>% count("PCR_round") %>% pull()
PCR_batch7_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(PCR_round == "7") %>% count("PCR_round") %>% pull()
PCR_batch8_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(PCR_round == "8") %>% count("PCR_round") %>% pull()
PCR_batch9_count <- meta(p.comb.filt.s.1500.80.true.nozero) %>% filter(PCR_round == "9") %>% count("PCR_round") %>% pull()
#plot librarysize colored by PCR_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(PCR_round))) + geom_point()  + theme_bw() + ggtitle("Library size - p.comb.filt.s.1500.80.true.nozero")+scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c(paste("Batch 1\nn =", PCR_batch1_count, sep = " "),paste("Batch 2\nn =", PCR_batch2_count, sep = " "),paste("Batch 3\nn =", PCR_batch3_count, sep = " "),paste("Batch 4\nn =", PCR_batch4_count, sep = " "),paste("Batch 5\nn =", PCR_batch5_count, sep = " "),paste("Batch 6\nn =", PCR_batch6_count, sep = " "),paste("Batch 7\nn =", PCR_batch7_count, sep = " "),paste("Batch 8\nn =", PCR_batch8_count, sep = " "),paste("Batch 9\nn =", PCR_batch9_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))

#Boxplots of library size - grouped by tissue type, colored by DNA extraction batch / PCR batch
#set levels
df$sample_side <- factor(df$sample_side)
#create comb
comb <- split(t(combn(levels(df$sample_side), 2)),seq(nrow(t(combn(levels(df$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df, aes(x = sample_side, y = LibrarySize)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by sample side") +
  theme(text = element_text(size = 10))
#set levels
df$DNAex_round <- factor(df$DNAex_round)
#create comb: pairwise comparisons
comb <- split(t(combn(levels(df$DNAex_round), 2)),seq(nrow(t(combn(levels(df$DNAex_round), 2)))))
#boxplot comparing librarysizes of different DNAex batches
ggplot(df, aes(x = DNAex_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "DNA extraction round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.02,comparisons = comb, map_signif_level = FALSE,y_position = seq.int(4000,21000,900),color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by DNA extraction batch") +
    theme(text = element_text(size = 10))
#set levels
df$PCR_round <- factor(df$PCR_round)
#create comb: pairwise comparisons
comb <- split(t(combn(levels(df$PCR_round), 2)),seq(nrow(t(combn(levels(df$PCR_round), 2)))))
#boxplot comparing librarysizes of different PCR batches
ggplot(df, aes(x = PCR_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "PCR round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.1,comparisons = comb, map_signif_level = FALSE,color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by PCR extraction batch") +
    theme(text = element_text(size = 10))

# Microbial composition
physeq.fam <- transform_sample_counts(p.comb.filt.s.1500.80.true.nozero, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.125,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "magenta2", "palegreen", "salmon", "maroon", "cyan2")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
#microbial composition at species level - grouped by DNA Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level - grouped by DNA Batch")+
  facet_wrap(vars(DNAex_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
#microbial composition at species level - grouped by PCR Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level - grouped by PCR Batch")+
  facet_wrap(vars(PCR_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```

###Preperation for diversity metrics
```{r ContaminantRemover: Preperation for diversity metrics}
zerosamplesum.true <- which(sample_sums(p.comb.filt.s.1500.80.true.nozero) == 0) #which samples have sample_sum = 0 -> these samples need to be removed to calculate distance metric, dissimilarity matrix
nzerosamplesum.true <- length(zerosamplesum.true) #amount of samples with total counts of 0

#create new phyloseq object: removing samples with 0 total read counts
p.comb.filt.s.1500.80.true.nozero.both <- subset_samples(p.comb.filt.s.1500.80.true.nozero, !(sample_names(p.comb.filt.s.1500.80.true.nozero) %in% sample_names(p.comb.filt.s.1500.80.true.nozero)[zerosamplesum.true]))
```

### Alpha diversity
```{r ContaminantRemover: Alpha diversity}
# Alpha diversity
aindex <- estimate_richness(physeq = otu_table(p.comb.filt.s.1500.80.true.nozero.both), measures = c("Observed","Shannon","Simpson","InvSimpson", "Chao1","ACE"))

aindex$sample_side <- meta(p.comb.filt.s.1500.80.true.nozero.both)$sample_side
aindex$sample_side <- factor(aindex$sample_side)

comb <- split(t(combn(levels(aindex$sample_side), 2)),seq(nrow(t(combn(levels(aindex$sample_side), 2)))))
data <- reshape2::melt(aindex)
ggplot(data, aes(x = sample_side,y = value)) +
  # Outliers are removed, because otherwise each data point would be plotted twice; 
  # as an outlier of boxplot and as a point of dotplot.
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + 
  ggtitle("ContaminantRemover approach - alpha diversity") +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE,) +
  theme(text = element_text(size = 10)) +
  facet_wrap(~ variable, scales = "free_y")
```

### Beta diversity - tissue samples - sample side
```{r ContaminantRemover: BrayCurtis - sample side}
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.true.nozero.both, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.true.nozero.both, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.true.nozero.both)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.true.nozero.both, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ContaminantRemover - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)
#introduce Pairwise.adonis for dissimilarity matrix - compairing tumor vs. normal
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(p.comb.filt.s.1500.80.true.nozero.both)$sample_side)
p.adonis.pairwise
```

### Beta diversity - tissue samples - DNA extraction batch
```{r ContaminantRemover: Bray Curtis - DNA extraction batch}
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.true.nozero.both, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.true.nozero.both, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.true.nozero.both)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.true.nozero.both, bc_ordination, color="DNAex_round", shape = "sample_side") + 
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 4_5","Batch 5","Batch 6","Batch 7","Batch NA"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ContaminantRemover - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)
#introduce Pairwise.adonis for dissimilarity matrix - pairwise compairing DNA extraction batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.true.nozero.both)$DNAex_round)
p.adonis.pairwise
```

### Beta diversity - tissue samples - PCR batch
```{r ContaminantRemover: Bray Curtis - PCR batch}
# color by pcr batch
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.true.nozero.both, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.true.nozero.both, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.true.nozero.both)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.true.nozero.both, bc_ordination, color="PCR_round", shape = "sample_side") + 
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 5","Batch 6","Batch 7","Batch 8","Batch 9"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ContaminantRemover - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)
```

# Thoughts
```{r}
#honest hypothesis: ConQuR, lower limit on read count {< 1e^-04, PERFect → REread paper: they order taxa based on appearance}, Nejman Filter-approach
```
# Solo-PERFect
## Sim_PERFect
```{r}
# Perfect prevalence filtering
otu = as(otu_table(p.comb.filt.s.1500.80.true), "matrix")
if(taxa_are_rows(p.comb.filt.s.1500.80.true)){otu <- t(otu)}
otu = as_tibble(otu)
res_sim <- PERFect_sim(X = otu,alpha = 0.05)
dim(res_sim$filtX)
ids.sim<- colnames(res_sim$filtX)
physeq.control <-  prune_taxa(ids.sim, p.comb.filt.s.1500.80.true)
# alpha at 0.15 keeps 841
# alpha at 0.1 keeps 144
# alpha at 0.05 keeps 129
```
### Library Size
```{r}
#prepare dataframe for plotting
df <- meta(physeq.control)
df$LibrarySize <- sample_sums(physeq.control)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
#plot librarysize 
ggplot(data=df, aes(x=Index, y=LibrarySize)) + geom_point()  + theme_bw() + ggtitle("Library size -  solo_PERFect")+
  scale_color_brewer(palette = "Set1")+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare DNAex_batch counts for plotting
DNAex_batch1_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "1") %>% count("DNAex_round") %>% pull()
DNAex_batch2_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "2") %>% count("DNAex_round") %>% pull()
DNAex_batch3_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "3") %>% count("DNAex_round") %>% pull()
DNAex_batch4_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "4") %>% count("DNAex_round") %>% pull()
DNAex_batch5_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "5") %>% count("DNAex_round") %>% pull()
DNAex_batch6_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "6") %>% count("DNAex_round") %>% pull()
DNAex_batch7_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "7") %>% count("DNAex_round") %>% pull()
DNAex_batch4_5_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "4_5") %>% count("DNAex_round") %>% pull()
DNAex_batchNA_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "NA") %>% count("DNAex_round") %>% pull()
#plot librarysize colored by DNAex_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(DNAex_round))) + geom_point()  + theme_bw() + ggtitle("Library size -  solo_PERFect")+scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c(paste("Batch 1\nn =", DNAex_batch1_count, sep = " "),paste("Batch 2\nn =", DNAex_batch2_count, sep = " "),paste("Batch 3\nn =", DNAex_batch3_count, sep = " "),paste("Batch 4\nn =", DNAex_batch4_count, sep = " "),paste("Batch 4_5\nn =", DNAex_batch4_5_count, sep = " "),paste("Batch 5\nn =", DNAex_batch5_count, sep = " "),paste("Batch 6\nn =", DNAex_batch6_count, sep = " "),paste("Batch 7\nn =", DNAex_batch7_count, sep = " "),paste("Batch NA\nn =", DNAex_batchNA_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare PCR_batch counts for plotting
PCR_batch1_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "1") %>% count("PCR_round") %>% pull()
PCR_batch2_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "2") %>% count("PCR_round") %>% pull()
PCR_batch3_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "3") %>% count("PCR_round") %>% pull()
PCR_batch4_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "4") %>% count("PCR_round") %>% pull()
PCR_batch5_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "5") %>% count("PCR_round") %>% pull()
PCR_batch6_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "6") %>% count("PCR_round") %>% pull()
PCR_batch7_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "7") %>% count("PCR_round") %>% pull()
PCR_batch8_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "8") %>% count("PCR_round") %>% pull()
PCR_batch9_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "9") %>% count("PCR_round") %>% pull()
#plot librarysize colored by PCR_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(PCR_round))) + geom_point()  + theme_bw() + ggtitle("Library size -  solo_PERFect")+scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c(paste("Batch 1\nn =", PCR_batch1_count, sep = " "),paste("Batch 2\nn =", PCR_batch2_count, sep = " "),paste("Batch 3\nn =", PCR_batch3_count, sep = " "),paste("Batch 4\nn =", PCR_batch4_count, sep = " "),paste("Batch 5\nn =", PCR_batch5_count, sep = " "),paste("Batch 6\nn =", PCR_batch6_count, sep = " "),paste("Batch 7\nn =", PCR_batch7_count, sep = " "),paste("Batch 8\nn =", PCR_batch8_count, sep = " "),paste("Batch 9\nn =", PCR_batch9_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
```
### Microbial Composition at species level
```{r}
physeq.fam <- transform_sample_counts(physeq.control, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.04,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
header2 <- c("normal" = "Normal pancreas","tumor" = "Tumor tissue") # for plot
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "magenta2", "palegreen", "salmon", "maroon", "cyan2")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~sample_side, scales = "free", labeller = as_labeller(header2))+
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
#microbial composition at species level - grouped by DNA Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level - grouped by DNA Batch")+
  facet_wrap(vars(DNAex_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

#microbial composition at species level - grouped by PCR Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level - grouped by PCR Batch")+
  facet_wrap(vars(PCR_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```
### Microbial Composition at genus level
```{r}
physeq.fam <- transform_sample_counts(physeq.control, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$genus.v2 <- ifelse(df.fam.melt$Abundance>0.04,as.character(df.fam.melt$genus), "others") # define cut-off for genus to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$genus.v2 <- factor(df.fam.melt$genus.v2, levels=rev(unique(df.fam.melt$genus.v2)))
header2 <- c("normal" = "Normal pancreas","tumor" = "Tumor tissue") # for plot
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at genus level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~sample_side, scales = "free", labeller = as_labeller(header2))+
  ggtitle("Microbial composition at genus level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
#microbial composition at genus level - grouped by DNA Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at genus level - grouped by DNA Batch")+
  facet_wrap(vars(DNAex_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

#microbial composition at genus level - grouped by PCR Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at genus level - grouped by PCR Batch")+
  facet_wrap(vars(PCR_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```
### Alpha diversity
```{r}
aindex <- estimate_richness(physeq = otu_table(physeq.control), measures = c("Observed","Shannon","Simpson","InvSimpson", "Chao1","ACE"))

aindex$sample_side <- meta(physeq.control)$sample_side
aindex$sample_side <- factor(aindex$sample_side)

comb <- split(t(combn(levels(aindex$sample_side), 2)),seq(nrow(t(combn(levels(aindex$sample_side), 2)))))
data <- reshape2::melt(aindex)
ggplot(data, aes(x = sample_side,y = value)) +
  # Outliers are removed, because otherwise each data point would be plotted twice; 
  # as an outlier of boxplot and as a point of dotplot.
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + 
  ggtitle("ContaminantRemover approach - alpha diversity") +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE,) +
  theme(text = element_text(size = 10)) +
  facet_wrap(~ variable, scales = "free_y")
```

### Beta diversity
```{r}
bray_dist = phyloseq::distance(physeq.control, method="bray")
bc_ordination = ordinate(physeq.control, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(physeq.control)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(physeq.control, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control)$sample_side)
p.adonis.pairwise
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control)$DNAex_round)
p.adonis.pairwise
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control)$PCR_round)
p.adonis.pairwise

p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control)$sample_side)
p.adonis.pairwise
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control)$DNAex_round)
p.adonis.pairwise
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control)$PCR_round)
p.adonis.pairwise
```
## Perm_PERFect
```{r}
otu = as(otu_table(p.comb.filt.s.1500.80.true), "matrix")
if(taxa_are_rows(p.comb.filt.s.1500.80.true)){otu <- t(otu)}
otu = as_tibble(otu)
res_sim <- PERFect_sim(X = otu)
dim(res_sim$filtX)
ids.sim<- colnames(res_sim$filtX)
physeq.control <-  prune_taxa(ids.sim, p.comb.filt.s.1500.80.true)

ord.pval <- pvals_Order(Counts = otu,res_sim = res_sim )

otu = as(otu_table(p.comb.filt.s.1500.80.true), "matrix")
if(taxa_are_rows(p.comb.filt.s.1500.80.true)){otu <- t(otu)}
otu = as_tibble(otu)
res_perm <- PERFect_perm(X = otu, Order.user = ord.pval,alpha = 0.1)
dim(res_perm$filtX)
ids.sim<- colnames(res_perm$filtX)
physeq.control2 <-  prune_taxa(ids.sim, p.comb.filt.s.1500.80.true)
physeq.control2
```

### LibrarySize
```{r}
#prepare dataframe for plotting
df <- meta(physeq.control2)
df$LibrarySize <- sample_sums(physeq.control2)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
#plot librarysize 
ggplot(data=df, aes(x=Index, y=LibrarySize)) + geom_point()  + theme_bw() + ggtitle("Library size -  solo_PERFect")+
  scale_color_brewer(palette = "Set1")+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare DNAex_batch counts for plotting
DNAex_batch1_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "1") %>% count("DNAex_round") %>% pull()
DNAex_batch2_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "2") %>% count("DNAex_round") %>% pull()
DNAex_batch3_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "3") %>% count("DNAex_round") %>% pull()
DNAex_batch4_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "4") %>% count("DNAex_round") %>% pull()
DNAex_batch5_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "5") %>% count("DNAex_round") %>% pull()
DNAex_batch6_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "6") %>% count("DNAex_round") %>% pull()
DNAex_batch7_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "7") %>% count("DNAex_round") %>% pull()
DNAex_batch4_5_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "4_5") %>% count("DNAex_round") %>% pull()
DNAex_batchNA_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "NA") %>% count("DNAex_round") %>% pull()
#plot librarysize colored by DNAex_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(DNAex_round))) + geom_point()  + theme_bw() + ggtitle("Library size -  solo_PERFect")+scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c(paste("Batch 1\nn =", DNAex_batch1_count, sep = " "),paste("Batch 2\nn =", DNAex_batch2_count, sep = " "),paste("Batch 3\nn =", DNAex_batch3_count, sep = " "),paste("Batch 4\nn =", DNAex_batch4_count, sep = " "),paste("Batch 4_5\nn =", DNAex_batch4_5_count, sep = " "),paste("Batch 5\nn =", DNAex_batch5_count, sep = " "),paste("Batch 6\nn =", DNAex_batch6_count, sep = " "),paste("Batch 7\nn =", DNAex_batch7_count, sep = " "),paste("Batch NA\nn =", DNAex_batchNA_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare PCR_batch counts for plotting
PCR_batch1_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "1") %>% count("PCR_round") %>% pull()
PCR_batch2_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "2") %>% count("PCR_round") %>% pull()
PCR_batch3_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "3") %>% count("PCR_round") %>% pull()
PCR_batch4_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "4") %>% count("PCR_round") %>% pull()
PCR_batch5_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "5") %>% count("PCR_round") %>% pull()
PCR_batch6_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "6") %>% count("PCR_round") %>% pull()
PCR_batch7_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "7") %>% count("PCR_round") %>% pull()
PCR_batch8_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "8") %>% count("PCR_round") %>% pull()
PCR_batch9_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "9") %>% count("PCR_round") %>% pull()
#plot librarysize colored by PCR_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(PCR_round))) + geom_point()  + theme_bw() + ggtitle("Library size -  solo_PERFect")+scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c(paste("Batch 1\nn =", PCR_batch1_count, sep = " "),paste("Batch 2\nn =", PCR_batch2_count, sep = " "),paste("Batch 3\nn =", PCR_batch3_count, sep = " "),paste("Batch 4\nn =", PCR_batch4_count, sep = " "),paste("Batch 5\nn =", PCR_batch5_count, sep = " "),paste("Batch 6\nn =", PCR_batch6_count, sep = " "),paste("Batch 7\nn =", PCR_batch7_count, sep = " "),paste("Batch 8\nn =", PCR_batch8_count, sep = " "),paste("Batch 9\nn =", PCR_batch9_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
```

### Microbial Composition at species level
```{r}
physeq.fam <- transform_sample_counts(physeq.control2, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.08,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
header2 <- c("normal" = "Normal pancreas","tumor" = "Tumor tissue") # for plot
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~sample_side, scales = "free", labeller = as_labeller(header2))+
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
#microbial composition at species level - grouped by DNA Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level - grouped by DNA Batch")+
  facet_wrap(vars(DNAex_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

#microbial composition at species level - grouped by PCR Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level - grouped by PCR Batch")+
  facet_wrap(vars(PCR_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```

### Microbial Composition at genus level
```{r}
physeq.fam <- transform_sample_counts(physeq.control2, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$genus.v2 <- ifelse(df.fam.melt$Abundance>0.02,as.character(df.fam.melt$genus), "others") # define cut-off for genus to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$genus.v2 <- factor(df.fam.melt$genus.v2, levels=rev(unique(df.fam.melt$genus.v2)))
header2 <- c("normal" = "Normal pancreas","tumor" = "Tumor tissue") # for plot
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at genus level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~sample_side, scales = "free", labeller = as_labeller(header2))+
  ggtitle("Microbial composition at genus level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
#microbial composition at genus level - grouped by DNA Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at genus level - grouped by DNA Batch")+
  facet_wrap(vars(DNAex_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

#microbial composition at genus level - grouped by PCR Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at genus level - grouped by PCR Batch")+
  facet_wrap(vars(PCR_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```

### Alpha diversity
```{r}
aindex <- estimate_richness(physeq = otu_table(physeq.control2), measures = c("Observed","Shannon","Simpson","InvSimpson", "Chao1","ACE"))

aindex$sample_side <- meta(physeq.control2)$sample_side
aindex$sample_side <- factor(aindex$sample_side)

comb <- split(t(combn(levels(aindex$sample_side), 2)),seq(nrow(t(combn(levels(aindex$sample_side), 2)))))
data <- reshape2::melt(aindex)
ggplot(data, aes(x = sample_side,y = value)) +
  # Outliers are removed, because otherwise each data point would be plotted twice; 
  # as an outlier of boxplot and as a point of dotplot.
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + 
  ggtitle("ContaminantRemover approach - alpha diversity") +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE,) +
  theme(text = element_text(size = 10)) +
  facet_wrap(~ variable, scales = "free_y")
```

### Beta diversity
```{r}
bray_dist = phyloseq::distance(physeq.control2, method="bray")
bc_ordination = ordinate(physeq.control2, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(physeq.control2)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(physeq.control2, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control2)$sample_side)
p.adonis.pairwise
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control2)$DNAex_round)
p.adonis.pairwise
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control2)$PCR_round)
p.adonis.pairwise

p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control2)$sample_side)
p.adonis.pairwise
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control2)$DNAex_round)
p.adonis.pairwise
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(physeq.control2)$PCR_round)
p.adonis.pairwise
```


# Nejman as control mechanism
```{r}
#Check our data for results from nejman
#Nejman found for PDAC: Fusobacterium nucleatum, Enterobacter asburiae, Klebsiella pneumoniae, Citrobacter freundii, Enterobacter cloacae

# Fusobacterium nucleatum
taxa_position <- which(tax_table(p.comb.filt.s.1500.80)[,7] == "Fusobacterium nucleatum") #which OTU encodes the species name
##ggplot?
plot(x = seq(1,nsamples(p.comb.filt.s.1500.80)),y = otu_table(p.comb.filt.s.1500.80)[taxa_position,])
sample_position <- which(otu_table(p.comb.filt.s.1500.80)[taxa_position,] > 0) #index of which samples are positive for the species | we use the phyloseq object containing both NCT samples and Tissue samples
sample_names(p.comb.filt.s.1500.80)[sample_position] #names of samples positive for species

physeq.melt <- psmelt(p.comb.filt.s.1500.80)
df.phy.melt <- physeq.melt %>% filter(species == c("Fusobacterium nucleatum","Enterobacter asburiae","Klebsiella pneumoniae","Citrobacter freundii","Enterobacter cloacae"))
df.phy.melt

ggplot(data = df.phy.melt,)
```
# ConQuR
## 1. ConQuR
### Application + PCoA control
```{r}
#transform otu_table into matrix
otu = as(otu_table(p.comb.filt.s.1500.80.true), "matrix")
#transpose matrix such that
if(taxa_are_rows(p.comb.filt.s.1500.80.true)){otu <- t(otu)}
#create df based on sample_data. give factors for variables 
df.mod <- meta(p.comb.filt.s.1500.80.true)
df.mod <- meta(p.comb.filt.s.1500.80.true) %>% mutate(sample_sidev2 = as.factor(sample_side),
                                                      sample_nrv2 = as.factor(sample_nr),
                                                      PCR_inputv2 = as.factor(PCR_input),
                                                      ng_in_fcv2 = as.numeric(gsub(",",".",x = df.mod$ng_in_fc)),
                                                      PCR.eluate.ng.µLv2 = as.numeric(gsub(",",".",x = df.mod$PCR.eluate.ng.µL)),
                                                      µL.in.libraryv2 = as.numeric(gsub(",",".",x = df.mod$µL.in.library)),
                                                      PCR_roundv2 = as.factor(PCR_round),
                                                      DNAex_roundv2 = as.factor(DNAex_round))
#define batch for ConQuR - here DNA extraction batch
batchid = df.mod$DNAex_roundv2
#define covariates for ConQuR
covar = df.mod[,c("sample_sidev2")]
#suppress benign warning
options(warn=-1)
#perform ConQuR - for DNA extraction batches we choose batch no. 1 as reference batch
taxa_corrected1 = ConQuR(tax_tab = otu, batchid = batchid, covariates = covar, batch_ref = "1")

#Plot PCoA using BrayCurtis for original data (otu) and modified data (taxa_corrected1)
Plot_PCoA(TAX = otu,factor = batchid,main="Before Correction, Bray Curtis")
Plot_PCoA(TAX = taxa_corrected1,factor = batchid ,main = "ConQuR, Bray Curtis")
#Plot PCoA using Aitchison for original data (otu) and modified data (taxa_corrected1)
Plot_PCoA(TAX = otu,factor = batchid,dissimilarity = "Aitch",main="Before Correction, Aitchison")
Plot_PCoA(TAX = taxa_corrected1,factor = batchid,dissimilarity = "Aitch",main = "ConQuR, Aitchison")
```
### newPhyloseq
```{r}
#based on new data create new phyloseq
p.comb.filt.s.1500.80.postConQuR1 <- merge_phyloseq(otu_table(t(taxa_corrected1),taxa_are_rows = TRUE), tax_table(p.comb.filt.s.1500.80.true), sample_data(p.comb.filt.s.1500.80.true))

p.comb.filt.s.1500.80.postConQuR1 <- prune_taxa(taxa = (taxa_names(p.comb.filt.s.1500.80.postConQuR1)[-which(taxa_sums(p.comb.filt.s.1500.80.postConQuR1) == 0)]),x = p.comb.filt.s.1500.80.postConQuR1)
```
### Adonis2 / PERMANOVA
```{r}
#calculate bray_dist of our new phyloseq
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.postConQuR1, method="bray")
#omnibus testing for DNA extraction batch, PCR batch, sample_side
adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR1)$DNAex_round)
adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR1)$PCR_round)
adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR1)$sample_side)
#pairwise testing inside DNA extraction batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.postConQuR1)$DNAex_round)
p.adonis.pairwise
#pairwise testing inside PCR batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.postConQuR1)$PCR_round)
p.adonis.pairwise
#pairwise testing of "sample_side" {Redundant as sample_side consists of "normal" & "tumor"}
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.postConQuR1)$sample_side)
p.adonis.pairwise
```
### Beta diversity
```{r}
#beta diversity of new data - colored by sample side
bc_ordination = ordinate(p.comb.filt.s.1500.80.postConQuR1, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR1)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.postConQuR1, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - ConQuR1 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)
#beta diversity of new data - colored by DNA extraction batch
bc_ordination = ordinate(p.comb.filt.s.1500.80.postConQuR1, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR1)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.postConQuR1, bc_ordination, color="DNAex_round") +
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 4_5","Batch 5","Batch 6","Batch 7","Batch NA"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - ConQuR1 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)
#beta diversity of new data - colored by PCR batch
bc_ordination = ordinate(p.comb.filt.s.1500.80.postConQuR1, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR1)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.postConQuR1, bc_ordination, color="PCR_round") +
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 5","Batch 6","Batch 7","Batch 8","Batch 9"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - ConQuR1 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)
```

## 2. ConQuR
### Application + PCoA control
```{r}
#use the same df as for 1. ConQuR
df.mod = df.mod
#define batch as PCR
batchid = df.mod$PCR_roundv2 
#define covariates
covar = df.mod[,c("sample_sidev2")]
#perform ConQuR - for PCR batches we choose batch no. 1 as reference batch
#suppress benign warning
options(warn=-1)
# use tax_tab = taxa_corrected1 as we build on pre ConQuR´d data
taxa_corrected2 = ConQuR(tax_tab = taxa_corrected1, batchid = batchid, covariates = covar, batch_ref = "8")

Plot_PCoA(TAX = otu,factor = batchid,main="Before Correction, Bray Curtis")
Plot_PCoA(TAX = taxa_corrected1,factor = batchid ,main = "ConQuR, DNAex-corrected, Bray Curtis")
Plot_PCoA(TAX = taxa_corrected2,factor = batchid ,main = "ConQuR, all-corrected, Bray Curtis")
Plot_PCoA(TAX = otu,factor = batchid,dissimilarity = "Aitch",main="Before Correction, Aitchison")
Plot_PCoA(TAX = taxa_corrected1,factor = batchid,dissimilarity = "Aitch",main = "ConQuR, DNAex-corrected, Aitchison")
Plot_PCoA(TAX = taxa_corrected2,factor = batchid,dissimilarity = "Aitch",main = "ConQuR, all-corrected, Aitchison")

#suppress benign warning
options(warn=-1)
taxa_corrected = ConQuR(tax_tab = taxa_corrected1, batchid = batchid, covariates = covar, batch_ref = "1")

Plot_PCoA(TAX = otu,factor = batchid,main="Before Correction, Bray Curtis")
Plot_PCoA(TAX = taxa_corrected1,factor = batchid ,main = "ConQuR, DNAex-corrected, Bray Curtis")
Plot_PCoA(TAX = taxa_corrected,factor = batchid ,main = "ConQuR, all-corrected, Bray Curtis")
Plot_PCoA(TAX = otu,factor = batchid,dissimilarity = "Aitch",main="Before Correction, Aitchison")
Plot_PCoA(TAX = taxa_corrected1,factor = batchid,dissimilarity = "Aitch",main = "ConQuR, DNAex-corrected, Aitchison")
Plot_PCoA(TAX = taxa_corrected,factor = batchid,dissimilarity = "Aitch",main = "ConQuR, all-corrected, Aitchison")
```
###newPhyloseq
```{r}
p.comb.filt.s.1500.80.postConQuR2 <- merge_phyloseq(otu_table(t(taxa_corrected2),taxa_are_rows = TRUE), tax_table(p.comb.filt.s.1500.80.true), sample_data(p.comb.filt.s.1500.80.true))

p.comb.filt.s.1500.80.postConQuR2 <- prune_taxa(taxa = (taxa_names(p.comb.filt.s.1500.80.postConQuR2)[-which(taxa_sums(p.comb.filt.s.1500.80.postConQuR2) == 0)]),x = p.comb.filt.s.1500.80.postConQuR2)


p.comb.filt.s.1500.80.postConQuR <- merge_phyloseq(otu_table(t(taxa_corrected),taxa_are_rows = TRUE), tax_table(p.comb.filt.s.1500.80.true), sample_data(p.comb.filt.s.1500.80.true))

p.comb.filt.s.1500.80.postConQuR <- prune_taxa(taxa = (taxa_names(p.comb.filt.s.1500.80.postConQuR)[-which(taxa_sums(p.comb.filt.s.1500.80.postConQuR) == 0)]),x = p.comb.filt.s.1500.80.postConQuR)
```

### Adonis2 / PERMANOVA
```{r}
#calculate bray_dist of our new phyloseq
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.postConQuR2, method="bray")
#omnibus testing for DNA extraction batch, PCR batch, sample_side
adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR2)$DNAex_round)
adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR2)$PCR_round)
adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR2)$sample_side)
#pairwise testing inside DNA extraction batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.postConQuR2)$DNAex_round)
p.adonis.pairwise
#pairwise testing inside PCR batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.postConQuR2)$PCR_round)
p.adonis.pairwise
#pairwise testing of "sample_side" {Redundant as sample_side consists of "normal" & "tumor"}
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.postConQuR2)$sample_side)
p.adonis.pairwise
```


```{r}
#beta diversity of new data - colored by sample side
bc_ordination = ordinate(p.comb.filt.s.1500.80.postConQuR2, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR2)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.postConQuR2, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - ConQuR2 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)
#beta diversity of new data - colored by DNA extraction batch
bc_ordination = ordinate(p.comb.filt.s.1500.80.postConQuR2, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR2)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.postConQuR2, bc_ordination, color="DNAex_round", shape = "sample_side") +
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 4_5","Batch 5","Batch 6","Batch 7","Batch NA"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - ConQuR2 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)
#beta diversity of new data - colored by PCR batch
bc_ordination = ordinate(p.comb.filt.s.1500.80.postConQuR2, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.postConQuR2)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.postConQuR2, bc_ordination, color="PCR_round", shape = "sample_side") +
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 5","Batch 6","Batch 7","Batch 8","Batch 9"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - ConQuR2 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)


#covar = df.mod[,c("sample_sidev2","sample_nrv2","PCR_inputv2","ng_in_fcv2","PCR.eluate.ng.µLv2","µL.in.libraryv2")]
#covar = df.mod[,c("sample_sidev2","sample_nrv2")]
#define covariates for ConQuR - here sample_side. NOTE: as seen above I have also tried other combinations of covariates (ConQuR´s GitHub uses "sex","age","race") - if we however use these - nothing changes
```
# Decontam

Even in the low-biomass regime, it is still expected that non-contaminants will appear in a larger fraction of true samples than in negative control samples.
Rows should correspond to samples, and columns to sequences (or OTUs). If a phyloseq object is provided, the otu-table component will be extracted.
```{r}
p.comb.filt.s.1500.80.r <- transform(p.comb.filt.s.1500.80, "compositional")
threshold <- 0.4
#set levels
df <- meta(p.comb.filt.s.1500.80.r)

contamdf.prev <- isContaminant(p.comb.filt.s.1500.80.r, method="prevalence", neg="is.neg", threshold = threshold, normalize = FALSE,batch = df$PCR_round)
# null-hypothesis: not a contaminant , alternate hypothesis: contaminant
hist(contamdf.prev$p, 100, ylim = c(0,300), xlim = c(0,1))

which(contamdf.prev$contaminant)
length(which(contamdf.prev$contaminant))

noncontamdf.prev <- isNotContaminant(p.comb.filt.s.1500.80.r, method = "prevalence", neg="is.neg",threshold=0.5,detailed = TRUE)
hist(noncontamdf.prev$p, 100, ylim = c(0,300), xlim = c(0,1))

p.comb.filt.s.1500.80.r <- prune_taxa(taxa = taxa_names(p.comb.filt.s.1500.80.r)[!(contamdf.prev$contaminant)],x = p.comb.filt.s.1500.80.r)

# no raw data

p.comb.try.save <- merge_phyloseq(p.comb.filt.s.1500.80.nct,p.comb.filt.s.1500.80.postConQuR) # merge the phyloseq objects containing true samples (p.true.filt.s.1500.80) and nct samples (physeq.filt.nct.s.1500.80) #ConQuR based on DNAex_round = 1 & PCR_round = 1
p.comb.try <- merge_phyloseq(p.comb.filt.s.1500.80.nct,p.comb.filt.s.1500.80.postConQuR2)
p.comb.s.try <- tax_glom(p.comb.try,taxrank = "species", NArm = TRUE)

p.comb.s.try.r <- transform(p.comb.s.try, "compositional")
noncontamdf.prev <- isNotContaminant(p.comb.s.try.r, method = "prevalence", neg="is.neg",threshold=0.5,detailed = TRUE) # null-hypothesis: is contaminant / below threshold we reject null-hypothesis
hist(noncontamdf.prev$p, 100, ylim = c(0,300), xlim = c(0,1))
length(which(noncontamdf.prev$not.contaminant))

p.comb.filt.s.1500.80.r <- prune_taxa(taxa = taxa_names(p.comb.s.try)[(noncontamdf.prev$not.contaminant)],x = p.comb.s.try)
p.comb.filt.s.1500.80.r <- subset_samples(p.comb.filt.s.1500.80.r,TRUE_control=="TRUE")
df <- meta(p.comb.filt.s.1500.80.r)
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80.r)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
PCR_batch1_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "1") %>% count("PCR_round") %>% pull()
PCR_batch2_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "2") %>% count("PCR_round") %>% pull()
PCR_batch3_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "3") %>% count("PCR_round") %>% pull()
PCR_batch4_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "4") %>% count("PCR_round") %>% pull()
PCR_batch5_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "5") %>% count("PCR_round") %>% pull()
PCR_batch6_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "6") %>% count("PCR_round") %>% pull()
PCR_batch7_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "7") %>% count("PCR_round") %>% pull()
PCR_batch8_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "8") %>% count("PCR_round") %>% pull()
PCR_batch9_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "9") %>% count("PCR_round") %>% pull()
#plot librarysize colored by PCR_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(PCR_round))) + geom_point()  + theme_bw() + ggtitle("Library size - p.comb.filt.s.1500.80.r")+scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c(paste("Batch 1\nn =", PCR_batch1_count, sep = " "),paste("Batch 2\nn =", PCR_batch2_count, sep = " "),paste("Batch 3\nn =", PCR_batch3_count, sep = " "),paste("Batch 4\nn =", PCR_batch4_count, sep = " "),paste("Batch 5\nn =", PCR_batch5_count, sep = " "),paste("Batch 6\nn =", PCR_batch6_count, sep = " "),paste("Batch 7\nn =", PCR_batch7_count, sep = " "),paste("Batch 8\nn =", PCR_batch8_count, sep = " "),paste("Batch 9\nn =", PCR_batch9_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))

#prepare dataframe for plotting
df <- meta(p.comb.filt.s.1500.80.r) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80.r)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))

#set levels
df$sample_side <- factor(df$sample_side)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$sample_side), 2)),seq(nrow(t(combn(levels(df$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df, aes(x = sample_side, y = LibrarySize)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by sample side") +
  theme(text = element_text(size = 10))

#set levels
df$DNAex_round <- factor(df$DNAex_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$DNAex_round), 2)),seq(nrow(t(combn(levels(df$DNAex_round), 2)))))
#boxplot comparing librarysizes of different DNAex batches
ggplot(df, aes(x = DNAex_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "DNA extraction round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.02,comparisons = comb, map_signif_level = FALSE,y_position = seq.int(55000,200000,8000),color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by DNA extraction batch") +
    theme(text = element_text(size = 10))

#set levels
df$PCR_round <- factor(df$PCR_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$PCR_round), 2)),seq(nrow(t(combn(levels(df$PCR_round), 2)))))
#boxplot comparing librarysizes of different PCR batches
ggplot(df, aes(x = PCR_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "PCR round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.1,comparisons = comb, map_signif_level = FALSE,color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by PCR extraction batch") +
    theme(text = element_text(size = 10))

bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.r, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.r, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.r)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.r, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.r, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.r, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.r)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.r, bc_ordination, color="DNAex_round") + 
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.r, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.r, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.r)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.r, bc_ordination, color="PCR_round") + 
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

aindex <- estimate_richness(physeq = otu_table(p.comb.filt.s.1500.80.r), measures = c("Observed","Shannon","Simpson","InvSimpson", "Chao1","ACE"))

aindex$sample_side <- meta(p.comb.filt.s.1500.80.r)$sample_side
aindex$sample_side <- factor(aindex$sample_side)

comb <- split(t(combn(levels(aindex$sample_side), 2)),seq(nrow(t(combn(levels(aindex$sample_side), 2)))))
data <- reshape2::melt(aindex)
ggplot(data, aes(x = sample_side,y = value)) +
  # Outliers are removed, because otherwise each data point would be plotted twice; 
  # as an outlier of boxplot and as a point of dotplot.
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + 
  ggtitle("ContaminantRemover approach - alpha diversity") +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE,) +
  theme(text = element_text(size = 10)) +
  facet_wrap(~ variable, scales = "free_y")

physeq.fam <- transform_sample_counts(p.comb.filt.s.1500.80.r, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.03,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))


p.comb <- prune_taxa(taxa = taxa_names(p.comb.s.try)[(noncontamdf.prev$not.contaminant)],x = p.comb.s.try)
p.comb.new.true <- subset_samples(p.comb, TRUE_control == "TRUE") # only true samples
p.comb.new.nct <- subset_samples(p.comb, TRUE_control == "control") # only negative samples
  #plot how many taxa are present in atleast x% of negative controls
prev_ntc = tibble(
  prev.ntc = prevalence(p.comb.new.nct, detection  = 0, sort = TRUE, count = FALSE),
  OTU =as.factor(names(prevalence(p.comb.new.nct , detection = 0, sort = TRUE, count = FALSE))) 
)

prev_true = tibble(
  prev.true = prevalence(p.comb.new.true, detection  = 0, sort = TRUE, count = FALSE),
  OTU =as.factor(names(prevalence(p.comb.new.true , detection = 0, sort = TRUE, count = FALSE))) 
)

prev <- prev_true %>% 
  left_join(prev_ntc) %>% 
  mutate(prev.ntc = ifelse(is.na(prev.ntc), 0, prev.ntc),
         prev.true = ifelse(is.na(prev.true), 0, prev.true))

prev_count.mod <- prev %>% # rebuild "prev_count" object 
  count("prev.ntc") %>% # rebuild "prev_count" object 
  mutate(prop = freq/nrow(prev)) %>% # rebuild "prev_count" object 
  arrange(-prev.ntc) %>% #orders the tibble starting with highest prev.ntc value and ending with lowest prev.ntc
  mutate(csum = cumsum(freq)) %>% #create column containing the cumulative sum of n
  mutate(cumprop = csum/nrow(prev)) # take the cumsum and divide by the amount of taxa
# Idea behind this:
# In previous plots we used to plot the exact number / proportion of bacteria contained in x% of negative controls yielding nejman.figure.jpg / Graph 1 in notepad.
# Now we plot the proportion of bacteria contained in at least x% of negative controls.
# x = % of negative controls | y = (sum of all bacteria present in more than x negative controls + bacteria present in x negative controls) / (all bacteria)


ggplot(prev_count.mod, aes(x= prev.ntc, y = cumprop))+
  geom_point()+
  xlab ("% negative controls") +
  ylab("Proportion bacteria") +
  theme_bw() +
  ggtitle("% of bacteria in at least x ntc - % of negative controls included")+
  theme( axis.text.x = element_text(size=12),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

filter <- 0.27 #CP setting threshold to 27% -> MAYBE USE 23%
filtered_taxa1 <- tibble(
  Cont = genefilter_sample(p.comb.new.nct, filterfun_sample(function(x) x > 0), A=filter*nsamples(p.comb.new.nct)),
  OTU = names(genefilter_sample(p.comb.new.nct, filterfun_sample(function(x) x > 0), A=filter*nsamples(p.comb.new.nct)))
) # this keeps ordering of taxa in line with taxonomyTable
head(!(filtered_taxa1$Cont))
length(which(filtered_taxa1$Cont))
p.comb <- prune_taxa(taxa = c(!(filtered_taxa1$Cont)),x = p.comb) #remove highly prevelent bacteria
p.comb.true <- subset_samples(p.comb, TRUE_control == "TRUE")


IDs <- unique(meta(p.comb)$DNAex_round)
df.condition1.meta <- as.data.frame(matrix(0,nrow = 8,ncol = length(IDs)),row.names = c("n_true_samples","n_nct_samples","n_paraffin_tumor_nct_samples","n_paraffin_normal_nct_samples","n_paraffin_nct_samples","n_buffer_nct_samples","n_pcr_nct_samples","batch_name"))
vec0 <- rep(0,length(IDs))
for (i in seq(1,length(IDs))) {
  vec0[i] <- paste0("DNAex","b",i)
}
colnames(df.condition1.meta) <- vec0
for (i in seq(1,length(IDs))) {
  b <- subset_samples(p.comb, DNAex_round == IDs[i])
  b <- prune_taxa(taxa = (taxa_names(b)[-which(taxa_sums(b) == 0)]),x = b) #this removes taxa with taxa_sums == 0 from the overall batch
  check1 <- length(which(meta(b)$TRUE_control == "TRUE")) #checks whether the batch contains true_samples
  check2 <- length(which(meta(b)$TRUE_control == "control")) #checks whether the batch contains nct_sample
  check3 <- length(which(meta(b)$sample_side == "buffer")) #checks whether the batch contains buffer_nct_samples
  check4 <- length(which(meta(b)$sample_side == "PCR_ctrl_H2O")) #checks whether the batch contains pcrH2O_nct_samples
  check5 <- length(which(meta(b)$sample_side == "paraffin_normal")) #checks whether the batch contains paraffinnormal_nct_samples
  check6 <- length(which(meta(b)$sample_side == "paraffin_tumor")) #checks whether the batch contains paraffintumor_nct_samples
  df.condition1.meta[1,i] <- check1
  df.condition1.meta[2,i] <- check2
  df.condition1.meta[3,i] <- check6
  df.condition1.meta[4,i] <- check5
  df.condition1.meta[5,i] <- (check5+check6)
  df.condition1.meta[6,i] <- check3
  df.condition1.meta[7,i] <- check4
  df.condition1.meta[8,i] <- IDs[i]
  if (check1 == 0) {
    print(paste0("check1 :",IDs[i]))
    next}
  if (check2 == 0) {
    print(paste0("check2 :",IDs[i]))
    next}
  if (check3 != 0) {
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "buffer") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "buffer") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt1-buffer-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p1-buffer-",IDs[i])
    assign(x,df.p)
  }
  if (check4 != 0) {
    #for DNAex this is mostly irrelevant as PCR_ctrl_H2O has DNAex_round: NA
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "PCR_ctrl_H2O") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "PCR_ctrl_H2O") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt1-PCR_ctrl_H2O-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p1-PCR_ctrl_H2O-",IDs[i])
    assign(x,df.p)
  }
  n_paraffin <- (check5+check6)
  if (n_paraffin != 0) {
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "paraffin_tumor" |  sample_side == "paraffin_normal") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "paraffin_tumor" |  sample_side == "paraffin_normal") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt1-paraffin-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p1-paraffin-",IDs[i])
    assign(x,df.p)
  }
}

IDs <- unique(meta(p.comb)$PCR_round)
df.condition2.meta <- as.data.frame(matrix(0,nrow = 8,ncol = length(IDs)),row.names = c("n_true_samples","n_nct_samples","n_paraffin_tumor_nct_samples","n_paraffin_normal_nct_samples","n_paraffin_nct_samples","n_buffer_nct_samples","n_pcr_nct_samples","batch_name"))
vec0 <- rep(0,length(IDs))
for (i in seq(1,length(IDs))) {
  vec0[i] <- paste0("PCR","b",i)
}
colnames(df.condition2.meta) <- vec0
for (i in seq(1,length(IDs))) {
  b <- subset_samples(p.comb, PCR_round == IDs[i])
  b <- prune_taxa(taxa = (taxa_names(b)[-which(taxa_sums(b) == 0)]),x = b) #this removes taxa with taxa_sums == 0 from the overall batch
  check1 <- length(which(meta(b)$TRUE_control == "TRUE")) #checks whether the batch contains true_samples
  check2 <- length(which(meta(b)$TRUE_control == "control")) #checks whether the batch contains nct_sample
  check3 <- length(which(meta(b)$sample_side == "buffer")) #checks whether the batch contains buffer_nct_samples
  check4 <- length(which(meta(b)$sample_side == "PCR_ctrl_H2O")) #checks whether the batch contains pcrH2O_nct_samples
  check5 <- length(which(meta(b)$sample_side == "paraffin_normal")) #checks whether the batch contains paraffinnormal_nct_samples
  check6 <- length(which(meta(b)$sample_side == "paraffin_tumor")) #checks whether the batch contains paraffintumor_nct_samples
  df.condition2.meta[1,i] <- check1
  df.condition2.meta[2,i] <- check2
  df.condition2.meta[3,i] <- check6
  df.condition2.meta[4,i] <- check5
  df.condition2.meta[5,i] <- (check5+check6)
  df.condition2.meta[6,i] <- check3
  df.condition2.meta[7,i] <- check4
  df.condition2.meta[8,i] <- IDs[i]
  if (check1 == 0) {
    print(paste0("check1 :",IDs[i]))
    next}
  if (check2 == 0) {
    print(paste0("check2 :",IDs[i]))
    next}
  if (check3 != 0) {
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "buffer") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "buffer") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt2-buffer-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p2-buffer-",IDs[i])
    assign(x,df.p)
  }
  if (check4 != 0) {
    #for DNAex this is mostly irrelevant as PCR_ctrl_H2O has DNAex_round: NA
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "PCR_ctrl_H2O") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "PCR_ctrl_H2O") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt2-PCR_ctrl_H2O-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p2-PCR_ctrl_H2O-",IDs[i])
    assign(x,df.p)
  }
  n_paraffin <- (check5+check6)
  if (n_paraffin != 0) {
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "paraffin_tumor" |  sample_side == "paraffin_normal") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "paraffin_tumor" |  sample_side == "paraffin_normal") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt2-paraffin-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p2-paraffin-",IDs[i])
    assign(x,df.p)
  }
}

bad_taxa <- list()
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-1`[which(x = `df.p1-buffer-1`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-2`[which(x = `df.p1-buffer-2`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-3`[which(x = `df.p1-buffer-3`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-4`[which(x = `df.p1-buffer-4`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-5`[which(x = `df.p1-buffer-5`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-6`[which(x = `df.p1-buffer-6`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-1`[which(x = `df.p1-paraffin-1`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-2`[which(x = `df.p1-paraffin-2`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-3`[which(x = `df.p1-paraffin-3`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-4`[which(x = `df.p1-paraffin-4`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-5`[which(x = `df.p1-paraffin-5`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-6`[which(x = `df.p1-paraffin-6`$pvalue >= 0.05),1])

bad_taxa <- list.append(bad_taxa,`df.p2-buffer-8`[which(x = `df.p2-buffer-8`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p2-PCR_ctrl_H2O-1`[which(x = `df.p2-PCR_ctrl_H2O-1`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p2-PCR_ctrl_H2O-2`[which(x = `df.p2-PCR_ctrl_H2O-2`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p2-paraffin-8`[which(x = `df.p2-paraffin-8`$pvalue >= 0.05),1])

a <- union(bad_taxa[[1]],bad_taxa[[2]])
b <- union(bad_taxa[[3]],bad_taxa[[4]])
c <- union(bad_taxa[[5]],bad_taxa[[6]])
d <- union(bad_taxa[[7]],bad_taxa[[8]])
e <- union(bad_taxa[[9]],bad_taxa[[10]])
f <- union(bad_taxa[[11]],bad_taxa[[12]])
g <- union(bad_taxa[[13]],bad_taxa[[14]])
h <- union(bad_taxa[[15]],bad_taxa[[16]])

a <- union(a,b)
c <- union(c,d)
e <- union(e,f)
g <- union(g,h)

a <- union(a,c)
e <- union(e,g)

bad_taxa <- union(a,e)

p.newfilter2 <- subset_taxa(p.comb, !(taxa_names(p.comb) %in% bad_taxa))
p.newfilter2.true <- subset_samples(p.newfilter2,TRUE_control == "TRUE")

physeq.fam <- transform_sample_counts(p.newfilter2.true, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.04,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

p.test <- merge_phyloseq(p.newfilter2.true,p.comb.new.nct)
p.test <- tax_glom(physeq = p.test,taxrank = "species")
noncontamdf.prev <- isNotContaminant(p.test, method = "prevalence", neg="is.neg",threshold=0.2,detailed = TRUE) # null-hypothesis: is contaminant / below threshold we reject null-hypothesis
hist(noncontamdf.prev$p, 100, ylim = c(0,300), xlim = c(0,1))
noncontamdf.prev$OTU <- rownames(noncontamdf.prev)

p.fin <- prune_taxa(taxa = taxa_names(p.test)[(noncontamdf.prev$not.contaminant)],x = p.test)
p.fin <- subset_samples(p.fin,TRUE_control=="TRUE")
df <- meta(p.fin)
df$LibrarySize <- sample_sums(p.fin)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
PCR_batch1_count <- meta(p.fin) %>% filter(PCR_round == "1") %>% count("PCR_round") %>% pull()
PCR_batch2_count <- meta(p.fin) %>% filter(PCR_round == "2") %>% count("PCR_round") %>% pull()
PCR_batch3_count <- meta(p.fin) %>% filter(PCR_round == "3") %>% count("PCR_round") %>% pull()
PCR_batch4_count <- meta(p.fin) %>% filter(PCR_round == "4") %>% count("PCR_round") %>% pull()
PCR_batch5_count <- meta(p.fin) %>% filter(PCR_round == "5") %>% count("PCR_round") %>% pull()
PCR_batch6_count <- meta(p.fin) %>% filter(PCR_round == "6") %>% count("PCR_round") %>% pull()
PCR_batch7_count <- meta(p.fin) %>% filter(PCR_round == "7") %>% count("PCR_round") %>% pull()
PCR_batch8_count <- meta(p.fin) %>% filter(PCR_round == "8") %>% count("PCR_round") %>% pull()
PCR_batch9_count <- meta(p.fin) %>% filter(PCR_round == "9") %>% count("PCR_round") %>% pull()
#plot librarysize colored by PCR_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(PCR_round))) + geom_point()  + theme_bw() + ggtitle("Library size - p.fin")+scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c(paste("Batch 1\nn =", PCR_batch1_count, sep = " "),paste("Batch 2\nn =", PCR_batch2_count, sep = " "),paste("Batch 3\nn =", PCR_batch3_count, sep = " "),paste("Batch 4\nn =", PCR_batch4_count, sep = " "),paste("Batch 5\nn =", PCR_batch5_count, sep = " "),paste("Batch 6\nn =", PCR_batch6_count, sep = " "),paste("Batch 7\nn =", PCR_batch7_count, sep = " "),paste("Batch 8\nn =", PCR_batch8_count, sep = " "),paste("Batch 9\nn =", PCR_batch9_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare dataframe for plotting
df <- meta(p.fin) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.fin)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))

#set levels
df$sample_side <- factor(df$sample_side)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$sample_side), 2)),seq(nrow(t(combn(levels(df$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df, aes(x = sample_side, y = LibrarySize)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by sample side") +
  theme(text = element_text(size = 10))

#set levels
df$DNAex_round <- factor(df$DNAex_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$DNAex_round), 2)),seq(nrow(t(combn(levels(df$DNAex_round), 2)))))
#boxplot comparing librarysizes of different DNAex batches
ggplot(df, aes(x = DNAex_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "DNA extraction round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.02,comparisons = comb, map_signif_level = FALSE,y_position = seq.int(55000,200000,8000),color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by DNA extraction batch") +
    theme(text = element_text(size = 10))

#set levels
df$PCR_round <- factor(df$PCR_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$PCR_round), 2)),seq(nrow(t(combn(levels(df$PCR_round), 2)))))
#boxplot comparing librarysizes of different PCR batches
ggplot(df, aes(x = PCR_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "PCR round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.1,comparisons = comb, map_signif_level = FALSE,color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by PCR extraction batch") +
    theme(text = element_text(size = 10))

bray_dist = phyloseq::distance(p.fin, method="bray")
bc_ordination = ordinate(p.fin, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.fin)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.fin, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

bray_dist = phyloseq::distance(p.fin, method="bray")
bc_ordination = ordinate(p.fin, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.fin)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.fin, bc_ordination, color="DNAex_round") + 
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

bray_dist = phyloseq::distance(p.fin, method="bray")
bc_ordination = ordinate(p.fin, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.fin)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.fin, bc_ordination, color="PCR_round") + 
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

aindex <- estimate_richness(physeq = otu_table(p.fin), measures = c("Observed","Shannon","Simpson","InvSimpson", "Chao1","ACE"))

aindex$sample_side <- meta(p.fin)$sample_side
aindex$sample_side <- factor(aindex$sample_side)

comb <- split(t(combn(levels(aindex$sample_side), 2)),seq(nrow(t(combn(levels(aindex$sample_side), 2)))))
data <- reshape2::melt(aindex)
ggplot(data, aes(x = sample_side,y = value)) +
  # Outliers are removed, because otherwise each data point would be plotted twice; 
  # as an outlier of boxplot and as a point of dotplot.
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + 
  ggtitle("ContaminantRemover approach - alpha diversity") +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE,) +
  theme(text = element_text(size = 10)) +
  facet_wrap(~ variable, scales = "free_y")

physeq.fam <- transform_sample_counts(p.fin, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.02,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~sample_side, scales = "free", labeller = as_labeller(header2))+
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```
# Decontam 2
```{r}
#same as above but new version of ConQuR
p.comb.try.save <- merge_phyloseq(p.comb.filt.s.1500.80.nct,p.comb.filt.s.1500.80.postConQuR)
p.comb.s.try <- tax_glom(p.comb.try.save,taxrank = "species", NArm = TRUE)

p.comb.s.try.r <- transform(p.comb.s.try, "compositional")
noncontamdf.prev <- isNotContaminant(p.comb.s.try.r, method = "prevalence", neg="is.neg",threshold=0.5,detailed = TRUE) # null-hypothesis: is contaminant / below threshold we reject null-hypothesis
hist(noncontamdf.prev$p, 100, ylim = c(0,300), xlim = c(0,1))
length(which(noncontamdf.prev$not.contaminant))


p.comb.filt.s.1500.80.r <- prune_taxa(taxa = taxa_names(p.comb.s.try)[(noncontamdf.prev$not.contaminant)],x = p.comb.s.try)
p.comb.filt.s.1500.80.r <- subset_samples(p.comb.filt.s.1500.80.r,TRUE_control=="TRUE")
df <- meta(p.comb.filt.s.1500.80.r)
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80.r)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
PCR_batch1_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "1") %>% count("PCR_round") %>% pull()
PCR_batch2_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "2") %>% count("PCR_round") %>% pull()
PCR_batch3_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "3") %>% count("PCR_round") %>% pull()
PCR_batch4_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "4") %>% count("PCR_round") %>% pull()
PCR_batch5_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "5") %>% count("PCR_round") %>% pull()
PCR_batch6_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "6") %>% count("PCR_round") %>% pull()
PCR_batch7_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "7") %>% count("PCR_round") %>% pull()
PCR_batch8_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "8") %>% count("PCR_round") %>% pull()
PCR_batch9_count <- meta(p.comb.filt.s.1500.80.r) %>% filter(PCR_round == "9") %>% count("PCR_round") %>% pull()
#plot librarysize colored by PCR_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(PCR_round))) + geom_point()  + theme_bw() + ggtitle("Library size - p.comb.filt.s.1500.80.r")+scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c(paste("Batch 1\nn =", PCR_batch1_count, sep = " "),paste("Batch 2\nn =", PCR_batch2_count, sep = " "),paste("Batch 3\nn =", PCR_batch3_count, sep = " "),paste("Batch 4\nn =", PCR_batch4_count, sep = " "),paste("Batch 5\nn =", PCR_batch5_count, sep = " "),paste("Batch 6\nn =", PCR_batch6_count, sep = " "),paste("Batch 7\nn =", PCR_batch7_count, sep = " "),paste("Batch 8\nn =", PCR_batch8_count, sep = " "),paste("Batch 9\nn =", PCR_batch9_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))

#prepare dataframe for plotting
df <- meta(p.comb.filt.s.1500.80.r) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80.r)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))

#set levels
df$sample_side <- factor(df$sample_side)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$sample_side), 2)),seq(nrow(t(combn(levels(df$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df, aes(x = sample_side, y = LibrarySize)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by sample side") +
  theme(text = element_text(size = 10))

#set levels
df$DNAex_round <- factor(df$DNAex_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$DNAex_round), 2)),seq(nrow(t(combn(levels(df$DNAex_round), 2)))))
#boxplot comparing librarysizes of different DNAex batches
ggplot(df, aes(x = DNAex_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "DNA extraction round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.02,comparisons = comb, map_signif_level = FALSE,y_position = seq.int(55000,200000,8000),color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by DNA extraction batch") +
    theme(text = element_text(size = 10))

#set levels
df$PCR_round <- factor(df$PCR_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$PCR_round), 2)),seq(nrow(t(combn(levels(df$PCR_round), 2)))))
#boxplot comparing librarysizes of different PCR batches
ggplot(df, aes(x = PCR_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "PCR round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.1,comparisons = comb, map_signif_level = FALSE,color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by PCR extraction batch") +
    theme(text = element_text(size = 10))

bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.r, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.r, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.r)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.r, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.r, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.r, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.r)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.r, bc_ordination, color="DNAex_round") + 
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.r, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.r, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.r)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.r, bc_ordination, color="PCR_round") + 
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

aindex <- estimate_richness(physeq = otu_table(p.comb.filt.s.1500.80.r), measures = c("Observed","Shannon","Simpson","InvSimpson", "Chao1","ACE"))

aindex$sample_side <- meta(p.comb.filt.s.1500.80.r)$sample_side
aindex$sample_side <- factor(aindex$sample_side)

comb <- split(t(combn(levels(aindex$sample_side), 2)),seq(nrow(t(combn(levels(aindex$sample_side), 2)))))
data <- reshape2::melt(aindex)
ggplot(data, aes(x = sample_side,y = value)) +
  # Outliers are removed, because otherwise each data point would be plotted twice; 
  # as an outlier of boxplot and as a point of dotplot.
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + 
  ggtitle("ContaminantRemover approach - alpha diversity") +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE,) +
  theme(text = element_text(size = 10)) +
  facet_wrap(~ variable, scales = "free_y")

physeq.fam <- transform_sample_counts(p.comb.filt.s.1500.80.r, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.04,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

p.comb <- prune_taxa(taxa = taxa_names(p.comb.s.try)[(noncontamdf.prev$not.contaminant)],x = p.comb.s.try)
p.comb.new.true <- subset_samples(p.comb, TRUE_control == "TRUE") # only true samples
p.comb.new.nct <- subset_samples(p.comb, TRUE_control == "control") # only negative samples
  #plot how many taxa are present in atleast x% of negative controls
prev_ntc = tibble(
  prev.ntc = prevalence(p.comb.new.nct, detection  = 0, sort = TRUE, count = FALSE),
  OTU =as.factor(names(prevalence(p.comb.new.nct , detection = 0, sort = TRUE, count = FALSE))) 
)

prev_true = tibble(
  prev.true = prevalence(p.comb.new.true, detection  = 0, sort = TRUE, count = FALSE),
  OTU =as.factor(names(prevalence(p.comb.new.true , detection = 0, sort = TRUE, count = FALSE))) 
)

prev <- prev_true %>% 
  left_join(prev_ntc) %>% 
  mutate(prev.ntc = ifelse(is.na(prev.ntc), 0, prev.ntc),
         prev.true = ifelse(is.na(prev.true), 0, prev.true))

prev_count.mod <- prev %>% # rebuild "prev_count" object 
  count("prev.ntc") %>% # rebuild "prev_count" object 
  mutate(prop = freq/nrow(prev)) %>% # rebuild "prev_count" object 
  arrange(-prev.ntc) %>% #orders the tibble starting with highest prev.ntc value and ending with lowest prev.ntc
  mutate(csum = cumsum(freq)) %>% #create column containing the cumulative sum of n
  mutate(cumprop = csum/nrow(prev)) # take the cumsum and divide by the amount of taxa
# Idea behind this:
# In previous plots we used to plot the exact number / proportion of bacteria contained in x% of negative controls yielding nejman.figure.jpg / Graph 1 in notepad.
# Now we plot the proportion of bacteria contained in at least x% of negative controls.
# x = % of negative controls | y = (sum of all bacteria present in more than x negative controls + bacteria present in x negative controls) / (all bacteria)


ggplot(prev_count.mod, aes(x= prev.ntc, y = cumprop))+
  geom_point()+
  xlab ("% negative controls") +
  ylab("Proportion bacteria") +
  theme_bw() +
  ggtitle("% of bacteria in at least x ntc - % of negative controls included")+
  theme( axis.text.x = element_text(size=12),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

filter <- 0.27 #CP setting threshold to 27% -> MAYBE USE 23%
filtered_taxa1 <- tibble(
  Cont = genefilter_sample(p.comb.new.nct, filterfun_sample(function(x) x > 0), A=filter*nsamples(p.comb.new.nct)),
  OTU = names(genefilter_sample(p.comb.new.nct, filterfun_sample(function(x) x > 0), A=filter*nsamples(p.comb.new.nct)))
) # this keeps ordering of taxa in line with taxonomyTable
head(!(filtered_taxa1$Cont))
length(which(filtered_taxa1$Cont))
p.comb <- prune_taxa(taxa = c(!(filtered_taxa1$Cont)),x = p.comb) #remove highly prevelent bacteria
p.comb.true <- subset_samples(p.comb, TRUE_control == "TRUE")

IDs <- unique(meta(p.comb)$DNAex_round)
df.condition1.meta <- as.data.frame(matrix(0,nrow = 8,ncol = length(IDs)),row.names = c("n_true_samples","n_nct_samples","n_paraffin_tumor_nct_samples","n_paraffin_normal_nct_samples","n_paraffin_nct_samples","n_buffer_nct_samples","n_pcr_nct_samples","batch_name"))
vec0 <- rep(0,length(IDs))
for (i in seq(1,length(IDs))) {
  vec0[i] <- paste0("DNAex","b",i)
}
colnames(df.condition1.meta) <- vec0
for (i in seq(1,length(IDs))) {
  b <- subset_samples(p.comb, DNAex_round == IDs[i])
  b <- prune_taxa(taxa = (taxa_names(b)[-which(taxa_sums(b) == 0)]),x = b) #this removes taxa with taxa_sums == 0 from the overall batch
  check1 <- length(which(meta(b)$TRUE_control == "TRUE")) #checks whether the batch contains true_samples
  check2 <- length(which(meta(b)$TRUE_control == "control")) #checks whether the batch contains nct_sample
  check3 <- length(which(meta(b)$sample_side == "buffer")) #checks whether the batch contains buffer_nct_samples
  check4 <- length(which(meta(b)$sample_side == "PCR_ctrl_H2O")) #checks whether the batch contains pcrH2O_nct_samples
  check5 <- length(which(meta(b)$sample_side == "paraffin_normal")) #checks whether the batch contains paraffinnormal_nct_samples
  check6 <- length(which(meta(b)$sample_side == "paraffin_tumor")) #checks whether the batch contains paraffintumor_nct_samples
  df.condition1.meta[1,i] <- check1
  df.condition1.meta[2,i] <- check2
  df.condition1.meta[3,i] <- check6
  df.condition1.meta[4,i] <- check5
  df.condition1.meta[5,i] <- (check5+check6)
  df.condition1.meta[6,i] <- check3
  df.condition1.meta[7,i] <- check4
  df.condition1.meta[8,i] <- IDs[i]
  if (check1 == 0) {
    print(paste0("check1 :",IDs[i]))
    next}
  if (check2 == 0) {
    print(paste0("check2 :",IDs[i]))
    next}
  if (check3 != 0) {
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "buffer") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "buffer") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt1-buffer-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p1-buffer-",IDs[i])
    assign(x,df.p)
  }
  if (check4 != 0) {
    #for DNAex this is mostly irrelevant as PCR_ctrl_H2O has DNAex_round: NA
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "PCR_ctrl_H2O") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "PCR_ctrl_H2O") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt1-PCR_ctrl_H2O-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p1-PCR_ctrl_H2O-",IDs[i])
    assign(x,df.p)
  }
  n_paraffin <- (check5+check6)
  if (n_paraffin != 0) {
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "paraffin_tumor" |  sample_side == "paraffin_normal") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "paraffin_tumor" |  sample_side == "paraffin_normal") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt1-paraffin-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p1-paraffin-",IDs[i])
    assign(x,df.p)
  }
}

IDs <- unique(meta(p.comb)$PCR_round)
df.condition2.meta <- as.data.frame(matrix(0,nrow = 8,ncol = length(IDs)),row.names = c("n_true_samples","n_nct_samples","n_paraffin_tumor_nct_samples","n_paraffin_normal_nct_samples","n_paraffin_nct_samples","n_buffer_nct_samples","n_pcr_nct_samples","batch_name"))
vec0 <- rep(0,length(IDs))
for (i in seq(1,length(IDs))) {
  vec0[i] <- paste0("PCR","b",i)
}
colnames(df.condition2.meta) <- vec0
for (i in seq(1,length(IDs))) {
  b <- subset_samples(p.comb, PCR_round == IDs[i])
  b <- prune_taxa(taxa = (taxa_names(b)[-which(taxa_sums(b) == 0)]),x = b) #this removes taxa with taxa_sums == 0 from the overall batch
  check1 <- length(which(meta(b)$TRUE_control == "TRUE")) #checks whether the batch contains true_samples
  check2 <- length(which(meta(b)$TRUE_control == "control")) #checks whether the batch contains nct_sample
  check3 <- length(which(meta(b)$sample_side == "buffer")) #checks whether the batch contains buffer_nct_samples
  check4 <- length(which(meta(b)$sample_side == "PCR_ctrl_H2O")) #checks whether the batch contains pcrH2O_nct_samples
  check5 <- length(which(meta(b)$sample_side == "paraffin_normal")) #checks whether the batch contains paraffinnormal_nct_samples
  check6 <- length(which(meta(b)$sample_side == "paraffin_tumor")) #checks whether the batch contains paraffintumor_nct_samples
  df.condition2.meta[1,i] <- check1
  df.condition2.meta[2,i] <- check2
  df.condition2.meta[3,i] <- check6
  df.condition2.meta[4,i] <- check5
  df.condition2.meta[5,i] <- (check5+check6)
  df.condition2.meta[6,i] <- check3
  df.condition2.meta[7,i] <- check4
  df.condition2.meta[8,i] <- IDs[i]
  if (check1 == 0) {
    print(paste0("check1 :",IDs[i]))
    next}
  if (check2 == 0) {
    print(paste0("check2 :",IDs[i]))
    next}
  if (check3 != 0) {
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "buffer") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "buffer") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt2-buffer-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p2-buffer-",IDs[i])
    assign(x,df.p)
  }
  if (check4 != 0) {
    #for DNAex this is mostly irrelevant as PCR_ctrl_H2O has DNAex_round: NA
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "PCR_ctrl_H2O") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "PCR_ctrl_H2O") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt2-PCR_ctrl_H2O-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p2-PCR_ctrl_H2O-",IDs[i])
    assign(x,df.p)
  }
  n_paraffin <- (check5+check6)
  if (n_paraffin != 0) {
    #perform test
    b.true <- subset_samples(b,TRUE_control == "TRUE") #subset of true samples in batch
    b.condition <- subset_samples(b,sample_side == "paraffin_tumor" |  sample_side == "paraffin_normal") #subset of batch_nct_samples in batch
    b.sub <- merge_phyloseq(b.true,b.condition) # we take the subsets we want, merge them to allow for thorough removal of bacteria
    if (length(which(taxa_sums(b.sub) == 0)) != 0) {
    b.sub <- prune_taxa(taxa = (taxa_names(b.sub)[-which(taxa_sums(b.sub) == 0)]),x = b.sub) #this removes taxa with taxa_sums == 0 from our batch subset
    }
    b.true <- subset_samples(b.sub,TRUE_control == "TRUE") #create new subset of true samples in batch
    b.condition <- subset_samples(b.sub,sample_side == "paraffin_tumor" |  sample_side == "paraffin_normal") #create new subset of batch_nct_samples in batch  
    # Im honest: this is a bit redundant
    df.filt <- data.frame(x = prevalence(b.true, detection = 0, count = TRUE),
                        n = nsamples(b.true),
                        p = prevalence(b.condition, detection = 0, count = FALSE)) #df stores x,n,p for binom.test
    vec <- seq(1,nrow(df.filt))
    for (j in vec) {
    vec[j] <- binom.test(x = df.filt[j,"x"], n = df.filt[j,"n"], p = df.filt[j,"p"],alternative = "greater")$p.value
    }
    y <- paste0("df.filt2-paraffin-",IDs[i])
    assign(y,df.filt)
    df.p <- data.frame(id = rownames(df.filt),
                       pvalue = vec)
    x <- paste0("df.p2-paraffin-",IDs[i])
    assign(x,df.p)
  }
}

bad_taxa <- list()
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-1`[which(x = `df.p1-buffer-1`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-2`[which(x = `df.p1-buffer-2`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-3`[which(x = `df.p1-buffer-3`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-4`[which(x = `df.p1-buffer-4`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-5`[which(x = `df.p1-buffer-5`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-buffer-6`[which(x = `df.p1-buffer-6`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-1`[which(x = `df.p1-paraffin-1`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-2`[which(x = `df.p1-paraffin-2`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-3`[which(x = `df.p1-paraffin-3`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-4`[which(x = `df.p1-paraffin-4`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-5`[which(x = `df.p1-paraffin-5`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p1-paraffin-6`[which(x = `df.p1-paraffin-6`$pvalue >= 0.05),1])

bad_taxa <- list.append(bad_taxa,`df.p2-buffer-8`[which(x = `df.p2-buffer-8`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p2-PCR_ctrl_H2O-1`[which(x = `df.p2-PCR_ctrl_H2O-1`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p2-PCR_ctrl_H2O-2`[which(x = `df.p2-PCR_ctrl_H2O-2`$pvalue >= 0.05),1])
bad_taxa <- list.append(bad_taxa,`df.p2-paraffin-8`[which(x = `df.p2-paraffin-8`$pvalue >= 0.05),1])

a <- union(bad_taxa[[1]],bad_taxa[[2]])
b <- union(bad_taxa[[3]],bad_taxa[[4]])
c <- union(bad_taxa[[5]],bad_taxa[[6]])
d <- union(bad_taxa[[7]],bad_taxa[[8]])
e <- union(bad_taxa[[9]],bad_taxa[[10]])
f <- union(bad_taxa[[11]],bad_taxa[[12]])
g <- union(bad_taxa[[13]],bad_taxa[[14]])
h <- union(bad_taxa[[15]],bad_taxa[[16]])

a <- union(a,b)
c <- union(c,d)
e <- union(e,f)
g <- union(g,h)

a <- union(a,c)
e <- union(e,g)

bad_taxa <- union(a,e)

p.newfilter2 <- subset_taxa(p.comb, !(taxa_names(p.comb) %in% bad_taxa))
p.newfilter2.true <- subset_samples(p.newfilter2,TRUE_control == "TRUE")

physeq.fam <- transform_sample_counts(p.newfilter2.true, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.04,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

p.test <- merge_phyloseq(p.newfilter2.true,p.comb.new.nct)
p.test <- tax_glom(physeq = p.test,taxrank = "species")
noncontamdf.prev <- isNotContaminant(p.test, method = "prevalence", neg="is.neg",threshold=0.3,detailed = TRUE) # null-hypothesis: is contaminant / below threshold we reject null-hypothesis
hist(noncontamdf.prev$p, 100, ylim = c(0,300), xlim = c(0,1))
noncontamdf.prev$OTU <- rownames(noncontamdf.prev)

p.fin <- prune_taxa(taxa = taxa_names(p.test)[(noncontamdf.prev$not.contaminant)],x = p.test)
p.fin <- subset_samples(p.fin,TRUE_control=="TRUE")
df <- meta(p.fin)
df$LibrarySize <- sample_sums(p.fin)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
PCR_batch1_count <- meta(p.fin) %>% filter(PCR_round == "1") %>% count("PCR_round") %>% pull()
PCR_batch2_count <- meta(p.fin) %>% filter(PCR_round == "2") %>% count("PCR_round") %>% pull()
PCR_batch3_count <- meta(p.fin) %>% filter(PCR_round == "3") %>% count("PCR_round") %>% pull()
PCR_batch4_count <- meta(p.fin) %>% filter(PCR_round == "4") %>% count("PCR_round") %>% pull()
PCR_batch5_count <- meta(p.fin) %>% filter(PCR_round == "5") %>% count("PCR_round") %>% pull()
PCR_batch6_count <- meta(p.fin) %>% filter(PCR_round == "6") %>% count("PCR_round") %>% pull()
PCR_batch7_count <- meta(p.fin) %>% filter(PCR_round == "7") %>% count("PCR_round") %>% pull()
PCR_batch8_count <- meta(p.fin) %>% filter(PCR_round == "8") %>% count("PCR_round") %>% pull()
PCR_batch9_count <- meta(p.fin) %>% filter(PCR_round == "9") %>% count("PCR_round") %>% pull()
#plot librarysize colored by PCR_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(PCR_round))) + geom_point()  + theme_bw() + ggtitle("Library size - p.fin")+scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c(paste("Batch 1\nn =", PCR_batch1_count, sep = " "),paste("Batch 2\nn =", PCR_batch2_count, sep = " "),paste("Batch 3\nn =", PCR_batch3_count, sep = " "),paste("Batch 4\nn =", PCR_batch4_count, sep = " "),paste("Batch 5\nn =", PCR_batch5_count, sep = " "),paste("Batch 6\nn =", PCR_batch6_count, sep = " "),paste("Batch 7\nn =", PCR_batch7_count, sep = " "),paste("Batch 8\nn =", PCR_batch8_count, sep = " "),paste("Batch 9\nn =", PCR_batch9_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare dataframe for plotting
df <- meta(p.fin) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.fin)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))

#set levels
df$sample_side <- factor(df$sample_side)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$sample_side), 2)),seq(nrow(t(combn(levels(df$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df, aes(x = sample_side, y = LibrarySize)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by sample side") +
  theme(text = element_text(size = 10))

#set levels
df$DNAex_round <- factor(df$DNAex_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$DNAex_round), 2)),seq(nrow(t(combn(levels(df$DNAex_round), 2)))))
#boxplot comparing librarysizes of different DNAex batches
ggplot(df, aes(x = DNAex_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "DNA extraction round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.02,comparisons = comb, map_signif_level = FALSE,y_position = seq.int(55000,200000,8000),color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by DNA extraction batch") +
    theme(text = element_text(size = 10))

#set levels
df$PCR_round <- factor(df$PCR_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$PCR_round), 2)),seq(nrow(t(combn(levels(df$PCR_round), 2)))))
#boxplot comparing librarysizes of different PCR batches
ggplot(df, aes(x = PCR_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "PCR round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.1,comparisons = comb, map_signif_level = FALSE,color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by PCR extraction batch") +
    theme(text = element_text(size = 10))

bray_dist = phyloseq::distance(p.fin, method="bray")
bc_ordination = ordinate(p.fin, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.fin)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.fin, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

bray_dist = phyloseq::distance(p.fin, method="bray")
bc_ordination = ordinate(p.fin, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.fin)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.fin, bc_ordination, color="DNAex_round") + 
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

bray_dist = phyloseq::distance(p.fin, method="bray")
bc_ordination = ordinate(p.fin, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.fin)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.fin, bc_ordination, color="PCR_round") + 
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

aindex <- estimate_richness(physeq = otu_table(p.fin), measures = c("Observed","Shannon","Simpson","InvSimpson", "Chao1","ACE"))

aindex$sample_side <- meta(p.fin)$sample_side
aindex$sample_side <- factor(aindex$sample_side)

comb <- split(t(combn(levels(aindex$sample_side), 2)),seq(nrow(t(combn(levels(aindex$sample_side), 2)))))
data <- reshape2::melt(aindex)
ggplot(data, aes(x = sample_side,y = value)) +
  # Outliers are removed, because otherwise each data point would be plotted twice; 
  # as an outlier of boxplot and as a point of dotplot.
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + 
  ggtitle("ContaminantRemover approach - alpha diversity") +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE,) +
  theme(text = element_text(size = 10)) +
  facet_wrap(~ variable, scales = "free_y")

physeq.fam <- transform_sample_counts(p.fin, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.02,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~sample_side, scales = "free", labeller = as_labeller(header2))+
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```
```{r}
otu = as(otu_table(p.fin), "matrix")
if(taxa_are_rows(p.fin)){otu <- t(otu)}
otu = as_tibble(otu)
res_sim <- PERFect_sim(X = otu,alpha = 0.1)
dim(res_sim$filtX)
ids.sim<- colnames(res_sim$filtX)
p.fin2 <-  prune_taxa(ids.sim, p.fin)

ord.pval <- pvals_Order(Counts = otu,res_sim = res_sim )

otu = as(otu_table(p.fin), "matrix")
if(taxa_are_rows(p.fin)){otu <- t(otu)}
otu = as_tibble(otu)
res_perm <- PERFect_perm(X = otu, Order.user = ord.pval,alpha = 0.1, rollmean = FALSE)
dim(res_sim$filtX)
ids.sim<- colnames(res_perm$filtX)
physeq.control <-  prune_taxa(ids.sim, p.comb.filt.s.1500.80.true)
physeq.control

physeq.fam <- transform_sample_counts(physeq.control, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.03,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
header2 <- c("normal" = "Normal pancreas","tumor" = "Tumor tissue") # for plot
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~sample_side, scales = "free", labeller = as_labeller(header2))+
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

aindex <- estimate_richness(physeq = otu_table(physeq.control), measures = c("Observed","Shannon","Simpson","InvSimpson", "Chao1","ACE"))

aindex$sample_side <- meta(physeq.control)$sample_side
aindex$sample_side <- factor(aindex$sample_side)

comb <- split(t(combn(levels(aindex$sample_side), 2)),seq(nrow(t(combn(levels(aindex$sample_side), 2)))))
data <- reshape2::melt(aindex)
ggplot(data, aes(x = sample_side,y = value)) +
  # Outliers are removed, because otherwise each data point would be plotted twice; 
  # as an outlier of boxplot and as a point of dotplot.
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + 
  ggtitle("ContaminantRemover approach - alpha diversity") +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE,) +
  theme(text = element_text(size = 10)) +
  facet_wrap(~ variable, scales = "free_y")


physeq.control2 <- prune_samples(samples = (sample_names(physeq.control)[-which(sample_sums(physeq.control) == 0)]),x = physeq.control)
bray_dist = phyloseq::distance(physeq.control2, method="bray")
bc_ordination = ordinate(physeq.control2, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(physeq.control2)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(physeq.control2, bc_ordination, color="DNAex_round") + 
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

bray_dist = phyloseq::distance(physeq.control2, method="bray")
bc_ordination = ordinate(physeq.control2, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(physeq.control2)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(physeq.control2, bc_ordination, color="PCR_round") + 
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

bray_dist = phyloseq::distance(physeq.control2, method="bray")
bc_ordination = ordinate(physeq.control2, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(physeq.control2)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(physeq.control2, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("ConQuR(2)-decontam50 - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#in this method Permutation PERFect did not help differentiate the data. It actually made the data worse by eliminating taxa that allowed a differentiation.
```


# Junk
```{r Nejman Filter 1}
#Understanding nejman et. al S9B: 
#we take all nct together - sum up the librarysize of the nct samples - samples_sums(1)+samples_sums(2)+samples_sums(3)+samples_sums(4)+.. {for all samples} → now we take all taxa {from true samples + nct samples} and calculate the prevalence of these taxa in our nct samples. - i.e. taxa 1: taxa_sums(taxa1)  / sum of sample_sums | in this case "taxa_sums(taxa1) refers to the sum of all read counts of taxon 1 in all nct samples | we perform this calculation for all taxa - generating a prevalence value {this prevalence refers not to a specific nct sample - rather to the general nct sample collection} | now we create a graph: on the x-axis we have the prevalence value {refering to the general prevalence of a taxon in controls - NOT the prevalence of a taxon inside a nct sample} - on the y-axis the proportion of taxa sharing the same prevalence value
sample_sums(physeq.filt.nct.s.1500.80) #sum of read counts in each nct control sample {a,b,c,d,...}
cumsum(sample_sums(physeq.filt.nct.s.1500.80)) # adding each sample´s sum of read counts together {a,a+b,a+b+c,a+b+c+d,...} 
#we now calculate the taxa_sums part: 
#[REMOVED] # we take the combined phyloseq-object and subset it for TRUE_control == control {that way we retain all 1010 taxa but only calculate the taxa_sums on negative controls}
df <- data.frame(taxa_sum = taxa_sums(p.comb.filt.s.1500.80.nct),row.names = taxa_names(p.comb.filt.s.1500.80.nct))
df <- df %>% mutate(prevalence = taxa_sum/cumsum(sample_sums(physeq.filt.nct.s.1500.80))[75])
#c(0,424,129,204,147,88,15,3)
#plot(x = c(0,10^-6,10^-5,10^-4,10^-3,10^-2,10^-1,1),y = c(0,424,129,204,147,88,15,3)) #We plot the 


p.comb.new.true <- subset_samples(p.comb.filt.s.new, TRUE_control == "TRUE") # only true samples
p.comb.new.nct <- subset_samples(p.comb.filt.s.new, TRUE_control == "control") # only negative samples
  #plot how many taxa are present in atleast x% of negative controls
prev_ntc = tibble(
  prev.ntc = prevalence(p.comb.new.nct, detection  = 0, sort = TRUE, count = FALSE),
  OTU =as.factor(names(prevalence(p.comb.new.nct , detection = 0, sort = TRUE, count = FALSE))) 
)

prev_true = tibble(
  prev.true = prevalence(p.comb.new.true, detection  = 0, sort = TRUE, count = FALSE),
  OTU =as.factor(names(prevalence(p.comb.new.true , detection = 0, sort = TRUE, count = FALSE))) 
)

prev <- prev_true %>% 
  left_join(prev_ntc) %>% 
  mutate(prev.ntc = ifelse(is.na(prev.ntc), 0, prev.ntc),
         prev.true = ifelse(is.na(prev.true), 0, prev.true))

prev_count <- prev %>%
  count(prev.ntc) %>%
  mutate(prop = n/nrow(prev))
```

