---
title: "Analysis of raw data"
author: "Christoph Petrynowski"
date: "2023-01-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

In this experiment, human tumor vs. normal pancreas FFPE tissue samples from the Pathology department are examined. The DNA had been isolated with the Invitrogen FFPE kit (by me). The sample size is n=24 for both tumor and normal pancreas (matched patients). Additionally, n=6 FFPE tumor tissue samples from the establishment experiment (patho_ffpe_self) without matched normal pancreas tissue are implemented (so we have n=30 tumor, n=24 normal pancreas). n=75 NTCs (n=23 paraffin tumor, n=24 paraffin normal, n=10 buffer-only, n=7 PCR CTRLs) plus n=11 NTCs from the establishment experiment (patho_ffpe_self) are implemented. Further, several samples <1000 reads were repeated and implemented as well (merge with their first sequencing).

{22.12.2022 CP} Who actually wrote that? Physeq1-5 alrady contain all 58 true samples - respective of removing 4 samples from phyloseq5

# Software-Library

```{r library, echo=FALSE}
library(PERFect)
library(phyloseq)
library(decontam)
library(tidyverse)
library(pairwiseAdonis)
library(kableExtra)
library(vegan)
library(tidytree)
library(ape)
library(ggrepel)
library(microbiome)
library(ggpubr) # for manual p-value assigning
library(rstatix) # for wilcox test
library(yingtools2)
#if( !require(ggtree) ){
#  install.packages(ggtree)
#}
library(readxl) # load excel data
if( !require(rlist) ){
  install.packages(rlist)
}
library(rlist) # work with lists
if( !require(ggsignif) ){
  install.packages(ggsignif)
}
library(ggsignif) # plot p.values on a ggplot
set.seed(3)
library(plyr) #for ldply - performs function on every element in list & combines the output into a dataframe
library(reshape2) # for melting data
if( !require(doParallel) ){
  install.packages("doParallel")
}
library(doParallel) # ConQuR´s GitHub recomends this step
#devtools::install_github("wdl2459/ConQuR")
library(ConQuR)
if( !require("VennDiagram") ){
  install.packages("VennDiagram")
}
library("VennDiagram")
if( !require("ggvenn") ){
  install.packages("ggvenn")
}
library("ggvenn")
```

# Raw Phyloseq 1500.80s
## Loading
### Loading phyloseq "1500.80s true samples" objects
```{r loading phyloseq "1500.80s true samples" objects}
physeq1.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_05_27_patho_FFPE_tum_panc_1/physeq_original_1500.80.rds")
physeq2.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_05_27_patho_FFPE_tum_panc_2/physeq_original_1500.80.rds")
physeq3.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_05_28_patho_FFPE_tum_panc_3/physeq_original_1500.80.rds")
physeq4.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_05_28_patho_FFPE_tum_panc_4/physeq_original_1500.80.rds")
physeq5.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2021_12_01_patho_FFPE_self_tumor/physeq_original_1500.80.rds")
physeq6.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_08_09_patho_FFPE_tum_panc_rep1/physeq_original_1500.80.rds") #09.08.22: rep1: 10 samples (tumor+normal panc mixed)
physeq7.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_08_09_patho_FFPE_tum_panc_rep2/physeq_original_1500.80.rds") #09.08.22: rep2: 10 samples (tumor+normal panc mixed)
p.true.original.1500.80 <- merge_phyloseq(physeq1.1500.80,physeq2.1500.80,physeq3.1500.80,physeq4.1500.80,physeq5.1500.80)#merge the different phyloseq objects into 1 phyloseq object
# sample_data of new phyloseq-object

p.true.filt.1500.80 <- subset_taxa(p.true.original.1500.80, superkingdom=="Bacteria")
p.true.filt.s.1500.80 <- tax_glom(p.true.filt.1500.80, taxrank = "species", NArm = TRUE)

rm(list = c("physeq1.1500.80", "physeq2.1500.80", "physeq3.1500.80", "physeq4.1500.80", "physeq5.1500.80","physeq6.1500.80","physeq7.1500.80"))
rm(p.true.filt.1500.80)
```

##LibrarySize
### LibrarySize tissue samples
```{r tissue samples´s librarysize of the 1500.80s}
normal_count <- as_tibble(sample_data(p.true.filt.s.1500.80)) %>% filter(sample_side == "normal") %>% count("sample_side") %>% pull(freq)
tumor_count <- as_tibble(sample_data(p.true.filt.s.1500.80)) %>% filter(sample_side == "tumor") %>% count("sample_side") %>% pull(freq)
df <- meta(p.true.filt.s.1500.80) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.true.filt.s.1500.80) # add LibrarySize for each sample
df <- df[order(df$LibrarySize),] # order samples (dataframe) based on LibrarySize in ascending order (this will be our y_value for ggplot)
df$Index <- seq(nrow(df)) # give each row an index (this will be our x_value for ggplot)
ggplot(data=df, aes(x=Index, y=LibrarySize, color=sample_side)) + geom_point()  + theme_bw() + ggtitle("Library size true samples 1500.80")+
  scale_color_brewer(palette = "Set1", labels=c(paste("Normal samples\nn =", normal_count, sep = " "), paste("Tumor samples\nn =", tumor_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12),
        legend.title = element_blank())
```

## Loading
### Loading phyloseq "1500.80s ntc samples" objects
```{r loading phyloseq "1500.80s ntc samples" objects}
#loading 1500.80s ntc samples
physeq.nct1.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_06_02_patho_FFPE_tum_panc_ntc_1/physeq_original_1500.80.rds")
physeq.nct2.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_06_02_patho_FFPE_tum_panc_ntc_2/physeq_original_1500.80.rds")
physeq.nct3.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_06_03_patho_FFPE_tum_panc_ntc_3/physeq_original_1500.80.rds")
physeq.nct4.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_06_03_patho_FFPE_tum_panc_ntc_4/physeq_original_1500.80.rds")
physeq.nct5.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_06_04_patho_FFPE_tum_panc_ntc_5/physeq_original_1500.80.rds")
physeq.nct6.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_06_04_patho_FFPE_tum_panc_ntc_6/physeq_original_1500.80.rds")
physeq.nct7.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_06_10_patho_FFPE_tum_panc_ntc_7/physeq_original_1500.80.rds")
physeq.nct8.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_07_13_patho_FFPE_tum_panc_ntc_8/physeq_original_1500.80.rds")
physeq.nct9.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_07_13_patho_FFPE_tum_panc_ntc_9/physeq_original_1500.80.rds")
physeq.nct10.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2021_12_02_patho_FFPE_self_ntc/physeq_original_1500.80.rds")
physeq.nct11.1500.80 <- readRDS("../raw_data/PhyloSeqObj/2022_01_24_patho_FFPE_self_ntc_2/physeq_original_1500.80.rds")

physeq.nct.original.1500.80 <- merge_phyloseq(physeq.nct1.1500.80, physeq.nct2.1500.80, physeq.nct3.1500.80, physeq.nct4.1500.80, physeq.nct5.1500.80, physeq.nct6.1500.80, physeq.nct7.1500.80, physeq.nct8.1500.80, physeq.nct9.1500.80,physeq.nct10.1500.80, physeq.nct11.1500.80)

physeq.filt.nct.1500.80 <- subset_taxa(physeq.nct.original.1500.80, superkingdom=="Bacteria")
physeq.filt.nct.s.1500.80 <- tax_glom(physeq.filt.nct.1500.80, taxrank = "species", NArm = TRUE)

rm(list = c('physeq.nct1.1500.80', 'physeq.nct2.1500.80', 'physeq.nct3.1500.80', 'physeq.nct4.1500.80', 'physeq.nct5.1500.80',
            'physeq.nct6.1500.80', 'physeq.nct7.1500.80', 'physeq.nct8.1500.80', 'physeq.nct9.1500.80', 'physeq.nct10.1500.80', 'physeq.nct11.1500.80'))
rm(physeq.filt.nct.1500.80)

#replace sample_side "paraffin" with "paraffin_tumor"
sample_data(physeq.filt.nct.s.1500.80)$sample_side <- ifelse(sample_data(physeq.filt.nct.s.1500.80)$sample_side == "paraffin", "paraffin_tumor", sample_data(physeq.filt.nct.s.1500.80)$sample_side)
```

## LibrarySize
### LibrarySize NCT samples
```{r NCT samples´s librarysize of the 1500.80s}
df <- meta(physeq.filt.nct.s.1500.80) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(physeq.filt.nct.s.1500.80) # add LibrarySize for each sample
df <- df[order(df$LibrarySize),] # order samples (dataframe) based on LibrarySize in ascending order (this will be our y_value for ggplot)
df$Index <- seq(nrow(df)) # give each row an index (this will be our x_value for ggplot)
ggplot(data=df, aes(x=Index, y=LibrarySize, color=sample_side)) + geom_point()  + theme_bw() + ggtitle("Library size negative controls 1500.80")+ scale_fill_discrete(name="Sample\nSide")+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12),
        legend.title = element_blank())
```

## Merge TRUE/NCT samples
```{r merge true/nct samples}
p.comb.filt.s.1500.80 <- merge_phyloseq(physeq.filt.nct.s.1500.80,p.true.filt.s.1500.80) # merge the phyloseq objects containing true samples (p.true.filt.s.1500.80) and nct samples (physeq.filt.nct.s.1500.80)
p.comb.filt.s.1500.80 <- tax_glom(p.comb.filt.s.1500.80,taxrank = "species", NArm = TRUE) # agglomerate taxa to species level. This combines different OTUs {hypothecially: 450 / 453} assigned to the same species into just 1 OTU {this progess accounts for } 
```
## Loading
### Loading Batch metadata
```{r introduce Batch metadata}
df <- read_excel("../raw_data/Metadata.xlsx", 
    col_types = c("numeric", "text", "text", 
        "text", "text", "text", "text", "text", 
        "text", "numeric", "text", "numeric", 
        "numeric", "text", "text")) #load metadata excel-sheet containing DNA extraction batch information & PCR batch infromation
df <- as.data.frame(df)

sd.new <- meta(p.comb.filt.s.1500.80) %>% arrange(uniqueID)
df <- df %>% arrange(uniqueID)
sd.new$DNAex_round <- df$DNAex_round
sd.new$PCR_round <- df$PCR_round
rownames(sd.new) <- sd.new$id

# the above code arranges sampledata of merged phyloseq-object by uniqueID & the metadata excel-sheet by uniqueID. This guarantees the ordering of the rows for both objects is the same. It follows that we can now create the columns "DNAex_round" & "PCR_round" in sd.new using our df. As sample_data from phyloseq-objects have rownames equivalent to the names of their samples we define the rownames of sd.new using the $id colums

p.comb.filt.s.1500.80 <- merge_phyloseq(otu_table(p.comb.filt.s.1500.80), tax_table(p.comb.filt.s.1500.80), sample_data(sd.new)) #we merge the otu_table and tax_table of our old phyloseq-object and the newly created sample_data - now containing the DNA extraction batch information and PCR batch information
sample_data(p.comb.filt.s.1500.80)$is.neg <- sample_data(p.comb.filt.s.1500.80)$TRUE_control == "control"#first identify your negative controls -> decontam!
```

## LibrarySize
### LibrarySize COMB - colored by TRUE/control, DNAex_round, PCR_round
```{r LibrarySize Tissue/NCT - colored by TRUE/control, DNAex_round, PCR_round }
#prepare dataframe for plotting
df <- meta(p.comb.filt.s.1500.80)
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
#prepare TRUE_control counts for plotting
true_count <- meta(p.comb.filt.s.1500.80) %>% filter(TRUE_control == "TRUE") %>% count("TRUE_control") %>% pull()
control_count <- meta(p.comb.filt.s.1500.80) %>% filter(TRUE_control == "control") %>% count("TRUE_control") %>% pull()
#plot librarysize colored by TRUE_control
ggplot(data=df, aes(x=Index, y=LibrarySize, color=TRUE_control)) + geom_point()  + theme_bw() + ggtitle("Library size p.comb.filt.s.1500.80")+
  scale_color_brewer(palette = "Set1", labels=c(paste("Negative controls\nn =", control_count, sep = " "), paste("Tissue samples\nn =", true_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare DNAex_batch counts for plotting
DNAex_batch1_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "1") %>% count("DNAex_round") %>% pull()
DNAex_batch2_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "2") %>% count("DNAex_round") %>% pull()
DNAex_batch3_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "3") %>% count("DNAex_round") %>% pull()
DNAex_batch4_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "4") %>% count("DNAex_round") %>% pull()
DNAex_batch5_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "5") %>% count("DNAex_round") %>% pull()
DNAex_batch6_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "6") %>% count("DNAex_round") %>% pull()
DNAex_batch7_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "7") %>% count("DNAex_round") %>% pull()
DNAex_batch4_5_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "4_5") %>% count("DNAex_round") %>% pull()
DNAex_batchNA_count <- meta(p.comb.filt.s.1500.80) %>% filter(DNAex_round == "NA") %>% count("DNAex_round") %>% pull()
#plot librarysize colored by DNAex_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(DNAex_round))) + geom_point()  + theme_bw() + ggtitle("Library size -  p.comb.filt.s.1500.80")+scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c(paste("Batch 1\nn =", DNAex_batch1_count, sep = " "),paste("Batch 2\nn =", DNAex_batch2_count, sep = " "),paste("Batch 3\nn =", DNAex_batch3_count, sep = " "),paste("Batch 4\nn =", DNAex_batch4_count, sep = " "),paste("Batch 4_5\nn =", DNAex_batch4_5_count, sep = " "),paste("Batch 5\nn =", DNAex_batch5_count, sep = " "),paste("Batch 6\nn =", DNAex_batch6_count, sep = " "),paste("Batch 7\nn =", DNAex_batch7_count, sep = " "),paste("Batch NA\nn =", DNAex_batchNA_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#prepare PCR_batch counts for plotting
PCR_batch1_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "1") %>% count("PCR_round") %>% pull()
PCR_batch2_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "2") %>% count("PCR_round") %>% pull()
PCR_batch3_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "3") %>% count("PCR_round") %>% pull()
PCR_batch4_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "4") %>% count("PCR_round") %>% pull()
PCR_batch5_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "5") %>% count("PCR_round") %>% pull()
PCR_batch6_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "6") %>% count("PCR_round") %>% pull()
PCR_batch7_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "7") %>% count("PCR_round") %>% pull()
PCR_batch8_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "8") %>% count("PCR_round") %>% pull()
PCR_batch9_count <- meta(p.comb.filt.s.1500.80) %>% filter(PCR_round == "9") %>% count("PCR_round") %>% pull()
#plot librarysize colored by PCR_round
ggplot(data=df, aes(x=Index, y=LibrarySize, color=as.factor(PCR_round))) + geom_point()  + theme_bw() + ggtitle("Library size - p.comb.filt.s.1500.80")+scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c(paste("Batch 1\nn =", PCR_batch1_count, sep = " "),paste("Batch 2\nn =", PCR_batch2_count, sep = " "),paste("Batch 3\nn =", PCR_batch3_count, sep = " "),paste("Batch 4\nn =", PCR_batch4_count, sep = " "),paste("Batch 5\nn =", PCR_batch5_count, sep = " "),paste("Batch 6\nn =", PCR_batch6_count, sep = " "),paste("Batch 7\nn =", PCR_batch7_count, sep = " "),paste("Batch 8\nn =", PCR_batch8_count, sep = " "),paste("Batch 9\nn =", PCR_batch9_count, sep = " ")))+
  theme (axis.text=element_text(size=14),
        axis.title=element_text(size=16,face="bold"),
        legend.text = element_text(size = 12))
#https://stackoverflow.com/questions/43359050/error-continuous-value-supplied-to-discrete-scale-in-default-data-set-example#43359104

#p.comb.filt.s.1500.80 <- subset_samples(p.comb.filt.s.1500.80, sample_nr!="P19269-20" & sample_nr!="P15970-20" & sample_nr!="P12758-20" & sample_nr!="P3434-20") #we remove 4 samples as these patients do not have buccal/ rectal swaps for later analysis
```

## DEMERGE!
```{r}
p.comb.filt.s.1500.80.true <- subset_samples(p.comb.filt.s.1500.80,TRUE_control=="TRUE") #only true samples
p.comb.filt.s.1500.80.true <- prune_taxa(x = p.comb.filt.s.1500.80.true,taxa = taxa_names(p.comb.filt.s.1500.80.true)[-(which(taxa_sums(p.comb.filt.s.1500.80.true)==0))]) #only taxa found in true samples

p.comb.filt.s.1500.80.nct <- subset_samples(p.comb.filt.s.1500.80,TRUE_control == "control") #only NCT samples
p.comb.filt.s.1500.80.nct <- prune_taxa(x = p.comb.filt.s.1500.80.nct,taxa = taxa_names(p.comb.filt.s.1500.80.nct)[-(which(taxa_sums(p.comb.filt.s.1500.80.nct)==0))]) #only taxa found in NCT samples
```

### LibrarySize - tissue samples - boxplots
```{r visualize LibrarySize of tissue samples as a boxplot(normal vs tumor) - colored by DNAex_round, PCR_round}
p.comb.filt.s.1500.80.true <- subset_samples(p.comb.filt.s.1500.80,TRUE_control=="TRUE")
#prepare dataframe for plotting
df <- meta(p.comb.filt.s.1500.80.true) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80.true)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))

#set levels
df$sample_side <- factor(df$sample_side)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$sample_side), 2)),seq(nrow(t(combn(levels(df$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df, aes(x = sample_side, y = LibrarySize)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by sample side") +
  theme(text = element_text(size = 10))

#set levels
df$DNAex_round <- factor(df$DNAex_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$DNAex_round), 2)),seq(nrow(t(combn(levels(df$DNAex_round), 2)))))
#boxplot comparing librarysizes of different DNAex batches
ggplot(df, aes(x = DNAex_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "DNA extraction round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.02,comparisons = comb, map_signif_level = FALSE,y_position = seq.int(55000,200000,8000),color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by DNA extraction batch") +
    theme(text = element_text(size = 10))

#set levels
df$PCR_round <- factor(df$PCR_round)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df$PCR_round), 2)),seq(nrow(t(combn(levels(df$PCR_round), 2)))))
#boxplot comparing librarysizes of different PCR batches
ggplot(df, aes(x = PCR_round, y = LibrarySize, color = sample_side)) +
    geom_boxplot(outlier.shape = NA) +
    geom_jitter(width = 0.2) +
    labs(x = "PCR round") + 
    geom_signif(tip_length = 0.02,step_increase = 0.1,comparisons = comb, map_signif_level = FALSE,color = "black") +
  ggtitle("LibrarySize of tissue samples grouped by PCR extraction batch") +
    theme(text = element_text(size = 10))
```

We compare librarysizes of true samples via boxplots. 
Plot1: LibSizes of tissue samples grouped by sample side 
Plot2: LibSizes of tissue samples grouped by DNA extraction batch
Plot3: LibSizes of tissue samples grouped by PCR batch
Conclusion: Plot 1 scores a pvalue = 0.066 = 6.6% - i.e. just slightly not significant. In Plot2 and Plot3 we can clearly see the main driver of any differences in librarysizes is DNA batch 6 (Plot 2), PCR batch 8 (Plot 3). Both of these batches represent the same tissue samples (uniqueID: 326 - 335). Combining this information with the microbial composition of these 10 samples yields the conclusion that high amounts of contaminants are present in this batch. 
Visual intra-batch (not DNA extraction batch 6 / PCR batch 8) comparisons indicate no significant differences between librarysizes for normal samples versus tumor samples (i.e. Plot3: take PCR batch 1 and visually compare normal samples versus tumor samples)
Small note: Plot 2 and Plot 3 show that DNA extraction batch 6 / PCR batch 8 are made up of only tumor samples -> bad study design!

Building on our above conclusions: we now go one step further and compare normal samples versus tumor samples for each batch (DNA extraction, PCR) on a statistical basis - if possible (ie PCR batch 8 only contains tumor samples)
### Boxplot of tissue samples´LibrarySize comparing sample_side per DNA extraction batch, PCR batch
```{r Boxplot of tissue samples´LibrarySize comparing sample_side per DNA extraction batch, PCR batch}
#prepare dataframe for plotting
df <- meta(p.comb.filt.s.1500.80.true) # Put sample_data into a ggplot-friendly data.frame
df$LibrarySize <- sample_sums(p.comb.filt.s.1500.80.true)
df <- df[order(df$LibrarySize),]
df$Index <- seq(nrow(df))
#DNA
#now we need to subset the dataframe for each batch seperatly
#DNA batch 1
df_mod <- df %>% filter(DNAex_round == "1")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = DNAex_round)) +
  ggtitle("LibrarySize of DNA extraction batch 1: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
#DNA batch 2
df_mod <- df %>% filter(DNAex_round == "2")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb: vector containing groups for groupwise comparisons
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = DNAex_round)) +
  ggtitle("LibrarySize of DNA extraction batch 2: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
#DNA batch 3
df_mod <- df %>% filter(DNAex_round == "3")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = DNAex_round)) +
  ggtitle("LibrarySize of DNA extraction batch 3: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
#DNA batch 4
df_mod <- df %>% filter(DNAex_round == "4")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = DNAex_round)) +
  ggtitle("LibrarySize of DNA extraction batch 4: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
#DNA batch 5
df_mod <- df %>% filter(DNAex_round == "5")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = DNAex_round)) +
  ggtitle("LibrarySize of DNA extraction batch 5: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
#PCR
#now we need to subset the dataframe for each batch seperatly
df_mod <- df %>% filter(PCR_round == "1")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = PCR_round)) +
  ggtitle("LibrarySize of PCR batch 1: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))
  
#now we need to subset the dataframe for each batch seperatly
df_mod <- df %>% filter(PCR_round == "2")
#set levels
df_mod$sample_side <- factor(df_mod$sample_side)
#create comb
comb <- split(t(combn(levels(df_mod$sample_side), 2)),seq(nrow(t(combn(levels(df_mod$sample_side), 2)))))
#boxplot comparing librarysizes of normal vs tumor samples
ggplot(df_mod, aes(x = sample_side, y = LibrarySize, color = PCR_round)) +
  ggtitle("LibrarySize of PCR batch 2: normal vs tumor") +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter(width = 0.2) +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black") +
  theme(text = element_text(size = 10))

#shorter code version using facet_wrap
data <- reshape2::melt(df,measure.vars = c("LibrarySize"),value.name = c("LibrarySize"))
data$sample_side <- factor(data$sample_side)
comb <- split(t(combn(levels(data$sample_side), 2)),seq(nrow(t(combn(levels(data$sample_side), 2)))))
ggplot(data, aes(x = sample_side,y = LibrarySize)) +geom_boxplot(outlier.shape = NA) + 
    geom_jitter(width = 0.2) + labs(x = "sample type") +
    geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black",y_position = c(14000)) + facet_wrap(vars(DNAex_round), scales = "free_y") +
    ggtitle("LibrarySize - DNA extraction batches grouped by sample side")

ggplot(data, aes(x = sample_side,y = LibrarySize)) +geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE, color = "black",step_increase = 0.1) + facet_wrap(vars(PCR_round), scales = "free_y") + 
  ggtitle("LibrarySize - PCR batches grouped by sample side")
#notice DNAex batch 6 / PCR batch 8 have no normal sample type 
#helpful https://www.statology.org/facet_wrap/
```

Plot1: LibrarySize of DNA extraction batch 1 grouped by sample side
Plot2: LibrarySize of DNA extraction batch 2 grouped by sample side
Plot3: LibrarySize of DNA extraction batch 3 grouped by sample side
Plot4: LibrarySize of DNA extraction batch 4 grouped by sample side
Plot5: LibrarySize of DNA extraction batch 5 grouped by sample side
Plot6: LibrarySize of PCR batch 1 grouped by sample side
Plot7: LibrarySize of PCR batch 2 grouped by sample side

As we expected from visualizing LibrarySize of tissue samples comparing normal tissue vs. tumor tissue - coloring by DNA extraction batch/ PCR batch (Plots 2 and 3) - the differences between LibrarySize for sample side in each batch are not significant.

## Analysis of prevalence
-### Decontam classification
```{r}
noncontamdf.prev <- isNotContaminant(p.comb.filt.s.1500.80, method = "prevalence", neg ="is.neg",threshold = 0.5,detailed = TRUE)
#threshold of 0.5 classifies all taxa present in more negative controls then tisse samples as contaminant
#intersting would be the comparison of this classification and our NEJMAN derived classification method
hist(noncontamdf.prev$p, 100, ylim = c(0,300), xlim = c(0,1))
```


### Prev.true vs. prev.ntc
```{r}
# Make phyloseq object of presence-absence in negative controls and true samples
ps.pa <- transform_sample_counts(p.comb.filt.s.1500.80, function(abund) 1*(abund>0))
ps.pa.neg <- prune_samples(sample_data(ps.pa)$TRUE_control == "control", ps.pa)
ps.pa.pos <- prune_samples(sample_data(ps.pa)$TRUE_control == "TRUE", ps.pa)
# Make data.frame of prevalence in positive and negative samples
df.pa <- psmelt(ps.pa)
df.pa$pa.pos <- taxa_sums(ps.pa.pos)
df.pa$pa.neg <- taxa_sums(ps.pa.neg)

ggplot(data=df.pa, aes(x=pa.neg/nsamples(ps.pa.neg), y=pa.pos/nsamples(ps.pa.pos))) + geom_point() +
  xlab("Prevalence (Negative Controls)") + ylab("Prevalence (True Samples)")

#compare this to DECONTAM github: https://benjjneb.github.io/decontam/vignettes/decontam_intro.html
```




## Microbial composition
### Microbial composition - species level - tissue samples - grouped by DNA, PCR batch
```{r Microbial composition of tissue samples, DNA extraction batch, PCR batch}

physeq.fam <- transform_sample_counts(p.comb.filt.s.1500.80.true, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$species.v2 <- ifelse(df.fam.melt$Abundance>0.04,as.character(df.fam.melt$species), "others") # define cut-off for species to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$species.v2 <- factor(df.fam.melt$species.v2, levels=rev(unique(df.fam.melt$species.v2)))
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at species level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

#microbial composition at species level - grouped by DNA Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level - grouped by DNA Batch")+
  facet_wrap(vars(DNAex_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

#microbial composition at species level - grouped by PCR Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=species.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at species level - grouped by PCR Batch")+
  facet_wrap(vars(PCR_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```

### Microbial Composition at genus level
```{r}
physeq.fam <- transform_sample_counts(p.comb.filt.s.1500.80.true, function(x){x/sum(x)}) #transform total read counts into per sample relative counts
df.fam.melt <- psmelt(physeq.fam) # use psmelt for generating a dataframe -> useful for ggplot

df.fam.melt$genus.v2 <- ifelse(df.fam.melt$Abundance>0.01,as.character(df.fam.melt$genus), "others") # define cut-off for genus to be named or else be grouped into "others
#ploting abundance > 5% | 1% requires 161 colors
df.fam.melt$genus.v2 <- factor(df.fam.melt$genus.v2, levels=rev(unique(df.fam.melt$genus.v2)))
header2 <- c("normal" = "Normal pancreas","tumor" = "Tumor tissue") # for plot
cols <- c("#9d547c","#56ca63","#a357d6","#419d2a","#525fd6","#8cbe3a","#c944aa","#5ba557","#9e66cb","#c1b735","#6d82ec","#e69728","#6654b0","#799330","#da7fdf","#3c782c","#e44586","#63c996","#dc3f53","#49cbc8","#cf3f29","#4fabda","#da6c2b","#598bd1","#b78c24","#8d4191","#a0b971","#b2386a","#479d71","#ae4341","#2ba198","#e07557","#5361a3","#dda353","#aa98df","#5b6114","#dc89bf","#327243","#e57b94","#277257","#9b62a0","#bbab59","#98495a","#526229","#d8827d","#857624","#9a4a22","#7c7d46","#e3a073","#9e6b33", "gray74", "lawngreen", "orchid2", "olivedrab1", "ivory3", "darkseagreen", "bisque2", "darkgoldenrod2", "blue2", "skyblue", "seashell2", "turquoise", "tan1", "seagreen2", "palevioletred3", "linen", "steelblue4", "limegreen", "purple3", "khaki3", "snow3", "darkslategray", "lavender", "magenta2", "palegreen", "salmon", "maroon", "cyan2","#671408","#FAEBD7","#7FFFD4","#F0FFFF","#A52A2A","burlywood","cadetblue","#7FFF00","chocolate","cornsilk","#FF7F50","#008B8B","darkgoldenrod1","darkolivegreen","darkorange4","darkslategray3","navajowhite3","orchid4","gray25","#F0924D")
#microbial composition at genus level
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  facet_wrap(.~sample_side, scales = "free", labeller = as_labeller(header2))+
  ggtitle("Microbial composition at genus level")+
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
#microbial composition at genus level - grouped by DNA Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at genus level - grouped by DNA Batch")+
  facet_wrap(vars(DNAex_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))

#microbial composition at genus level - grouped by PCR Batch
ggplot(df.fam.melt, aes(x=id, y=Abundance, fill=genus.v2)) +
  geom_bar(stat="identity") +
  scale_y_continuous(labels = scales::percent, expand = c(0, 0), limits = c(0, 1.02)) +
  scale_fill_manual(values = cols) +
  xlab ("") +
  ylab("Relative abundance") +
  theme_bw() +
  ggtitle("Microbial composition at genus level - grouped by PCR Batch")+
  facet_wrap(vars(PCR_round), scales = "free_y") +
  theme( axis.text.x = element_blank(),
        axis.text.y = element_text (size=12),
        axis.title = element_text(size=14, face="bold"))+
  theme(legend.position = "bottom",
        legend.title = element_blank(),
        legend.background = element_rect(size=0.25, linetype="solid", colour ="black"),
        legend.key.size = unit(4,"mm"))
```

## Alpha diversity
```{r}
aindex <- estimate_richness(physeq = otu_table(p.comb.filt.s.1500.80.true), measures = c("Observed","Shannon","Simpson","InvSimpson", "Chao1","ACE"))

aindex$sample_side <- meta(p.comb.filt.s.1500.80.true)$sample_side
aindex$sample_side <- factor(aindex$sample_side)

comb <- split(t(combn(levels(aindex$sample_side), 2)),seq(nrow(t(combn(levels(aindex$sample_side), 2)))))
data <- reshape2::melt(aindex)
ggplot(data, aes(x = sample_side,y = value)) +
  # Outliers are removed, because otherwise each data point would be plotted twice; 
  # as an outlier of boxplot and as a point of dotplot.
  geom_boxplot(outlier.shape = NA) + 
  geom_jitter(width = 0.2) + 
  ggtitle("raw data - alpha diversity") +
  labs(x = "sample type") +
  geom_signif(comparisons = comb, map_signif_level = FALSE,) +
  theme(text = element_text(size = 10)) +
  facet_wrap(~ variable, scales = "free_y")

plot_richness(physeq = p.comb.filt.s.1500.80.true,x = "sample_side")
plot_richness(physeq = p.comb.filt.s.1500.80.true,x = "sample_side",color = "PCR_round")
plot_richness(physeq = p.comb.filt.s.1500.80.true,x = "sample_side",color = "DNAex_round")
```

## Beta diversity
### Beta diversity - tissue samples - sample side
```{r TRUE - BrayCurtis - sample side}
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.true, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.true, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.true)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.true, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix of p.comb.filt.s.1500.80.true, compairing tumor vs. normal
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors = sample_data(p.comb.filt.s.1500.80.true)$sample_side)
p.adonis.pairwise
```

### Beta diversity - tissue samples - DNA extraction batch
```{r TRUE - BrayCurtis - DNAex batches}
#Bray-Curtis for True samples only
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.true, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.true, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.true)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.true, bc_ordination, color="DNAex_round") + 
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 4_5","Batch 5","Batch 6","Batch 7","Batch NA"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix - pairwise compairing DNA extraction batches
#p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.true)$DNAex_round)
#p.adonis.pairwise
```

### Beta diversity - tissue samples - PCR batch
```{r TRUE - BrayCurtis - PCR batches}
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.true, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.true, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.true)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.true, bc_ordination, color="PCR_round") + 
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 5","Batch 6","Batch 7","Batch 8","Batch 9"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("True samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix of p.comb.filt.s.1500.80.true, pairwise compairing PCR batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.true)$PCR_round)
p.adonis.pairwise
```

### Beta diversity - NCT samples - sample_side
```{r NCT - BrayCurtis - sample side}
p.comb.filt.s.1500.80.nct <- subset_samples(p.comb.filt.s.1500.80,TRUE_control == "control")
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.nct, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.nct, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.nct)$sample_side)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.nct, bc_ordination, color="sample_side") + 
  geom_point(aes(colour=sample_side), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("NCT samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix of p.comb.filt.s.1500.80.nct, pairwise compairing sample_side
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.nct)$sample_side)
p.adonis.pairwise
```

### Beta diversity - NCT samples - DNA extraction batch
```{r NCT - BrayCurtis - DNAex batches}
p.comb.filt.s.1500.80.nct <- subset_samples(p.comb.filt.s.1500.80,TRUE_control == "control")
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.nct, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.nct, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.nct)$DNAex_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.nct, bc_ordination, color="DNAex_round") + 
  geom_point(aes(colour=DNAex_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="DNA batch",breaks=c("1","2","3","4","4_5","5","6","7","NA"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 4_5","Batch 5","Batch 6","Batch 7","Batch NA"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("NCT samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix of p.comb.filt.s.1500.80.true, pairwise compairing DNA extraction batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.nct)$DNAex_round)
p.adonis.pairwise
```

### Beta diversity - NCT samples - PCR batch
```{r NCT - BrayCurtis - PCR batches}
bray_dist = phyloseq::distance(p.comb.filt.s.1500.80.nct, method="bray")
bc_ordination = ordinate(p.comb.filt.s.1500.80.nct, method="PCoA", distance=bray_dist)
pcoa1 <- paste("PCoA 1", " [", round(bc_ordination[[3]]$Relative_eig[1], digits = 3)*100, "%]", sep = "")
pcoa2 <- paste("PCoA 2", " [", round(bc_ordination[[3]]$Relative_eig[2], digits = 3)*100, "%]", sep = "")
p.adonis <- adonis2(bray_dist ~ sample_data(p.comb.filt.s.1500.80.nct)$PCR_round)
p <- paste0("p = ",p.adonis$`Pr(>F)`[1])
annotations <- data.frame(
        xpos = c(-Inf),
        ypos =  c(Inf),
        annotateText = p,
        hjustvar = c(-0.2),
        vjustvar = c(1.5))
plot_ordination(p.comb.filt.s.1500.80.nct, bc_ordination, color="PCR_round") + 
  geom_point(aes(colour=PCR_round), size=3) +
  theme(aspect.ratio=1) +
  theme_bw()+
 scale_colour_discrete(name="PCR batch",breaks=c("1","2","3","4","5","6","7","8","9"),labels=c("Batch 1","Batch 2","Batch 3","Batch 4","Batch 5","Batch 6","Batch 7","Batch 8","Batch 9"))+
  stat_ellipse() +
  xlab(pcoa1)+
  ylab(pcoa2)+
  theme(panel.grid =  element_blank())+
  ggtitle("NCT samples - Bray curtis") +
  theme (axis.text=element_text(size=14),
         axis.title=element_text(size=16,face="bold"))+
  theme(
        # legend.justification=c(1.15,-0.15), legend.position=c(1,0),
        legend.text = element_text(size = 12))+
  geom_text(data=annotations,aes(x=xpos,y=ypos,hjust=hjustvar,vjust=vjustvar,label=annotateText), inherit.aes = FALSE)

#introduce Pairwise.adonis for dissimilarity matrix of p.comb.filt.s.1500.80.true, pairwise compairing PCR batches
p.adonis.pairwise <- pairwise.adonis(x = bray_dist,factors =sample_data(p.comb.filt.s.1500.80.nct)$PCR_round)
p.adonis.pairwise
```